---
title: "Estimation of VAF of tumors across multiple timepoints for the PBTA Cohort"
author: 'Antonia Chroni <chronia@chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---


#### Tumor evolution project 

# Background
We are investigating the change of VAFs across multiple timepoints in the PBTA cohort.

### Data used 
In this notebook, we are looking into the v12 histologies file (last updates on May 1st, 2023) and subset to the PBTA cohort.
We use the output files from the "analyses/sample-distribution-analysis/results".

### Usage
This notebook is intended to be run via the command line from the directory of the module as follows:

```
Rscript -e "rmarkdown::render('01-process-data.Rmd', clean = TRUE)"
```

# Set up

```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
  library(plyr)
})
```

## Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
scratch_dir <- file.path(root_dir, "scratch")
data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "create-corplots") # create analysis_dir
input_dir <- file.path(analysis_dir, "input") # create input_dir
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results") # create files_dir

# Input files
v12_file <- file.path(data_dir, "histologies.tsv")
maf_file <- file.path(data_dir, "snv-consensus-plus-hotspots.maf.tsv.gz")
tmb_file <- file.path(input_dir, "snv-mutation-tmb-coding.tsv") # tmb file
genomic_assays_matched_time_points_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") #  generated from the add-sample-distribution analysis
```

## Read in data and process

```{r load-process-inputs-please-wait}
# Make a vector of coding gene mutations
maf_nonsynonymous <- c(
  "Missense_Mutation",
  "Frame_Shift_Del",
  "In_Frame_Ins",
  "Frame_Shift_Ins",
  "Splice_Site",
  "Nonsense_Mutation",
  "In_Frame_Del",
  "Nonstop_Mutation",
  "Translation_Start_Site"
)

# Read in maf file, calculate VAF, and retain only coding gene mutations
maf <- data.table::fread(maf_file, data.table = F) %>% 
  filter(Tumor_Sample_Barcode %in% pbta$Kids_First_Biospecimen_ID) %>%
  dplyr::rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>% # change the column name of the biospecimen column to match the column name from the histologies files
  dplyr::mutate(VAF = t_alt_count / (t_ref_count + t_alt_count)) %>% # calculate VAF 
  mutate(gene_protein = paste(Hugo_Symbol, HGVSp_Short, sep = "_")) %>%  # concatenate gene and protein information
  filter(Variant_Classification %in% maf_nonsynonymous) %>%
  select(Kids_First_Biospecimen_ID, Hugo_Symbol, Chromosome, Start_Position, Variant_Classification, VAF, HGVSc, HGVSp, HGVSp_Short, gene_protein) 

# Read in tmb file
tmb <- readr::read_tsv(tmb_file, guess_max = 100000, show_col_types = FALSE)

# Read in histologies file and filter for the pbta cohort
pbta <- readr::read_tsv(v12_file, guess_max = 100000, show_col_types = FALSE) %>%
  filter(cohort == "PBTA",
         !is.na(pathology_diagnosis),
         !composition %in% c("Derived Cell Line", "PDX"),
         !experimental_strategy %in% c("Methylation", "Targeted Sequencing")) %>%
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, experimental_strategy, tumor_descriptor, age_at_diagnosis_days, age_last_update_days, cancer_predispositions, broad_histology, short_histology, cancer_group) 

# Read in genomic_assays_matched_time_points file and get the list of patients with matched time points
patient_list <- readr::read_tsv(genomic_assays_matched_time_points_file, guess_max = 100000, show_col_types = FALSE) %>%
                                      subset(select = c("Kids_First_Participant_ID", "descriptors"))

# How many patients have paired time samples?
print(length(unique(patient_list$Kids_First_Participant_ID)))

# Add metadata from pbta to patient list and keep only autopsy samples
patient_list_metadata <- patient_list %>% 
  left_join(pbta, by = "Kids_First_Participant_ID") %>%
  filter(!(experimental_strategy == "RNA-Seq")) %>%
  mutate(tumor_descriptor = case_when(grepl("Primary Tumor", tumor_descriptor) ~ "Diagnosis",
                                 grepl("Initial CNS Tumor", tumor_descriptor) ~ "Diagnosis",
                                 grepl('Progressive Disease Post-Mortem', tumor_descriptor) ~ 'Deceased', 
                                 TRUE ~ tumor_descriptor)) %>%
  mutate(Deceased = case_when(grepl("Deceased", descriptors) ~ "Deceased")) %>%
  filter(Deceased == "Deceased") 

# How many patients have paired time samples with autopsy time point present?
print(length(unique(patient_list_metadata$Kids_First_Participant_ID)))
```

# Create maf file for autopsy samples

```{r maf_autopsy}
# There are cases with more than two paired time points. 
# We will need to create a column to distinguish for those cases (descriptors_new_number). 
# This will allow us to filter and create multiple corplots per patient.
maf_autopsy <- patient_list_metadata %>% 
  left_join(maf, by=c('Kids_First_Biospecimen_ID')) %>%
  filter(!is.na(Kids_First_Participant_ID),
         !is.na(VAF),
         !VAF == 0) %>%
  mutate(descriptors_new = case_when(descriptors == "Deceased, Diagnosis, Diagnosis, Progressive, Progressive" ~ "dx_pro",
                                      descriptors == "Deceased, Diagnosis, Recurrence" ~ "dx_rec",
                                      descriptors %in% c("Deceased, Progressive, Recurrence", "Deceased, Progressive, Progressive, Recurrence", "Deceased, Progressive, Recurrence, Recurrence") ~ "pro_rec",
                                      descriptors == "Deceased, Diagnosis, Progressive, Recurrence" ~ "dx_pro_rec",
                                      descriptors %in% c("Deceased, Diagnosis", "Deceased, Diagnosis, Diagnosis", "Deceased, Diagnosis, Diagnosis, Diagnosis, Diagnosis") ~ "dx", 
                                      descriptors %in% c("Deceased, Progressive", "Deceased, Progressive, Progressive") ~ "pro",
                                      descriptors == "Deceased, Recurrence" ~ "rec")) %>%
  mutate(descriptors_new_number = case_when(descriptors_new %in% c("dx_pro", "dx_rec", "pro_rec") ~ "2",
                                            descriptors_new == "dx_pro_rec" ~ "3",
                                            descriptors_new %in% c("dx", "pro", "rec") ~ "1")) %>%
  write_tsv(file.path(scratch_dir, "maf_autopsy.tsv"))
```

```{r echo=TRUE}
sessionInfo()
```