---
title: "Estimation of VAFs of tumors across multiple timepoints for the PBTA Cohort"
author: 'Antonia Chroni <chronia@chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---


#### Tumor evolution project 

# Background
We are investigating the change of VAFs across multiple timepoints in the PBTA cohort.

### Data used 
In this notebook, we are looking into the v12 histologies file (last updates on May 1st, 2023) and subset to the PBTA cohort.
We use the output files from the "analyses/sample-distribution-analysis/results".

### Method used 


### Usage

This notebook is intended to be run via the command line from the top directory of the repository as follows:

```
Rscript -e "rmarkdown::render('1-process-data_VAF.Rmd', clean = TRUE)"
```

# Set up

```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(flextable)
  library(plyr)
  library(ggplot2)
  library(ggpubr)
  library(ggrepel)
})
```

## Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no
# matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
setwd(root_dir)

analysis_dir <- file.path(root_dir, "analyses", "create-corplots")

# File path to input directory
input_dir <- file.path(analysis_dir, "input")
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")
data_dir <- file.path(root_dir, "data")

# Inputs
maf <- file.path(data_dir, "snv-consensus-plus-hotspots.maf.tsv.gz")

# The following files were generated from the add-sample-distribution analysis
v12 <- file.path(data_dir, "histologies.tsv")
input_genomic_df_filter <- file.path(files_dir, "genomic_df_filter.tsv")
input_tmb <- file.path(input_dir, "snv-mutation-tmb-coding.tsv")

# File path to results directory
results_dir <-
  file.path(analysis_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# File path to results directory
plots_dir <-
  file.path(analysis_dir, "plots")
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

source(paste0(root_dir, "/figures/theme.R"))
```

## Load and Process data
```{r load-inputs-please-wait}
## Read MAF file
snv <- data.table::fread(maf, data.table = F) %>%
  filter(Tumor_Sample_Barcode %in% pbta_histologies$Kids_First_Biospecimen_ID) %>%
  dplyr::rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>% # We will first change the column name of the biospecimen column to match the column name from the histologies files.
  dplyr::mutate(VAF = t_alt_count / (t_ref_count + t_alt_count)) %>% # using formula for VAF calculation
  select(Kids_First_Biospecimen_ID, Hugo_Symbol, Chromosome, Start_Position, Variant_Classification, VAF)

# pbta cohort
pbta <- readr::read_tsv(v12, guess_max = 100000, show_col_types = FALSE) %>%
  filter(cohort == "PBTA",
         !is.na(pathology_diagnosis),
         !composition %in% c("Derived Cell Line", "PDX"),
         !experimental_strategy %in% c("Methylation", "Targeted Sequencing")) 

## Read in genomic_df_filter file
genomic_df_filter <- readr::read_tsv(input_genomic_df_filter, guess_max = 100000, show_col_types = FALSE)

# Read file with TMB information
tmb <- readr::read_tsv(input_tmb, guess_max = 100000, show_col_types = FALSE)
```

# Process Data: snv 

```{r echo=TRUE}
# Merge genomic_df_filter with pbta_subset to get sample_id information
# Create match_id2 to count for multiple samples per patient, if any
genomic_df_filter_merge <- genomic_df_filter %>% 
  left_join(pbta, by=c('Kids_First_Participant_ID'))

genomic_df_subset <- genomic_df_filter_merge %>%
  subset(select = c(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, sample_id, composition, match_id, tumor_descriptor, descriptors, cancer_group.x, experimental_strategy.y)) %>%
  filter(!(experimental_strategy.y == "RNA-Seq")) %>%
  mutate(match_id2 = paste(Kids_First_Participant_ID, sample_id, composition, sep = "_")) %>%
  mutate(tumor_descriptor = case_when(grepl("Primary Tumor", tumor_descriptor) ~ "Diagnosis",
                                 grepl("Initial CNS Tumor", tumor_descriptor) ~ "Diagnosis",
                                 grepl('Progressive Disease Post-Mortem', tumor_descriptor) ~ 'Deceased', 
                                 TRUE ~ tumor_descriptor)) %>%
  write_tsv(file.path(results_dir, "genomic_df_subset.tsv"))

genomic_df_subset %>%
  group_by(Kids_First_Participant_ID, match_id2, descriptors) %>%
  tally() %>%
  regulartable() %>%
  fontsize(size = 12, part = "all")
```


```{r echo=TRUE}
#maf_files <- snv_vaf_subset %>% 
#  left_join(genomic_df_subset, by=c('Kids_First_Biospecimen_ID')) %>%
#  filter(!is.na(Kids_First_Participant_ID))
  
# Create dir to save MAF files
#temp_path_csv <-
#  file.path(results_dir, "01_MAF_files_Per_Patient")
#if (!dir.exists(temp_path_csv)) {
#  dir.create(temp_path_csv)
#}

# we will change the dir here otherwise files are saved outside of the desired folder
#setwd(temp_path_csv)

# we will split masf data by the "Kids_First_Participant_ID.x" column
#maf_files %>% 
#  group_split(Kids_First_Participant_ID) %>% 
#  walk(~write_csv(.x, paste0(.x$Kids_First_Participant_ID[1], "-vaf.csv")))

# let's change back to the root_dir for the rest of the module
#setwd(root_dir)
```

# Autopsy samples 
We will focus on patient cases containing autopsy samples and save these maf files separately.
We will do this for different pairs.


```{r echo=TRUE}
# Create dir to save MAF files separately for each patient case as well
scratch <-
  file.path(results_dir, "scratch")
if (!dir.exists(scratch)) {
  dir.create(scratch)
}

# we will change the dir here otherwise files are saved outside of the desired folder
setwd(scratch)

maf_files_Deceased <- snv_vaf_subset %>% 
  left_join(genomic_df_subset, by=c('Kids_First_Biospecimen_ID')) %>%
  filter(!is.na(Kids_First_Participant_ID)) %>%
  mutate(Deceased = case_when(grepl("Deceased", descriptors) ~ "Deceased")) %>%
  filter(Deceased == "Deceased") %>%
  write_tsv(file.path(scratch, "maf_files_Deceased.tsv"))


##########################################################################################
# Storing all Dx-Deceased Sample pairs in a list so each file can be processed iteratively
scratch_Dx_Dec <-
  file.path(scratch, "Dx_Dec")
if (!dir.exists(scratch_Dx_Dec)) {
  dir.create(scratch_Dx_Dec)
}

setwd(scratch_Dx_Dec)
data <- maf_files_Deceased %>% 
  mutate(data = case_when(grepl("Deceased, Diagnosis", descriptors) ~ "Dx_Dec")) %>%
  filter(data == "Dx_Dec")

# we will split maf data by the "Kids_First_Participant_ID.x" column
data %>% 
  group_split(Kids_First_Participant_ID) %>% 
  walk(~write_csv(.x, paste0(.x$Kids_First_Participant_ID[1], "-vaf.csv")))


##########################################################################################
# Storing all Pro_Dec Sample pairs in a list so each file can be processed iteratively
scratch_Pro_Dec <-
  file.path(scratch, "Pro_Dec")
if (!dir.exists(scratch_Pro_Dec)) {
  dir.create(scratch_Pro_Dec)
}

setwd(scratch_Pro_Dec)
data <- maf_files_Deceased %>% 
  filter(descriptors %in% c("Deceased, Progressive", "Deceased, Progressive, Progressive"))

# we will split maf data by the "Kids_First_Participant_ID.x" column
data %>% 
  group_split(Kids_First_Participant_ID) %>% 
  walk(~write_csv(.x, paste0(.x$Kids_First_Participant_ID[1], "-vaf.csv")))


##########################################################################################
# Storing all Deceased, Recurrence Sample pairs in a list so each file can be processed iteratively
scratch_Rec_Dec <-
  file.path(scratch, "Rec_Dec")
if (!dir.exists(scratch_Rec_Dec)) {
  dir.create(scratch_Rec_Dec)
}

setwd(scratch_Rec_Dec)
data <- maf_files_Deceased %>% 
  filter(descriptors == "Deceased, Recurrence")

# we will split maf data by the "Kids_First_Participant_ID.x" column
data %>% 
  group_split(Kids_First_Participant_ID) %>% 
  walk(~write_csv(.x, paste0(.x$Kids_First_Participant_ID[1], "-vaf.csv")))

##########################################################################################  
# Storing all Dx_Pro_Dec Sample pairs in a list so each file can be processed iteratively
#scratch_Dx_Pro_Dec <-
#  file.path(scratch, "Dx_Pro_Dec")
#if (!dir.exists(scratch_Dx_Pro_Dec)) {
#  dir.create(scratch_Dx_Pro_Dec)
#}

#setwd(scratch_Dx_Pro_Dec)
#data <- maf_files_Deceased %>% 
#  mutate(data = case_when(grepl("Deceased, Diagnosis, Diagnosis, Progressive, Progressive", descriptors) ~ "Dx_Pro_Dec")) %>%
# filter(data == "Dx_Pro_Dec")

# we will split maf data by the "Kids_First_Participant_ID.x" column
#data %>% 
#  group_split(Kids_First_Participant_ID) %>% 
#  walk(~write_csv(.x, paste0(.x$Kids_First_Participant_ID[1], "-vaf.csv")))


##########################################################################################
# Storing all Dx_Rec_Dec Sample pairs in a list so each file can be processed iteratively
#scratch_Dx_Rec_Dec <-
#  file.path(scratch, "Dx_Rec_Dec")
#if (!dir.exists(scratch_Dx_Rec_Dec)) {
#  dir.create(scratch_Dx_Rec_Dec)
#}

#setwd(scratch_Dx_Rec_Dec)
#data <- maf_files_Deceased %>% 
#  mutate(data = case_when(grepl("Deceased, Diagnosis, Recurrence", descriptors) ~ "Dx_Rec_Dec")) %>%
#  filter(data == "Dx_Rec_Dec")

# we will split maf data by the "Kids_First_Participant_ID.x" column
#data %>% 
#  group_split(Kids_First_Participant_ID) %>% 
#  walk(~write_csv(.x, paste0(.x$Kids_First_Participant_ID[1], "-vaf.csv")))


##########################################################################################
# Storing all Pro_Rec_Dec Sample pairs in a list so each file can be processed iteratively
#scratch_Pro_Rec_Dec <-
#  file.path(scratch, "Pro_Rec_Dec")
#if (!dir.exists(scratch_Pro_Rec_Dec)) {
#  dir.create(scratch_Pro_Rec_Dec)
#}

#setwd(scratch_Pro_Rec_Dec)
#data <- maf_files_Deceased %>% 
# mutate(data = case_when(grepl("Deceased, Progressive, Recurrence", descriptors) ~ "Pro_Rec_Dec")) %>%
#  filter(data == "Pro_Rec_Dec")

# we will split maf data by the "Kids_First_Participant_ID.x" column
#data %>% 
#  group_split(Kids_First_Participant_ID) %>% 
# walk(~write_csv(.x, paste0(.x$Kids_First_Participant_ID[1], "-vaf.csv")))
##########################################################################################
# let's change back to the root_dir for the rest of the module
setwd(root_dir)
```

## VAFs across all Matched Samples: Violin_plot

```{r Violin_plot}
# set the order for the x axis
maf_files_Deceased$tumor_descriptor <- factor(x = maf_files_Deceased$tumor_descriptor, levels = c("Diagnosis","Progressive", "Recurrence", "Deceased"))
#my_comparisons = list( c("Diagnosis","Progressive"), c("Diagnosis", "Recurrence"), c("Diagnosis", "Deceased"),
#                       c("Progressive", "Recurrence"), c("Progressive", "Deceased"),
#                       c("Recurrence", "Deceased"))

print(ggplot(maf_files_Deceased, aes(x= tumor_descriptor, y= VAF, fill = tumor_descriptor)) + 
  geom_violin() +
  #stat_compare_means(comparisons = my_comparisons, label.y = c(29, 35, 40))+ # Add pairwise comparisons p-value
  stat_compare_means(label.y = 55)  +   # Add global p-value
  stat_compare_means(method = "wilcox.test")+ # Add significance levels
  theme_Publication() +
  scale_y_continuous(limits=c(0,1))) +
  labs(title = "", x = "Time points", y = "VAF")

ggsave(filename = "Violin_plot_tumor_descriptor_Deceased.pdf", path = plots_dir,, width=7 ,height =5 ,device = "pdf")
```

## Mutations across all Matched Samples: Stacked_barplot

In the following there are Kids_First_Biospecimen_ID missing in the TMB file so they wont show in the df and plot.

```{r Stacked_barplot}
colnames(tmb)[which(names(tmb) == "Tumor_Sample_Barcode")] <- "Kids_First_Biospecimen_ID"

genomic_df_subset_Deceased <- genomic_df_subset %>%
  mutate(Deceased = case_when(grepl("Deceased", descriptors) ~ "Deceased")) %>%
  filter(Deceased == "Deceased")

tmb_merge <- tmb %>% 
  left_join(genomic_df_subset_Deceased , by=c('Kids_First_Biospecimen_ID')) %>%
  filter(!is.na(tumor_descriptor))   %>%
  write_tsv(file.path(scratch, "tmb_merge.tsv"))

# set the order for the x axis
tmb_merge$tumor_descriptor <- factor(x = tmb_merge$tumor_descriptor, levels = c("Diagnosis","Progressive", "Recurrence", "Deceased"))

print(ggplot(tmb_merge, aes(x = factor(Kids_First_Participant_ID), y = mutation_count, fill = tumor_descriptor))+  
  geom_col(position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = c("gray", "pink", "firebrick3", "dodgerblue3")) +
  geom_bar( stat="identity", width = 0.9) + 
  theme_Publication() + 
  theme(axis.text.x = element_text(angle = 85, hjust = 1, vjust =1)) + 
  scale_y_continuous(limits=c(0,12800)) +
  labs(title = "", x = "Kids_First_Participant_ID", y = "Total Mutations"))

ggsave(filename = "Stacked_barplot_tumor_descriptor_Deceased.pdf", path = plots_dir,, width=10 ,height =6 ,device = "pdf")

# Remove PT_3CHB9PK5 due to high TMB to make plot look nicer
tmb_merge_rm_PT_3CHB9PK5 <- tmb_merge %>% 
  filter(!(Kids_First_Participant_ID == "PT_3CHB9PK5"))   

# set the order for the x axis
tmb_merge_rm_PT_3CHB9PK5$tumor_descriptor <- factor(x = tmb_merge_rm_PT_3CHB9PK5$tumor_descriptor, levels = c("Diagnosis","Progressive", "Recurrence", "Deceased"))

print(ggplot(tmb_merge_rm_PT_3CHB9PK5, aes(x = factor(Kids_First_Participant_ID), y = mutation_count, fill = tumor_descriptor))+  
  geom_col(position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = c("gray", "pink", "firebrick3", "dodgerblue3")) +
  geom_bar( stat="identity", width = 0.9) + 
  theme_Publication() + 
  theme(axis.text.x = element_text(angle = 85, hjust = 1, vjust =1)) + 
  scale_y_continuous(limits=c(0,500)) +
  labs(title = "", x = "Kids_First_Participant_ID", y = "Total Mutations"))

ggsave(filename = "Stacked_barplot_tumor_descriptor_Deceased_rm_PT_3CHB9PK5.pdf", path = plots_dir, width=10 ,height =6 ,device = "pdf")
```

Attention, the sample PT_6N825561 has no Kids_First_Biospecimen_ID in the TMB file, hence no Stacked_barplot is shown.
The following are missing: Kids_First_Biospecimen_ID = BS_0YWFFRYV; BS_3V7ZNBTG; BS_E0S2Y0TS; BS_YTCYVVY9

## Corplots across all Matched Samples

We will look into cases with two time points.

### Let's look into the Diagnosis - Deceased matched samples.

```{r Dx_Dec plot}
# Make list
file_list <-  list.files (path = scratch_Dx_Dec, # Identify all CSV files
                       pattern = "*-vaf.csv", full.names = TRUE)

# File path to results directory
plots_dir_Dx_Dec <-
  file.path(plots_dir, "Dx_Dec")
if (!dir.exists(plots_dir_Dx_Dec)) {
  dir.create(plots_dir_Dx_Dec)
}

# Run loop for all files in the dir
for (x in file_list){
  
  # Let's get patient ID
  base = strsplit(x,"-vaf.csv")[[1]]  # splitting to save Dx and relapse sample names
  words = strsplit(base,"/")[[1]] 
  p = words[12]
  print(p)
  
  # Reading files iteratively and storing in a dataframe
  dat1 <- read.csv(file_list[1])
  dat1[is.na(dat1)] = 0  # Replace all na's with 0

  # Let's create columns for each time point examined
  dat1$Diagnosis <- ifelse(dat1$tumor_descriptor == dat1$tumor_descriptor, dat1$VAF, NA)
  dat1$Deceased <- ifelse(dat1$tumor_descriptor == dat1$tumor_descriptor, dat1$VAF, NA)
  
  # Label plot based on patient ID
  plot_title = p
  
  # set the order for the time points
  dat1$tumor_descriptor <- factor(x = dat1$tumor_descriptor, levels = c("Diagnosis", "Deceased"))
  
  # Plotting scatterplot for all Dx-Deceased Samples 
  print(ggplot(dat1, aes(Diagnosis, Deceased, color = tumor_descriptor)) + 
    geom_point( size = 10,fill = 4, alpha = 1/6) + 
    scale_colour_manual(values = c("dodgerblue3", "gray34")) + 
    labs(title = paste(plot_title)) + 
    geom_vline(xintercept = 0.1, linetype = "dashed") + # Adding vertical intercept
    geom_hline(yintercept = 0.1, linetype = "dashed") + # Adding horizontal intercept line
    geom_text_repel(aes(label =paste("",Hugo_Symbol,"")),size = 3.5,hjust = 0,vjust = 0, nudge_x = 0.005,point.padding = NA,segment.color = NA, show.legend = FALSE, xlim = c(0.02,NA),ylim = c(0.025,0.96)) + # if label column is one, then label the point with gene name and it's corresponding Protein Change value
    theme_Publication() + 
    #theme(plot.title = element_text(hjust = 0.5))+ 
    xlim(0,1) + 
    ylim(0,1))
  
  # Save as PDF
  ggsave(filename = paste0("Corplot_", p, ".pdf"), path = plots_dir_Dx_Dec, width=8 ,height =6 ,device = "pdf", useDingbats=F)
  
}
```

### Let's look into the Progression - Deceased matched samples.

```{r Pro_Dec plot}
# Make list
file_list <-  list.files (path = scratch_Pro_Dec, # Identify all CSV files
                       pattern = "*-vaf.csv", full.names = TRUE)

# File path to results directory
plots_dir_Pro_Dec <-
  file.path(plots_dir, "Pro_Dec")
if (!dir.exists(plots_dir_Pro_Dec)) {
  dir.create(plots_dir_Pro_Dec)
}

# Run loop for all files in the dir
for (x in file_list){
  
  # Let's get patient ID
  base = strsplit(x,"-vaf.csv")[[1]]  # splitting to save Dx and relapse sample names
  words = strsplit(base,"/")[[1]] 
  p = words[12]
  print(p)
  
  # Reading files iteratively and storing in a dataframe
  dat1 <- read.csv(file_list[1])
  dat1[is.na(dat1)] = 0  # Replace all na's with 0

  # Let's create columns for each time point examined
  dat1$Progressive <- ifelse(dat1$tumor_descriptor == dat1$tumor_descriptor, dat1$VAF, NA)
  dat1$Deceased <- ifelse(dat1$tumor_descriptor == dat1$tumor_descriptor, dat1$VAF, NA)
  
  # Label plot based on patient ID
  plot_title = p
  
  # set the order for the time points
  dat1$tumor_descriptor <- factor(x = dat1$tumor_descriptor, levels = c("Progressive", "Deceased"))
  
  # Plotting scatterplot for all Pro_Dec Samples 
  print(ggplot(dat1, aes(Progressive, Deceased, color = tumor_descriptor)) + 
    geom_point( size = 10,fill = 4, alpha = 1/6) + 
    scale_colour_manual(values = c("firebrick3", "gray34")) + 
    labs(title = paste(plot_title)) + 
    geom_vline(xintercept = 0.1, linetype = "dashed") + # Adding vertical intercept
    geom_hline(yintercept = 0.1, linetype = "dashed") + # Adding horizontal intercept line
    geom_text_repel(aes(label =paste("",Hugo_Symbol,"")),size = 3.5,hjust = 0,vjust = 0, nudge_x = 0.005,point.padding = NA,segment.color = NA, show.legend = FALSE, xlim = c(0.02,NA),ylim = c(0.025,0.96)) + # if label column is one, then label the point with gene name and it's corresponding Protein Change value
    theme_Publication() + 
    #theme(plot.title = element_text(hjust = 0.5))+ 
    xlim(0,1) + 
    ylim(0,1))
  
  # Save as PDF
  ggsave(filename = paste0("Corplot_", p, ".pdf"), path = plots_dir_Pro_Dec, width=8 ,height =6 ,device = "pdf", useDingbats=F)
  
}
```

### Let's look into the Recurrence - Deceased matched samples.

```{r Rec_Dec plot}
# Make list
file_list <-  list.files (path = scratch_Rec_Dec, # Identify all CSV files
                       pattern = "*-vaf.csv", full.names = TRUE)

# File path to results directory
plots_dir_Rec_Dec <-
  file.path(plots_dir, "Rec_Dec")
if (!dir.exists(plots_dir_Rec_Dec)) {
  dir.create(plots_dir_Rec_Dec)
}

# Run loop for all files in the dir
for (x in file_list){
  
  # Let's get patient ID
  base = strsplit(x,"-vaf.csv")[[1]]  # splitting to save Dx and relapse sample names
  words = strsplit(base,"/")[[1]] 
  p = words[12]
  print(p)
  
  # Reading files iteratively and storing in a dataframe
  dat1 <- read.csv(file_list[1])
  dat1[is.na(dat1)] = 0  # Replace all na's with 0

  # Let's create columns for each time point examined
  dat1$Recurrence <- ifelse(dat1$tumor_descriptor == dat1$tumor_descriptor, dat1$VAF, NA)
  dat1$Deceased <- ifelse(dat1$tumor_descriptor == dat1$tumor_descriptor, dat1$VAF, NA)
  
  # Label plot based on patient ID
  plot_title = p
  
  # set the order for the time points
  dat1$tumor_descriptor <- factor(x = dat1$tumor_descriptor, levels = c("Recurrence", "Deceased"))
  
  # Plotting scatterplot for all Rec_Dec Samples 
  print(ggplot(dat1, aes(Recurrence, Deceased, color = tumor_descriptor)) + 
    geom_point( size = 10,fill = 4, alpha = 1/6) + 
    scale_colour_manual(values = c("pink", "gray34")) + 
    labs(title = paste(plot_title)) + 
    geom_vline(xintercept = 0.1, linetype = "dashed") + # Adding vertical intercept
    geom_hline(yintercept = 0.1, linetype = "dashed") + # Adding horizontal intercept line
    geom_text_repel(aes(label =paste("",Hugo_Symbol,"")),size = 3.5,hjust = 0,vjust = 0, nudge_x = 0.005,point.padding = NA,segment.color = NA, show.legend = FALSE, xlim = c(0.02,NA),ylim = c(0.025,0.96)) + # if label column is one, then label the point with gene name and it's corresponding Protein Change value
    theme_Publication() + 
    #theme(plot.title = element_text(hjust = 0.5))+ 
    xlim(0,1) + 
    ylim(0,1))
  
  # Save as PDF
  ggsave(filename = paste0("Corplot_", p, ".pdf"), path = plots_dir_Rec_Dec, width=8 ,height =6 ,device = "pdf", useDingbats=F)
  
}
```

I did not look into cases with three time points, even though I split and save those data above in the scratch dir.
What should I do with those?

```{r echo=TRUE}
# Delete Dir scratch because it contains large files not needed anymore
unlink(scratch,recursive = TRUE)
```


```{r echo=TRUE}
sessionInfo()
```
  