---
title: "Longitudinal analysis for the pbta-tumor-evolution project and PBTA Cohort"
author: 'Antonia Chroni <chronia@chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---


#### Tumor evolution project 

# Background
This is an exploratory analysis of longitudinal data in the PBTA cohort.
We are investigating the number of histologies (cancer_type column): (1) per assay (experimental_strategy) and (2) pairs of genomic and transcriptomic assays.
We are also looking at the number of patient samples available per histology with paired genomic and transcriptomic assays for (1) each time point (disease_stage) and (2) when Diagnosis is present and paired with another point in time.

### Data used 
In this notebook, we are looking into the v12 histologies file (last updates on May 1st, 2023) and subset to the PBTA cohort.
We use tumor_descriptor and focus on disease_stages: Diagnosis, Progressive, Recurrence, and Deceased.

### Note
The colors in the plots do not match the ones of the PBTA cohort yet. ---TO DO---


### Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/01_sample_distribution_analysis/01-tumor-descriptor-and-assay-count-plot.Rmd', clean = TRUE)"
```

# Set up

```{r load-library}

suppressPackageStartupMessages({
  library(tidyverse)
  library(ggplot2)
  library(flextable)
  library(RColorBrewer) # to discard later
})


```

## Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no
# matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
setwd(root_dir)


# File path to input directory
input_dir <-
  file.path(root_dir, "input")


# Inputs
input_metadata <- file.path(input_dir, "histologies.tsv")
input_cols <- file.path(input_dir, "PBTA-germline-histology-groups-plot-mapping.tsv")


# File path to results directory
results_dir <-
  file.path(root_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}


# File path to plots directory
# plots_dir <-
#   file.path(root_dir, "plots")
# if (!dir.exists(plots_dir)) {
#   dir.create(plots_dir)
# }


# Outputs
output_PBTA_filter <- file.path(results_dir, "data_by_cohort_PBTA_filter.tsv")
output_PBTA_filter_count <- file.path(results_dir, "data_by_cohort_PBTA_filter_count.tsv")
output_Diagnosis_paired <- file.path(results_dir, "Diagnosis_paired.tsv")
output_Diagnosis_paired_dup <- file.path(results_dir, "Diagnosis_paired_dup.tsv")

output_Diagnosis_paired_cancer_group <- file.path(results_dir, "Diagnosis_paired_cancer_group.tsv")
output_Diagnosis_paired_cancer_group_duplicates <- file.path(results_dir, "Diagnosis_paired_cancer_group-duplicates.tsv")


# to add this later
# source publication theme
# source(file.path(root_dir, "figures", "theme.R"))
```


## Load data
```{r load-inputs-please-wait}
# Inputs
## Read in metadata
histologies <- readr::read_tsv(input_metadata, guess_max = 100000, show_col_types = FALSE)
# dim(data)

# Read table with colors per histology
cols <- readr::read_tsv(input_cols, guess_max = 100000, show_col_types = FALSE) # Load plotting colors

# Add colors palette to the histologies file
hist_cols <- histologies %>% left_join(cols, by=c('cancer_group','broad_histology'))  #to merge tables
# This is the tidyverse method. Try not to name variables the same as a function's name - that may cause coding errors downstream.


# to discard when add the col palette for PBTA cohort
mypalette <- brewer.pal(12,"Paired") # colorblindFriendly = TRUE
col5 <- colorRampPalette(mypalette)(5)
col9 <- colorRampPalette(mypalette)(9)

```

# Process Data

All tumor samples have matched normal samples. 
We need to filter normal samples out for further analysis. 
We are using "pathology_diagnosis" instead of "sample_type". This way we discard normal samples (as shown in "sample_type") in addition to undiagnosed tumor samples.
Also, we will discard Derived Cell Line, PDX from "composition".


```{r cohort PBTA, echo=TRUE}
tumor_df <- hist_cols  %>%
   filter(cohort == "PBTA",
          !is.na(pathology_diagnosis),
          !composition %in% c("Derived Cell Line", "PDX")) %>% 
         # !is.na(cancer_group) # Discard NAs in cancer type
   mutate(tumor_descriptor = ifelse(sample_id == "C4106355-Tumor", "Recurrence", # to be updated in v13 as "Recurrence"
                            ifelse(sample_id == "C4113612-Tumor", "Diagnosis", # to be updated in v13 as "Initial CNS Tumor"
                            ifelse(tumor_descriptor == "Primary Tumor", "Diagnosis",
                            ifelse(tumor_descriptor == "Initial CNS Tumor", "Diagnosis",
                            ifelse(tumor_descriptor == 'Progressive Disease Post-Mortem', 'Deceased',      
                                          tumor_descriptor))))))
                 
print (tumor_df %>% count(tumor_descriptor))

# Note
# The following samples (sample_id) are marked as Not Reported tumor_descriptor: 7316UP-3705, C4106355-Tumor, and C4113612-Tumor
# Jo Lynne found the following status for 2/3 in the data warehouse:
# C4106355-Tumor = Recurrence
# C4113612-Tumor = Initial CNS Tumor
# 7316UP-3705 = Not Reported
# manually annotate them here and these will be updated in the v13 - then discard that part of the code
# For now, you can leave this - we can fix the plot colors later - it might be that we have a longitudinal sample which comes from a benign tumor - we can possibly remove these later, but it would be good to get a source of truth for what our entire cohort looks like first before removing data,

# Save tumor_df after filter 
write_tsv(tumor_df, path = output_PBTA_filter)

count1 <- tumor_df %>% 
          group_by(cohort_participant_id, Kids_First_Participant_ID,
                 experimental_strategy, sample_type, tumor_descriptor, primary_site,
                 age_at_diagnosis_days, age_last_update_days, age_at_chemo_start, age_at_radiation_start,
                 pathology_diagnosis, cancer_group, broad_histology) %>% 
          dplyr::count(cancer_group) 

write_tsv(count1, output_PBTA_filter_count)

```

# Plots
## All assays

```{r plot1}
count1$experimental_strategy <- factor(x = count1$experimental_strategy, levels = c("WGS", "WXS", "RNA-Seq", "Targeted Sequencing", "Methylation"))

p <- ggplot(count1, aes(x=cancer_group, fill=experimental_strategy))+
        geom_bar()+
        scale_fill_manual(values=col5) +
        theme_classic()+
        ylab("Count cancer_group") +
        ggtitle("cancer_group vs experimental_strategy") +
        guides(color=guide_legend(override.aes=list(size=3))) +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
        print(p)
#ggsave("Barplot_cancer_group_experimental_strategy_PBTA.pdf", plot=p, path=plots_dir, device="pdf", height=7, width=10)

```

## Paired genomic, transcriptomic assays

```{r plot2}
count <- count1 %>% filter(experimental_strategy %in% c("WGS", "WXS", "RNA-Seq")) # Excluded: "Targeted Sequencing", Methylation
count$experimental_strategy <- factor(x = count$experimental_strategy, levels = c("WGS", "WXS", "RNA-Seq"))

p <- ggplot(count, aes(x=cancer_group, fill=experimental_strategy))+
        geom_bar()+
        scale_fill_manual(values=col5) + #scale_fill_manual(values=c(tumor_df$plot_group_hex)) 
        theme_classic()+
        ylab("Count cancer_group") +
        ggtitle("cancer_group vs experimental_strategy") +
        guides(color=guide_legend(override.aes=list(size=3))) +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
        print(p)
#ggsave("Barplot_cancer_group_experimental_strategy_PBTA_matched_DNA-RNA.pdf", plot=p, path=plots_dir, device="pdf", height=7, width=10)

```

## By cancer_group - all disease stages


```{r plot3}
count$tumor_descriptor <- factor(x = count$tumor_descriptor, levels = c("Diagnosis", "Progressive", "Recurrence", "Deceased", "Second Malignancy", "Metastatic", "Not Reported", "Residual", "Unavailable"))


p <- ggplot(count, aes(x=cancer_group, fill=tumor_descriptor))+
        geom_bar()+
        scale_fill_manual(values=col9) +
        theme_classic()+
        ylab("Count cancer_group") +
        ggtitle("cancer_group vs tumor_descriptor") +
        guides(color=guide_legend(override.aes=list(size=3))) +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
        print(p)
#ggsave("Barplot_cancer_group_tumor_descriptor_PBTA_matched_DNA-RNA.pdf", plot=p, path=plots_dir, device="pdf", height=7, width=10)

```

# Summary

In the following section, we will provide summary table across different disease stages, histologies, and patients samples.

## Number of each assay

First, we’ll examine how many of each type of assay we have (tumors only). This information is stored in the experimental_strategy column.

```{r}
tumor_df %>%
  group_by(experimental_strategy) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```



# Diagnosis vs. any other point in time (recurrence, progressive etc)

We’ll look at the breakdown of the tumor_descriptor column, separating the genomic assays from transcriptomic assays.

```{r}
tumor_descriptor_df <- tumor_df %>%
  select(Kids_First_Participant_ID,
         experimental_strategy,
         tumor_descriptor,
         cancer_group) %>%
  arrange(Kids_First_Participant_ID)
```


## Genomic assays
Setting aside the Panel sample for the moment and only looking at WXS and WGS assays. We’re collapsing the different values in tumor_descriptor to form a single descriptor when there are multiple types of tumors from the same individual.


```{r}
genomic_df <- tumor_descriptor_df %>%
  filter(experimental_strategy %in% c("WGS", "WXS")) %>%
  group_by(Kids_First_Participant_ID) %>%
  summarize(descriptors = paste(sort(unique(tumor_descriptor)),  
                                collapse = ", "),
            experimental_strategy = paste(sort(unique(experimental_strategy)),
                                          collapse = ", "),
            cancer_group = paste(sort(unique(cancer_group)),
                                          collapse = ", "))

genomic_df %>%
  group_by(descriptors) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```


## Transcriptomic assays
Looking only at RNA-seq samples and performing the same collapsing of the tumor_descriptor column.

```{r}
transcriptomic_df <- tumor_descriptor_df %>%
  filter(experimental_strategy == "RNA-Seq") %>%
  group_by(Kids_First_Participant_ID) %>%
  summarize(descriptors = paste(sort(unique(tumor_descriptor)),  
                                collapse = ", "),
            experimental_strategy = unique(experimental_strategy),
            cancer_group = paste(sort(unique(cancer_group)),
                                          collapse = ", "))

transcriptomic_df %>%
  group_by(descriptors) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")
```


## Paired genomic, transcriptomic assays
How many participants have paired genomic and transcriptomic samples for the same tumor_descriptor values?

```{r}
# if we perform a full join here using the Kids_First_Participant_ID and descriptors, we 
# will get NAs in columns where pairs don't exist and can use this information 
# to count
paired_df <-
  full_join(genomic_df, transcriptomic_df,
                   by = c("cancer_group", "Kids_First_Participant_ID", "descriptors"))

```

### All time points
Count the examples where all time points have both kinds of assays.

```{r}
# no NAs = both kinds of assays are present
paired_df %>%
  filter(complete.cases(.)) %>%
  group_by(descriptors) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part="all")
```


### Diagnosis paired with another point in time

```{r}
Diagnosis_paired <- paired_df  %>%
  filter(descriptors != "Diagnosis")  %>% 
  filter(grepl("Diagnosis", descriptors))
  
# Save Diagnosis_paired 
write_tsv(Diagnosis_paired, path = output_Diagnosis_paired)

Diagnosis_paired <- Diagnosis_paired[with(Diagnosis_paired, order(descriptors, cancer_group)), ]


Diagnosis_paired %>%
  regulartable() %>%
  fontsize(size = 12, part = "all")
```


### No RNA-seq
If the experimental_strategy.y column has an NA, RNA-seq is missing for that participant-descriptors pair.

```{r}
paired_df %>%
  filter(is.na(experimental_strategy.y)) %>%
  group_by(descriptors) %>%
  tally() %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")
```


### No WXS or WGS
If the experimental_strategy.x column has an NA, there is no WGS or WXS for that participant-descriptors pair.

```{r}
paired_df %>%
  filter(is.na(experimental_strategy.x)) %>%
  group_by(descriptors) %>%
  tally() %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")
```

### Duplicate participant IDs
There are cases where one timepoint (e.g., tumor_descriptor value) is missing from one kind of assay but not the other. This will show up as duplicated values in the Kids_First_Participant_ID column.

```{r}
ids_with_dups <- paired_df %>%
  filter(duplicated(Kids_First_Participant_ID)) %>%
  pull(Kids_First_Participant_ID)

Diagnosis_paired_dup <- paired_df %>%
  filter(Kids_First_Participant_ID %in% ids_with_dups) %>%
  arrange(Kids_First_Participant_ID)  

# Save Diagnosis_paired 
write_tsv(Diagnosis_paired_dup, path = output_Diagnosis_paired_dup)

Diagnosis_paired_dup %>%
  regulartable() %>%
  fontsize(size = 12, part = "all")
```



### Cohorts with low survival 
Let's estimate how many patient samples we have with Diagnosis paired with another point in time.
We will do this for cohorts with low survival first: Atypical Teratoid Rhabdoid Tumor (ATRT) and High-grade glioma (HGG).

We will first print list with cancer_groups to confirm the name of cancer groups we are interested as potential cohorts
and if there are any combinations of cancer_groups.

```{r}
list <-Diagnosis_paired %>% 
      count(cancer_group)
print (list, n)

list <-Diagnosis_paired_dup %>% 
      count(cancer_group)
print (list, n)

```

We notice for the High-grade glioma cases there are multiple diagnoses.
We will modify the cancer_group column to capture all possible combinations.

```{r}
Diagnosis_paired_modified <- Diagnosis_paired %>% 
  mutate(cancer_group_modified = ifelse(cancer_group == "High-grade glioma, Low-grade glioma", "High-grade glioma",
                                            cancer_group))
                              
                          
Diagnosis_paired_dup_modified <- Diagnosis_paired_dup %>% 
  mutate(cancer_group_modified = ifelse(cancer_group == "High-grade glioma, Infant-type hemispheric glioma", "High-grade glioma",
                        ifelse(cancer_group == "High-grade glioma, Low-grade glioma", "High-grade glioma",
                        ifelse(cancer_group == "High-grade glioma, Pilocytic astrocytoma", "High-grade glioma",
                                            cancer_group))))
  
```  

#### Atypical Teratoid Rhabdoid Tumor (ATRT)
Per patient sample

```{r}
cohort <- Diagnosis_paired_modified %>% 
  filter(cancer_group_modified == "Atypical Teratoid Rhabdoid Tumor")

cohort %>% 
  count(cancer_group_modified)


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```

Per patient sample and duplicates

```{r}

Diagnosis_paired_dup_modified %>% 
  filter(cancer_group_modified == "Atypical Teratoid Rhabdoid Tumor") %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```


#### High-grade glioma (HGG)
Per patient sample

```{r}
cohort <- Diagnosis_paired_modified %>% 
  filter(cancer_group_modified == "High-grade glioma")

cohort %>% 
  count(cancer_group_modified)


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```

Per patient sample and duplicates

```{r}

cohort_dup <-Diagnosis_paired_dup_modified %>% 
  filter(cancer_group_modified == "High-grade glioma") 

cohort_dup %>% 
  count(cancer_group_modified)

cohort_dup %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```


```{r}
sessionInfo()
```
  