---
title: "Longitudinal analysis for the PBTA Cohort (pbta-tumor-evolution project)"
author: 'Antonia Chroni <chronia@chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---


#### Tumor evolution project 

# Background
This is an exploratory analysis of longitudinal data in the PBTA cohort.
We are investigating the number of histologies (cancer_type column): (1) per assay (experimental_strategy) and (2) pairs of genomic and transcriptomic assays.
We are also looking at the number of patient samples available per histology with paired genomic and transcriptomic assays for (1) each time point (disease_stage) and (2) when Diagnosis is present and paired with another point in time.

### Data used 
In this notebook, we are looking into the v12 histologies file (last updates on May 1st, 2023) and subset to the PBTA cohort.
We use tumor_descriptor and focus on time points recorded.

### Note
The colors in the plots do not match the ones of the PBTA cohort yet. ---TO DO---


### Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/01_sample_distribution_analysis/01-tumor-descriptor-and-assay-count-plot.Rmd', clean = TRUE)"
```

# Set up

```{r load-library}

suppressPackageStartupMessages({
  library(tidyverse)
  library(ggplot2)
  library(flextable)
  library(RColorBrewer) # to discard later
})


```

## Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no
# matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
setwd(root_dir)


# File path to input directory
input_dir <-
  file.path(root_dir, "input")


# Inputs
input_metadata <- file.path(input_dir, "histologies.tsv")
input_cols <- file.path(input_dir, "PBTA-germline-histology-groups-plot-mapping.tsv")


# File path to results directory
results_dir <-
  file.path(root_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}


# File path to plots directory
# plots_dir <-
#   file.path(root_dir, "plots")
# if (!dir.exists(plots_dir)) {
#   dir.create(plots_dir)
# }


# Outputs
output_cohort_pbta <- file.path(results_dir, "cohort_pbta.tsv")

output_paired_assays <- file.path(results_dir, "paired_assays.tsv")
output_paired_assays_mult_events <- file.path(results_dir, "paired_assays_mult_events.tsv")

output_paired_assays_mult_events_filter <- file.path(results_dir, "paired_assays_mult_events_filter.tsv")

output_paired_Diagnosis <- file.path(results_dir, "paired_Diagnosis.tsv")


# to add this later
# source publication theme
# source(file.path(root_dir, "figures", "theme.R"))
```


## Load data
```{r load-inputs-please-wait}
# Inputs
## Read in metadata
v12 <- readr::read_tsv(input_metadata, guess_max = 100000, show_col_types = FALSE)
# dim(data)

# Read table with colors per histology
# cols <- readr::read_tsv(input_cols, guess_max = 100000, show_col_types = FALSE) # Load plotting colors

# Add colors palette to the histologies file
# hist_cols <- histologies %>% left_join(cols, by=c('cancer_group','broad_histology'))  #to merge tables
# This is the tidyverse method. Try not to name variables the same as a function's name - that may cause coding errors downstream.


# to discard when add the col palette for PBTA cohort
# mypalette <- brewer.pal(12,"Paired") # colorblindFriendly = TRUE
# col5 <- colorRampPalette(mypalette)(5)
# col9 <- colorRampPalette(mypalette)(9)

```

# Process Data

All tumor samples have matched normal samples. 
We need to filter normal samples out for further analysis. 
We are using "pathology_diagnosis" instead of "sample_type". This way we discard normal samples (as shown in "sample_type") in addition to undiagnosed tumor samples.
Also, we will discard Derived Cell Line, PDX from "composition".


```{r cohort PBTA, echo=TRUE}
# Note
# The following samples (sample_id) are marked as Not Reported tumor_descriptor: 7316UP-3705, C4106355-Tumor, and C4113612-Tumor
# Jo Lynne found the following status for 2/3 in the data warehouse:
# C4106355-Tumor = Recurrence
# C4113612-Tumor = Initial CNS Tumor
# 7316UP-3705 = Not Reported
# manually annotate them here and these will be updated in the v13 - then discard that part of the code
# For now, you can leave this - we can fix the plot colors later - it might be that we have a longitudinal sample which comes from a benign tumor - we can possibly remove these later, but it would be good to get a source of truth for what our entire cohort looks like first before removing data,

# I also added the cancer_group into the match_id to use this later in the script
pbta <- v12 %>%
  filter(cohort == "PBTA",
         !is.na(pathology_diagnosis),
         !composition %in% c("Derived Cell Line", "PDX"),
         !experimental_strategy %in% c("Methylation", "Targeted Sequencing")) %>%
           # !is.na(cancer_group) # Discard NAs in cancer type
  mutate(match_id = paste(cancer_group, sample_id, composition, tumor_descriptor, sep = "_")) %>% 
  mutate(tumor_descriptor_final = case_when(grepl("C4106355-Tumor", sample_id) ~ "Recurrence", # to be updated in v13 as "Recurrence"
                                            grepl("C4113612-Tumor", sample_id) ~ "Diagnosis", # to be updated in v13 as "Initial CNS Tumor"
                                            grepl("Primary Tumor", tumor_descriptor) ~ "Diagnosis",
                                            grepl("Initial CNS Tumor", tumor_descriptor) ~ "Diagnosis",
                                            grepl('Progressive Disease Post-Mortem', tumor_descriptor) ~ 'Deceased', 
                                            TRUE ~ tumor_descriptor))

print(pbta %>% count(tumor_descriptor_final))

# Save pbta after filter 
write_tsv(pbta, path = output_cohort_pbta)
```


## Filter based on both genomic and transcriptomic assays
Let's select all patient samples containing both genomic and transcriptomic assays.

```{r assays, echo=TRUE}

print(length(unique(pbta$Kids_First_Participant_ID)))
print(length(unique(pbta$match_id)))

assays <- pbta %>%
  select(Kids_First_Participant_ID, tumor_descriptor_final, match_id, experimental_strategy, cancer_group) %>%
  unique() %>%
  arrange(Kids_First_Participant_ID)


# we perform join here using the Kids_First_Participant_ID and match_id
paired_assays <- assays %>%
  group_by(Kids_First_Participant_ID, match_id) %>%
  summarise(experimental_strategy = str_c(experimental_strategy, collapse = ";")) %>%
  mutate(experimental_strategy = case_when(grepl("RNA-Seq;WGS", experimental_strategy) ~ "WGS;RNA-Seq", 
                                           grepl("RNA-Seq;WXS", experimental_strategy) ~ "WXS;RNA-Seq",
                                           grepl("WGS;RNA-Seq;WXS", experimental_strategy) ~ "WGS;WXS;RNA-Seq",
                                           grepl("WXS;WGS;RNA-Seq", experimental_strategy) ~ "WGS;WXS;RNA-Seq",
                                           grepl("RNA-Seq;WGS;WXS", experimental_strategy) ~ "WGS;WXS;RNA-Seq",
                                           grepl("RNA-Seq;WXS;WGS", experimental_strategy) ~ "WGS;WXS;RNA-Seq",
                                           TRUE ~ experimental_strategy))  

# let's summarize the assays
paired_assays <- paired_assays %>%
   mutate(experimental_strategy_sum = case_when(grepl("WGS;RNA-Seq", experimental_strategy) ~ "Both", 
                                                grepl("WXS;RNA-Seq", experimental_strategy) ~ "Both",
                                                grepl("WGS;WXS;RNA-Seq", experimental_strategy) ~ "Both",
                                                grepl("WGS", experimental_strategy) ~ "DNA",
                                                grepl("WXS", experimental_strategy) ~ "DNA", 
                                                grepl("RNA-Seq", experimental_strategy) ~ "RNA",
                                                TRUE ~ experimental_strategy)) 
  
# Save paired_assays 
write_tsv(paired_assays, path = output_paired_assays)
```

## Filter out duplicate events
We will identify duplicate events and discard them from further analysis.

```{r paired_assays_mult_events, echo=TRUE}

# identify multiple events per patient sample
paired_assays_mult_events <- paired_assays %>%
  group_by(Kids_First_Participant_ID) %>%
  summarise(match_id = str_c(match_id, collapse = "~"),
            experimental_strategy = str_c(experimental_strategy, collapse = "~"),
            experimental_strategy_sum = str_c(experimental_strategy_sum, collapse = "~")) %>%
  mutate(has_mult_events = case_when(grepl("~", match_id) ~ "yes", 
                               TRUE ~ "no"))

table(paired_assays_mult_events$has_mult_events)

# Save paired_assays_mult_events
write_tsv(paired_assays_mult_events, path = output_paired_assays_mult_events)

# filter out any multiple events based on has_mult_events column
paired_assays_mult_events <- paired_assays_mult_events %>%
  filter(has_mult_events == "no")

table(paired_assays_mult_events$has_mult_events)


```

# Summary

In the following section, we will provide summary table across different disease stages, histologies, and patients samples.

## Number of each assay

First, weâ€™ll examine how many of each type of assay we have (tumors only). This information is stored in the experimental_strategy column.

```{r}
paired_assays_mult_events %>%
  group_by(experimental_strategy_sum) %>%
  tally() %>%
  arrange(desc(n)) 

paired_assays_mult_events %>%
  regulartable() %>%
  fontsize(size = 12, part = "all")
```

## All assays

### All time points
Count the examples where all time points have both kinds of assays.

```{r}
# add column with descriptors by splitting match_id.
# The following script join with pbta df generates almost double samples (2892 obs of 61 variables)

# add columns with clinical information into the dataframe
# inner join 
# test2 <-
#  full_join(pbta, paired_assays_mult_events_filter,
#                   by = c("match_id", "Kids_First_Participant_ID")) %>%
#  filter(!is.na(has_mult_events)) 

#print(paired_df %>% count(tumor_descriptor_final))


# So, instead I did the following workaround to get the cancer_group and descriptors info in the new df

# no NAs = both kinds of assays are present
paired_assays_mult_events_all <- paired_assays_mult_events %>%
    select(Kids_First_Participant_ID, match_id, experimental_strategy, experimental_strategy_sum) %>%
    mutate(cancer_group = match_id) %>%
    mutate(cancer_group =gsub("_..*", "", cancer_group)) %>% 
    mutate(descriptors = match_id) %>% 
    mutate(descriptors=gsub("..*_", "", descriptors)) %>% 
    mutate(descriptors = case_when(grepl("Primary Tumor", descriptors) ~ "Diagnosis",
                                 grepl("Initial CNS Tumor", descriptors) ~ "Diagnosis",
                                 grepl('Progressive Disease Post-Mortem', descriptors) ~ 'Deceased', 
                                 TRUE ~ descriptors))  

paired_assays_mult_events_all_table <-  paired_assays_mult_events_all %>% filter(complete.cases(.)) %>%
  group_by(descriptors) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part="all")

#paired_df['Sum']=paired_df[descriptors].sum(axis=1) 
```


### Diagnosis paired with another point in time

```{r}

paired_Diagnosis <- paired_assays_mult_events_all  %>%
  mutate(descriptors_new=case_when(
    descriptors=="Diagnosis" & descriptors=="Progressive" ~ "Diagnosis-Progressive",
    descriptors=="Diagnosis" & descriptors=="Recurrence" ~ "Diagnosis-Recurrence",
    descriptors=="Diagnosis" & descriptors=="Deceased" ~ "Diagnosis-Deceased",
    descriptors=="Diagnosis" & descriptors=="Second Malignancy" ~ "Diagnosis-Second Malignancy",
    descriptors=="Diagnosis" & descriptors=="Not Reported" ~ "Diagnosis-Not Reported",
    descriptors=="Diagnosis" & descriptors=="NA" ~ "Diagnosis-NA",
    descriptors=="Diagnosis" & descriptors=="Residual" ~ "Diagnosis-Residual",
    descriptors=="Diagnosis" & descriptors=="Metastatic" ~ "Diagnosis-Metastatic",
    descriptors=="Diagnosis" & descriptors=="Unvailable" ~ "Diagnosis-Unvailable",
    TRUE ~ "Unmatched"
  ))


# paired_Diagnosis1 <- paired_assays_mult_events  %>%
#  filter(descriptors != "Diagnosis")  %>% 
#  filter(grepl("Diagnosis", descriptors))
  
# Save paired_Diagnosis 
write_tsv(paired_Diagnosis, path = output_paired_Diagnosis)


paired_Diagnosis %>%
  regulartable() %>%
  fontsize(size = 12, part = "all")
```


## Paired Genomic and transcriptomic assays
We will look only into pairs now.
We will discard any samples with no DNA/RNA counterpart (for now - we can later ask the pathology lab to sequence these).

```{r}
# keep only paired assays
# Same workaround as before to obtain cancer_group and descriptors information
paired_assays_mult_events_filter <- paired_assays_mult_events %>%
  filter(experimental_strategy_sum == "Both") %>% 
  mutate(cancer_group = match_id) %>%
    mutate(cancer_group =gsub("_..*", "", cancer_group)) %>% 
    mutate(descriptors = match_id) %>% 
    mutate(descriptors=gsub("..*_", "", descriptors)) %>% 
    mutate(descriptors = case_when(grepl("Primary Tumor", descriptors) ~ "Diagnosis",
                                 grepl("Initial CNS Tumor", descriptors) ~ "Diagnosis",
                                 grepl('Progressive Disease Post-Mortem', descriptors) ~ 'Deceased', 
                                 TRUE ~ descriptors))  

# Save paired_df
write_tsv(paired_assays_mult_events_filter, path = output_paired_assays_mult_events_filter)


```


### All time points
Count the examples where all time points have both kinds of assays.

```{r}
# no NAs = both kinds of assays are present
paired_assays_mult_events_filter %>%
  filter(complete.cases(.)) %>%
  group_by(descriptors) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part="all")

#paired_df['Sum']=paired_df[descriptors].sum(axis=1) 
```


### Diagnosis paired with another point in time

```{r}

paired_Diagnosis <- paired_assays_mult_events_filter  %>%
  mutate(descriptors_new=case_when(
    descriptors=="Diagnosis" & descriptors=="Progressive" ~ "Diagnosis-Progressive",
    descriptors=="Diagnosis" & descriptors=="Recurrence" ~ "Diagnosis-Recurrence",
    descriptors=="Diagnosis" & descriptors=="Deceased" ~ "Diagnosis-Deceased",
    descriptors=="Diagnosis" & descriptors=="Second Malignancy" ~ "Diagnosis-Second Malignancy",
    descriptors=="Diagnosis" & descriptors=="Not Reported" ~ "Diagnosis-Not Reported",
    descriptors=="Diagnosis" & descriptors=="NA" ~ "Diagnosis-NA",
    descriptors=="Diagnosis" & descriptors=="Residual" ~ "Diagnosis-Residual",
    descriptors=="Diagnosis" & descriptors=="Metastatic" ~ "Diagnosis-Metastatic",
    descriptors=="Diagnosis" & descriptors=="Unvailable" ~ "Diagnosis-Unvailable",
    TRUE ~ "Unmatched"
  ))


#paired_Diagnosis <- paired_assays_mult_events_filter  %>%
#  filter(descriptors != "Diagnosis")  %>% 
#  filter(grepl("Diagnosis", descriptors))
  
# Save paired_Diagnosis 
write_tsv(paired_Diagnosis, path = output_paired_Diagnosis)


paired_Diagnosis %>%
  regulartable() %>%
  fontsize(size = 12, part = "all")
```



### Paired assays and cancer_group
Let's investigate the number of tumor samples with paired assays per cancer_group.

```{r}
list <- paired_Diagnosis %>% 
      count(cancer_group) %>% 
      arrange(desc(n)) 
print (list, n)


```
We see that we have > 100 patient samples for Low-grade glioma, Medulloblastoma, High-grade glioma, and Ependymoma.

#### Low-grade glioma
Per patient sample

```{r}
cohort <- paired_Diagnosis %>% 
  filter(cancer_group == "Low-grade glioma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```

#### Medulloblastoma
Per patient sample

```{r}
cohort <- paired_Diagnosis %>% 
  filter(cancer_group == "Medulloblastoma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```


#### High-grade glioma
Per patient sample

```{r}
cohort <- paired_Diagnosis %>% 
  filter(cancer_group == "High-grade glioma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```

#### Ependymoma
Per patient sample

```{r}
cohort <- paired_Diagnosis %>% 
  filter(cancer_group == "Ependymoma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```

```{r}
sessionInfo()
```
  