---
title: "Inference of subclonal architecture of tumors across multiple timepoints for the PBTA Cohort (pbta-tumor-evolution project)"
author: 'Antonia Chroni <chronia@chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---


#### Tumor evolution project 

# Background
We are investigating the subclonal architecture of tumors across multiple timepoints in the PBTA cohort.
We will infer fishplots for each patient sample with multiple time points.

### Data used 
In this notebook, we are looking into the v12 histologies file (last updates on May 1st, 2023) and subset to the PBTA cohort.
We use the output files from the "analyses/sample-distribution-analysis/results".

### Method used 
For visualizing the tumor sublcones, we will use the fishplot R package as described in here: https://github.com/chrisamiller/fishplot


### Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/fishplot/01-fishplot.Rmd', clean = TRUE)"
```

# Set up

```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(R.utils)
  library(fishplot) # https://github.com/chrisamiller/fishplot
  library(clonevol)
  library(devtools)
  library(cDriver) # https://github.com/hanasusak/cDriver/
})
```

## Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no
# matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
#root_dir <- setwd("/Users/chronia/CHOP/projects/pbta-tumor-evolution/pbta-tumor-evolution-analysis/analyses/fishplot")
setwd(root_dir)


# File path to input directory
input_dir <-
  file.path(root_dir, "input")

# Inputs
input_snv <- source(file.path(root_dir, "data", "snv-consensus-plus-hotspots.maf.tsv.gz"))

# The following files were generated from the add-sample-distribution analysis
# to source files from there - TO FIX THIS LATER
input_cohort_pbta <- file.path(input_dir, "pbta.tsv")
#input_genomic_df <- file.path(input_dir, "genomic_df.tsv")
input_LGG <- file.path(input_dir, "LGG.tsv")

# File path to results directory
results_dir <-
  file.path(root_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}


# File path to plots directory
plots_dir <-
   file.path(root_dir, "plots")
 if (!dir.exists(plots_dir)) {
   dir.create(plots_dir)
 }


# to add this later
# source publication theme
# source(file.path(root_dir, "figures", "theme.R"))
```


## Load data
```{r load-inputs-please-wait}
# Inputs
## Read MAF file
snv <- data.table::fread(input_snv, data.table = F)

## Read in input_cohort_pbta file
pbta <- readr::read_tsv(input_cohort_pbta, guess_max = 100000, show_col_types = FALSE)

## Read in genomic_df file
# genomic_df <- readr::read_tsv(input_genomic_df, guess_max = 100000, show_col_types = FALSE)

## Read in LGG file
LGG <- readr::read_tsv(input_LGG, guess_max = 100000, show_col_types = FALSE)
```



# Process Data: snv

We need to process each input file and generate input files for the fishplot analysis.
We will first change the column name of the biospecimen column to match the one from the histolgies files.
Then, we will calculate VAFs and CCFs per tumor sample.

```{r process snv, echo=TRUE}
colnames(snv)[which(names(snv) == "Tumor_Sample_Barcode")] <- "Kids_First_Biospecimen_ID"
```  


```{r calculate VAF, echo=TRUE}
# using formula for VAF calculation
snv_vaf <- snv %>% mutate(VAF = t_alt_count / (t_ref_count + t_alt_count))
```

```{r calculate CCF, echo=TRUE}
# computational way by using cDriver R package
# if you encounter the following "Error: vector memory exhausted (limit reached?)"
# you may need to increase the memory in Renviron for this step
# https://stackoverflow.com/questions/51295402/r-on-macos-error-vector-memory-exhausted-limit-reached
# gc() #to check the memory used and/or available
snv_vaf_ccf <- CCF(snv_vaf)

# using formula for CCF calculation
snv_vaf_ccf_formula <- snv_vaf_ccf %>% mutate(CCF_formula = (2*t_alt_count)/ (t_ref_count + t_alt_count))
```

Both methods for CCF calculation seem to return the same values. The one using the formula is faster and requires less memory.
For now, we will proceed with the results from cDriver R package.

```{r echo=TRUE}
# we will subset to variables we will interested to focus on for downstream analysis
snv_vaf_ccf_subset <- subset(snv_vaf_ccf, 
                             select = c(Hugo_Symbol, Kids_First_Biospecimen_ID, Chromosome, Start_Position, End_Position, Variant_Type, Gene, VAF, CCF)) 
```


# Process Data: LGG

We will filter out from the LGG input file any not matched time point.
We will also remove any matches of time points with itself, e.g., Diagnosis-Diagnosis.

```{r process LGG, echo=TRUE}
LGG_filter <- LGG %>%
  mutate(descriptors_keep = case_when(grepl("Deceased, Progressive, Progressive", descriptors) ~ "keep",
                                      grepl("Diagnosis, Diagnosis, Diagnosis, Progressive, Progressive", descriptors) ~ "keep",
                                      grepl("Diagnosis, Progressive", descriptors) ~ "keep", 
                                      grepl("Diagnosis, Progressive, Recurrence", descriptors) ~ "keep",
                                      grepl("Diagnosis, Recurrence", descriptors) ~ "keep",
                                      grepl("Diagnosis, Recurrence, Recurrence", descriptors) ~ "keep",
                                      grepl("NA, Progressive", descriptors) ~ "keep",
                                      grepl("Progressive, Recurrence", descriptors) ~ "keep",
                                      grepl("Recurrence, Second Malignancy", descriptors) ~ "keep"))  %>%
  filter(descriptors_keep == "keep")  %>%
  write_tsv(file.path(results_dir, "genomic_df_filter.tsv"))
  
```

We have 27 patient samples with matched time points diagnosed with Low grade glioma (with both genomic and transcriptomic assays).


# Merge Data

We will merge all dataframes together.
We will first merge LGG_filter with pbta, and then, with snv_vaf_subset.

```{r merge df, echo=TRUE}
# we first need the age_at_event_days column to be added into the LGG_filter df
pbta_subset <- pbta %>%
  filter(cancer_group == "Low-grade glioma") %>%
  subset(select = c(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, 
                          tumor_descriptor, age_at_event_days, experimental_strategy,
                          broad_histology, molecular_subtype))

LGG_filter_pbta <- LGG_filter %>% 
  left_join(pbta_subset, by=c('Kids_First_Participant_ID')) 
  #distinct(Kids_First_Participant_ID, .keep_all = TRUE) %>%
  
LGG_filter_pbta <- LGG_filter_pbta[!duplicated(LGG_filter_pbta[c("Kids_First_Participant_ID", "age_at_event_days")]),] %>%
  write_tsv(file.path(results_dir, "LGG_filter_pbta.tsv"))

merge_df_all <- LGG_filter_pbta %>% 
  left_join(snv_vaf_ccf_subset, by=c('Kids_First_Biospecimen_ID')) %>%
  write_tsv(file.path(results_dir, "merge_df_all.tsv"))
```


# Run subclonal analysis to produce fishplot

We will use the fishplot R package to infer and visualize subclones across multiple time points.

The following is based on patient PT_962TCBVR with three time points present: Diagnosis, Progressive, Recurrence.
Also, for the "frac.table" and "parents" matrices, I am using a random example.
My rational was to make the function work until I figure out how to properly generate the matrices.
And it did :)

```{r subclonal analysis, echo=TRUE}
#provide a list of timepoints to plot
#You may need to add interpolated points to end up with the desired visualization. 

# The first timepoint needs to be set up at 0 in order to run fishplot function
# We will create age_at_event_days_mut column where we will replace with zeros the age for the first time point event.

merge_df_all_mut <- merge_df_all %>% 
  filter(!experimental_strategy.y == "RNA-Seq") %>%
  mutate(age_at_event_days_mut = age_at_event_days) %>%
  mutate(age_at_event_days_mut = case_when(descriptors == "Deceased, Progressive, Progressive" & tumor_descriptor == "Deceased" ~ age_at_event_days_mut == "0",
                                           descriptors == "Diagnosis, Diagnosis, Diagnosis, Progressive, Progressive" & tumor_descriptor == "Initial CNS Tumor" ~ age_at_event_days_mut == "0",
                                           descriptors == "Diagnosis, Progressive" & tumor_descriptor == "Initial CNS Tumor" ~ age_at_event_days_mut == "0",
                                           descriptors == "Diagnosis, Progressive, Recurrence" & tumor_descriptor == "Initial CNS Tumor" ~ age_at_event_days_mut == "0",
                                           descriptors == "Diagnosis, Recurrence" & tumor_descriptor == "Initial CNS Tumor" ~ age_at_event_days_mut == "0",
                                           descriptors == "Diagnosis, Recurrence, Recurrence" & tumor_descriptor == "Initial CNS Tumor" ~ age_at_event_days_mut == "0",
                                           descriptors == "NA, Progressive" & tumor_descriptor == "NA" ~ age_at_event_days_mut == "0",
                                           descriptors == "Progressive, Recurrence" & tumor_descriptor == "Progressive" ~ age_at_event_days_mut == "0",
                                           descriptors == "Recurrence, Second Malignancy" & tumor_descriptor == "Recurrence" ~ age_at_event_days_mut == "0", 
                                           TRUE ~ age_at_event_days_mut)) %>%
  write_tsv(file.path(results_dir, "merge_df_all_mut.tsv"))


# there are issues with patient samples not having all tumor_descriptor as they should
# that results into single events for age_at_event_days
# and fishplot fails
# I need to work on this more.


# Anyway, I focus for now on one patient (PT_962TCBVR) for which three time points were present

# BUT I need to figure out how to create frac.table from CCF per position - the one here is an example
# file:///Users/chronia/CHOP/software/clonevol/clonevol.pdf

# AND how to set up parents - - the one here is an example as well

test <- merge_df_all_mut %>%
  filter(Kids_First_Participant_ID == "PT_962TCBVR") %>%
  write_tsv(file.path(results_dir, "PT_962TCBVR.tsv"))
         
# timepoints <- merge_df_all_mut$age_at_event_days_mut   
timepoints <- c(0, 1717, 2402)

#provide a matrix with the fraction of each population
#present at each timepoint
#    frac.table = matrix(test$CCF,
#      ncol=length(timepoints))

frac.table = matrix(
      c(88, 10, 00,
         12, 00, 40, 
         00, 07, 45),
      ncol=length(timepoints))

    #provide a vector listing each clone's parent
    #(0 indicates no parent)
    parents = c(0,0,0)

    # I think that fishplot is random-number function
    # under the same parents vector listing I got errors
    set.seed(1234)
    #create a fish object
    fish = createFishObject(frac.table,parents,timepoints=timepoints, fix.missing.clones=TRUE)

    #calculate the layout of the drawing
    fish = layoutClones(fish)

    #draw the plot, using the splining method (recommended)
    #and providing both timepoints to label and a plot title
   pdf("./plots/fishplot_PT_962TCBVR.pdf",width=8,height=4)
   print(fishPlot(fish,shape="spline",title.btm="PT_962TCBVR",
             cex.title=0.5, vlines=c(0,1717, 2402), 
             vlab=c("day 0","day 1717", "day 2402")))
   dev=dev.off()
```




```{r}
sessionInfo()
```
  