---
title: "Inference of subclonal architecture of tumors across multiple timepoints in the paired longitudinal (PR) cohort"
author: 'Antonia Chroni <chronia@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---

# Background

This notebook is for preparing input files for pyclone-vi (https://github.com/Roth-Lab/pyclone-vi). The same files will be used as an input for FastClone (https://github.com/GuanLab/FastClone_GuanLab).


# Set up

```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
  library(readxl)
})
```

# Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory
# Use this as the root directory to ensure proper sourcing of functions 
# no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
analysis_dir <- file.path(root_dir, "analyses", "tumor-clone-inference")
input_dir <- file.path(analysis_dir, "input")
data_dir <- file.path(root_dir, "data")
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")
maf_files_dir <- file.path(root_dir, "analyses", "tmb-vaf-longitudinal", "results")

# Input files
#pbta_file <- file.path(files_dir, "pbta.tsv") # file from add-sample-distribution module
maf_file <- file.path(maf_files_dir, "tmb_vaf_genomic.tsv")
genomic_paired_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") # file from add-sample-distribution module
nautilus_dec_file <- file.path(input_dir, "deceased_samples.xlsx") 

# EXAMPLE 
# But to replace with dir with cns data inferred by CNVkit
PT_Z4BF2NSB_dir <- file.path(input_dir, "cnvkit_data_example/PT_Z4BF2NSB") 

# File path to results directory
results_dir <-
  file.path(analysis_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# File path to input directory
pyclonevi_input_dir <-
  file.path(analysis_dir, "results", "pyclonevi_input")
if (!dir.exists(pyclonevi_input_dir)) {
  dir.create(pyclonevi_input_dir)
}
```

# Load and process data

```{r load-process-inputs}
# Read maf
maf_df <- readr::read_tsv(maf_file, guess_max = 100000, show_col_types = FALSE) %>%
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, cg_id, tumor_descriptor,
         sample_id, aliquot_id, Chromosome, 
         Start_Position, Reference_Allele, Tumor_Seq_Allele2, 
         t_ref_count, t_alt_count, tmb, 
         tumor_fraction, tumor_ploidy) %>% 

  # Remove hypermutants
  filter(!tmb >= 10, 
  
  # There are alterations with high number of read counts.
  # We will exclude those with >1000 for now.
         !t_alt_count >= 1000,
         !t_ref_count >= 1000)  %>% 
  
  # Add `normal_cn`: Total copy number of segment in healthy tissue. 
  # For autosome this will be two and male sex chromosomes one.
  # See: https://github.com/Roth-Lab/pyclone-vi
  mutate(normal_cn = case_when(grepl("chrY", Chromosome) ~ "1", 
                                     TRUE ~ "2"))
  
# Read patient list
genomic_paired_df <- readr::read_tsv(genomic_paired_file, guess_max = 100000, show_col_types = FALSE) %>% 
  select(!c(cancer_group, experimental_strategy)) %>% 
  left_join(maf_df, by = c("Kids_First_Participant_ID")) 


# Let's count #specimens per sample 
# We will remove any samples with less than 2 specimens as phylogenies require at least 3 taxa
kids_specimens_n_df <- genomic_paired_df %>% 
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, tumor_descriptor) %>% 
  unique() %>% 
  dplyr::count(Kids_First_Participant_ID) %>% 
  dplyr::mutate(kids_specimens_n = glue::glue("{Kids_First_Participant_ID}  (N={n})")) %>%
  dplyr::rename(kids_specimens_number = n) %>% 
  filter(!kids_specimens_number <= 2) %>%

  left_join(genomic_paired_df, by = c("Kids_First_Participant_ID")) %>%
  #left_join(maf_df) %>% 
  filter(!is.na(Chromosome)) 
  
# Let's confirm and add the number of timepoints per sample
# In case we lost any from filtering hypermutants and high reads/alteration
timepoints_n_df <- kids_specimens_n_df %>% 
  select(Kids_First_Participant_ID, tumor_descriptor) %>% 
  unique() %>% 
  dplyr::count(Kids_First_Participant_ID) %>% 
  dplyr::mutate(kids_timepoints_n = glue::glue("{Kids_First_Participant_ID}  (N={n})")) %>%
  dplyr::rename(kids_timepoints_number = n) %>% 
  filter(!kids_timepoints_number <= 1)

df <- timepoints_n_df %>% 
  left_join(kids_specimens_n_df, by = c("Kids_First_Participant_ID")) %>% 
  write_tsv(file.path(results_dir, "samples_eligible_for_phylogeny.tsv"))

# List with samples eligible for phylogeny
# I added the information about to use for phylogenetic inferences
list_df <- df %>% 
  select(Kids_First_Participant_ID) %>% 
  unique() %>% 
  mutate(somatic_germline_phylogeny = case_when(grepl("PT_KZ56XHJT|PT_KTRJ8TFY", Kids_First_Participant_ID) ~ "yes",
                                        TRUE ~ "not_yet")) %>% 
  write_tsv(file.path(results_dir, "samples_eligible_for_phylogeny_list.tsv"))

```

## Add Nautilus location for Deceased specimens

```{r add-Nautilus-location}
nautilus_dec_df <- read_excel(nautilus_dec_file) %>% 
  right_join(df, by = c("sample_id", "aliquot_id", "tumor_descriptor")) %>% 
  write_tsv(file.path(results_dir, "nautilus_dec.tsv"))

list_df <- nautilus_dec_df %>% 
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, `Note field from Nautilus of initial parent`) %>% 
  unique() %>% 
  write_tsv(file.path(results_dir, "nautilus_dec_list.tsv"))

```

## Other, maybe to delete

``` {r other-things}

# Make list with samples with >2 biospecimens at Deceased timepoint
dec_n_df <- df %>% 
  filter(tumor_descriptor == "Deceased") %>% 
  select(Kids_First_Participant_ID, tumor_descriptor, Kids_First_Biospecimen_ID) %>% 
  unique() %>% 
  dplyr::count(Kids_First_Participant_ID) %>% 
 # dplyr::mutate(kids_dec_n = glue::glue("{Kids_First_Participant_ID}  (N={n})")) %>%
  dplyr::rename(kids_deceased_bs_number = n) %>% 
  filter(!kids_deceased_bs_number <= 1) %>% 
  write_tsv(file.path(results_dir, "kids_dec_multiple_bs_list.tsv"))

###-----------------------------------------------------
# let's look into PT_3KM9W8S8
PT_3KM9W8S8_df <- nautilus_dec_df %>% 
  filter(Kids_First_Participant_ID == "PT_3KM9W8S8",
         Kids_First_Biospecimen_ID == "BS_2NQXY528") %>% 
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, `Note field from Nautilus of initial parent`) %>% 
  unique() %>% 
  write_tsv(file.path(results_dir, "nautilus_dec_list.tsv"))
  

  
bs_id <- unique(PT_3KM9W8S8_df$Kids_First_Biospecimen_ID)


```


# Create pyclone input files 

We need to generate the input files according to the method's template. 
Phylogenetic methods require at least 2 samples per tumor site (multiregional sampling per anatomical site).
Here, we will consider kids samples with more than 2 timepoints with one or more biospecimens. 
We will compare later differences in samples with single vs multiple biospecimens.

```{r create-pyclone-all-samples-df}
# Create pyclone df for all samples
pyclone_all_samples_df <- nautilus_dec_df %>%
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, cg_id, 
         tumor_descriptor, Chromosome, Start_Position, 
         Reference_Allele, t_ref_count, t_alt_count,
         normal_cn, tumor_fraction) %>% 

  # create unique identifiers
  dplyr::mutate(mutation_id = paste(Kids_First_Participant_ID, Chromosome, Start_Position, Reference_Allele, sep = ":"),
                cg_id_kids = paste(cg_id, Kids_First_Participant_ID, sep = "_"),
                cg_id_kids = str_replace(cg_id_kids, "/|-", "_"),
                cg_id_kids = str_replace_all(cg_id_kids, " ", "_"),
                cg_id = str_replace(cg_id, "/|-", "_"),
                cg_id = str_replace_all(cg_id, " ", "_"),
                sample_id_bs = paste(tumor_descriptor, Kids_First_Biospecimen_ID, sep = ":")) %>% 
  
  # change names to match input requirements
  dplyr::rename("ref_counts" = "t_ref_count", 
                "alt_counts" = "t_alt_count",
                "tumour_content" = "tumor_fraction") %>%
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID,
         cg_id_kids, mutation_id, sample_id_bs, ref_counts, 
         alt_counts, normal_cn, tumour_content)
 

# I will test one HGG dataset for now
# PT_Z4BF2NSB

PT_Z4BF2NSB_DF <- pyclone_all_samples_df %>% 
  filter(Kids_First_Participant_ID == "PT_Z4BF2NSB") %>% 
  mutate(sample_id = case_when(grepl("Deceased:BS_2T4EJ6KN", sample_id_bs) ~ "Deceased1",
                               grepl("Deceased:BS_537YFJ06", sample_id_bs) ~ "Deceased2", 
                               
                               grepl("Deceased:BS_C6T8F9K7", sample_id_bs) ~ "Deceased3",
                               grepl("Deceased:BS_GSXNQNRY", sample_id_bs) ~ "Deceased4",
                               grepl("Deceased:BS_M29BNE7Z", sample_id_bs) ~ "Deceased5",
                               grepl("Deceased:BS_MRN9VDQ0", sample_id_bs) ~ "Deceased6",
                               grepl("Diagnosis:BS_W2QCHQ7E", sample_id_bs) ~ "Diagnosis1",
                               grepl("Recurrence:BS_EJP43CD9", sample_id_bs) ~ "Recurrence1"))
                               
                   
```

## Add major_cn, minor_cn columns from cns files

```{r process-cns-files}
data_dir <- dir(path = PT_Z4BF2NSB_dir,  pattern = ".call.cns", full.names = TRUE, recursive = TRUE)

# Create list 
data_list <- list()

for (i in 1:length(data_dir) ) { 
  data_list[[i]] <- read.csv(data_dir[i], header=T, sep="\t") 
   
  # Create file_name
  file_name <- gsub(".call.cns", "", str_split_fixed(data_dir[i], "/", 13)[,12])
  print(file_name)
  
  data_list[[i]] <- data_list[[i]] %>% 
    mutate(Kids_First_Biospecimen_ID = file_name)
  
  # The following code assigns name to df
  # But adds an extra df into the list - dont know yet why but let's remove, it's not necessary
  #df <- assign(file_name, data_list) 
  #data_list[[file_name]] <- df

}

# Bind all df from list 
data_list_bind <- dplyr::bind_rows(data_list) 

# Create and save pyclone_input!
pyclone_input <- data_list_bind %>% 
  
  # Rename to match input format
  # To figure out if the assignment is correct
  dplyr::rename("major_cn" = "cn1", 
                "minor_cn" = "cn2") %>% 

  left_join(PT_Z4BF2NSB_DF) %>% 
  filter(!is.na(major_cn),
         !is.na(minor_cn)) %>% 
  select(cg_id_kids, mutation_id, sample_id, ref_counts, 
         alt_counts, normal_cn, major_cn, minor_cn, tumour_content) %>% 
  write_tsv(file.path(pyclonevi_input_dir, "PT_Z4BF2NSB.tsv"))
  #write_delim(file.path(pyclonevi_input_dir, "PT_Z4BF2NSB.tsv"))
```


```{r}
sessionInfo()
```


