---
title: "Inference of subclonal architecture of tumors across multiple timepoints for the PBTA Cohort"
author: 'Antonia Chroni <chronia@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---

# Set up

```{r sload-library}
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(R.utils)
  library(clonevol)
  library(devtools)
  library(purrr)
  library(flextable)
  library(plyr)
})
```

# Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory
# Use this as the root directory to ensure proper sourcing of functions 
# no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
analysis_dir <- file.path(root_dir, "analyses", "tumor-clone-inference")
input_dir <- file.path(analysis_dir, "input")
data_dir <- file.path(root_dir, "data")
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")
maf_files_dir <- file.path(root_dir, "analyses", "tmb-vaf-longitudinal", "results")

# Input files
pbta_file <- file.path(files_dir, "pbta.tsv") # file from add-sample-distribution module
maf_file <- file.path(maf_files_dir, "maf.tsv")
genomic_paired_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") # file from add-sample-distribution module

# File path to results directory
results_dir <-
  file.path(analysis_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# File path to plots directory
plots_dir <-
   file.path(analysis_dir, "plots")
 if (!dir.exists(plots_dir)) {
   dir.create(plots_dir)
 }
```

# Load data

```{r load-inputs-please-wait}
maf_df <- readr::read_tsv(maf_file, guess_max = 100000, show_col_types = FALSE)

pbta_df <- readr::read_tsv(pbta_file, guess_max = 100000, show_col_types = FALSE)

df <- readr::read_tsv(genomic_paired_file, guess_max = 100000, show_col_types = FALSE) %>% 
  select(!cancer_group) %>% 
  left_join(pbta_df, by = c("Kids_First_Participant_ID", "experimental_strategy"))
```

# Process Data for CloneFinder

We need to generate the input files for CloneFinder according to the method's template.

## Step 1

First though, we need to ensure that we have enough number of tumor samples per time point and patient case.
This is because CloneFinder requires at least 2 tumor samples per anatomical site.
Thus, we will first identify the number of tumor samples per patient and we will keep only those with more than 2 tumor samples.

## Step 2
Input files for CloneFinder should contain the read counts for reference and alteration per each chromosomal position.
We will split these per patient sample based on the filtering from step 1.

```{r Process Data for CloneFinder-Step2}
# we will replace names in the column to follow input guidelines for CloneFinder

read_counts_all_samples <- subset(snv_subset, 
                             select = c(Kids_First_Biospecimen_ID, Chromosome, Start_Position, Reference_Allele, Tumor_Seq_Allele2, t_ref_count, t_alt_count))

read_counts_all_samples <- setnames(read_counts_all_samples, old = c('Chromosome','Start_Position','Reference_Allele', "Tumor_Seq_Allele2", "t_ref_count", "t_alt_count"), 
         new = c('Chr','Position','Wild', "Mut", "XX:ref", "XX:alt")) 

read_counts_all_samples_keep <- read_counts_all_samples %>% 
  left_join(keep, by=c('Kids_First_Biospecimen_ID')) %>%
  filter(!is.na(Kids_First_Participant_ID))
  
  
 # write_tsv(file.path(results_dir, "read_counts_all_samples_keep.tsv")) # this is a large file, if not needed in later stages, maybe there is no need to save
```

We will split and save read count data by the "Kids_First_Participant_ID" column

```{r echo=TRUE}
# Create dir to save read count data to be used as input for CloneFinder
temp_path_csv <-
  file.path(results_dir, "01_Read_count_Per_Patient")
if (!dir.exists(temp_path_csv)) {
  dir.create(temp_path_csv)
}

# we will change the dir here otherwise files are saved outside of the dsired folder
setwd(temp_path_csv)

# we will split Read_count data by the "Kids_First_Participant_ID.x" column
read_counts_all_samples_keep %>% 
  group_split(Kids_First_Participant_ID) %>% 
  walk(~write_csv(.x, paste0("01_", .x$Kids_First_Participant_ID[1], ".csv")))

# let's change back to the root_dir for the rest of the module
setwd(root_dir)
```

## Step 3 

Next we need to create different columns assigning the read counts for ref and alt per tumor sample ("Kids_First_Biospecimen_ID").
We will do this in three steps: (1) create wide df for ref variant, (2) create wide df for alt variant, and (3) merge wide df.

```{r Process Data for CloneFinder-Step3}
#create out_dir
    temp_path <-
      file.path(results_dir, "02_CloneFinder_input_files")
    if (!dir.exists(temp_path)) {
      dir.create(temp_path)
    } 

# create a list and read csv files
all_files <-  list.files(path = temp_path_csv, # Identify all CSV files
                       pattern = "*.csv", full.names = TRUE)
data_all <- lapply(all_files, read.csv)

##Create list of data frame names without the "CloneFinder_input_read_count_csv." part 
# get only the file names, not the full path.
names(data_all) <- gsub("01_", "",
                       list.files(temp_path_csv,full.names = FALSE),
                       fixed = TRUE)

# we will change the dir here otherwise files are saved outside of the desired folder
setwd(temp_path)

# create a loop for the elements in the 'all_files' list
# and create input files

for(i in 1:length(data_all)) {  
# i = "PT_KTRJ8TFY.csv"

# we need to have separate columns with the tumor sample containing read counts per variant
  
#####################
# we will first create a data frame for the read counts for the reference and alteration variant
# create unique identifier based on chromosome and chromosomal position for each variant
# let's also keep the information with tumor sample and time point together. This is useful for visualization purposes later.
# XX:ref  
  
df <- data_all[[i]] %>%
       mutate(match_id_unique = paste(Chr, Position, sep = "_")) %>%
       mutate(Kids_First_Biospecimen_ID_descriptor = paste(Kids_First_Biospecimen_ID, tumor_descriptor, sep = "_"))

  
df_ref <- df %>%
       mutate(Kids_First_Biospecimen_ID_descriptor = paste(Kids_First_Biospecimen_ID_descriptor, sep = ":", "ref")) %>% 
       subset(select = c(match_id_unique, Kids_First_Biospecimen_ID_descriptor, XX.ref)) 

# then we need to transform the data from long to wide
df_ref_wide <- list(dcast(setDT(df_ref), match_id_unique~Kids_First_Biospecimen_ID_descriptor))


#####################
# then we need to transform the data from long to wide
# XX:alt
df_alt <- df %>%
       mutate(Kids_First_Biospecimen_ID_descriptor = paste(Kids_First_Biospecimen_ID_descriptor, sep = ":", "alt")) %>% 
       subset(select = c(match_id_unique, Kids_First_Biospecimen_ID_descriptor, XX.alt)) 

# then we need to transform the data from long to wide
df_alt_wide = list(dcast(setDT(df_alt),match_id_unique~Kids_First_Biospecimen_ID_descriptor))

#####################
# merge data_all[[i]]$df_ref with data_all[[i]]$df_alt
# add columns: 'Chr','Position','Wild', "Mut" that are missing so it matches the CloneFinder input

#data_all[[i]]$merge_read_counts <- list(data_all[[i]]$df_ref_wide, left_join(data_all[[i]]$df_alt_wide)) %>% 
#                                          reduce(left_join, by=c('match_id_unique'))  %>% 
#                                          mutate(Chr =gsub("_..*", "", match_id_unique)) %>% 
#                                          mutate(Position =gsub("_", "", match_id_unique)) 

left_join_ref_alt_wide <- left_join(as.data.frame(df_ref_wide),as.data.frame(df_alt_wide), by = "match_id_unique") %>%
                          mutate(
                            Chr =gsub("_..*", "", 'match_id_unique'), 
                            Position =gsub("_", "", 'match_id_unique')
                          ) 
merge_read_counts <- list(left_join_ref_alt_wide)

# we will replace any NAs with zero
merge_read_counts[is.na(merge_read_counts)] <- 0  
  
# save edited files
write.table(merge_read_counts, paste0(temp_path, "_edited.txt"))
} 

# let's change back to the root_dir for the rest of the module
setwd(root_dir)
```

## Step 4 

```{r Process Data for CloneFinder-Step4}
# let's create a list of the samples and their directory path
# we will use this to run CloneFinder for all samples together

distinct_samples <- read_counts_all_samples_keep %>% distinct(Kids_First_Participant_ID)                   #removes the duplicate values
samples_list_path <-  subset(distinct_samples, 
                             select = c(Kids_First_Participant_ID)) %>%
           mutate(path = temp_path,
                 additional = 1) %>%
  write_csv(file.path(results_dir, "distinct_samples_list.csv"))
```


```{r echo=TRUE}
# Create directory to save CloneFinder results
CloneFinder_results <-
  file.path(results_dir, "CloneFinder_results")
if (!dir.exists(CloneFinder_results)) {
  dir.create(CloneFinder_results)
}
```

```{r}
sessionInfo()
```