---
title: "Inference of subclonal architecture of tumors across multiple timepoints in the paired longitudinal (PL) cohort"
author: 'Antonia Chroni <chronia@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---

# Background

This notebook is for preparing input files for [pyclone-vi](https://github.com/Roth-Lab/pyclone-vi). The same files with a minor modification will be used as an input for [FastClone](https://github.com/GuanLab/FastClone_GuanLab).


# Set up

```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
  library(readxl)
})
```

# Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory
# Use this as the root directory to ensure proper sourcing of functions 
# no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
analysis_dir <- file.path(root_dir, "analyses", "tumor-clone-inference")
input_dir <- file.path(analysis_dir, "input")
data_dir <- file.path(root_dir, "data")
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")
maf_files_dir <- file.path(root_dir, "analyses", "tmb-vaf-longitudinal", "results")

# Input files
pbta_file <- file.path(files_dir, "pbta.tsv") # file from add-sample-distribution module
maf_file <- file.path(data_dir, "snv-consensus-plus-hotspots.maf.tsv.gz")
tmb_file <- file.path(maf_files_dir, "tmb_vaf_genomic.tsv")
genomic_paired_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") # file from add-sample-distribution module
nautilus_dec_file <- file.path(input_dir, "deceased_samples.xlsx") 
palette_file <- file.path(root_dir, "figures", "palettes", "tumor_descriptor_color_palette.tsv")
cns_dir <- file.path(input_dir, "cnvkit_data") 

# File path to results directory
results_dir <-
  file.path(analysis_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# File path to input directory
pyclonevi_input_dir <-
  file.path(analysis_dir, "results", "pyclone-vi-input-synonymous")
if (!dir.exists(pyclonevi_input_dir)) {
  dir.create(pyclonevi_input_dir)
}

# File path to input directory
fastclone_input_dir <-
  file.path(analysis_dir, "results", "fastclone-input-synonymous")
if (!dir.exists(pyclonevi_input_dir)) {
  dir.create(pyclonevi_input_dir)
}


# File path to plot directory
pyclone_plots_dir <-
  file.path(analysis_dir, "plots", "pyclone-vi-synonymous")
if (!dir.exists(pyclone_plots_dir)) {
  dir.create(pyclone_plots_dir)
}


source(paste0(root_dir, "/figures/scripts/theme.R"))
```

# Load and process data

```{r load-process-inputs}
# Read in histologies file and filter for the pbta cohort
pbta <- readr::read_tsv(pbta_file, guess_max = 100000, show_col_types = FALSE) %>%
  filter(!(experimental_strategy == "RNA-Seq")) %>% 
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, cg_id, tumor_descriptor,
         sample_id, aliquot_id,
         tumor_fraction, tumor_ploidy)

# Read in tmb_all file
tmb_df <- readr::read_tsv(tmb_file, guess_max = 100000, show_col_types = FALSE) 

# Read maf
# Keep only synonymous mutations for the current analysis
maf_file_df <- vroom::vroom(maf_file, delim = "\t", col_names = TRUE)

# Make a vector of coding gene mutations
maf_nonsynonymous <- c(
  "Missense_Mutation",
  "Frame_Shift_Del",
  "In_Frame_Ins",
  "Frame_Shift_Ins",
  "Splice_Site",
  "Nonsense_Mutation",
  "In_Frame_Del",
  "Nonstop_Mutation",
  "Translation_Start_Site"
)

# Process maf
maf_df <- maf_file_df %>%
  filter(Tumor_Sample_Barcode %in% pbta$Kids_First_Biospecimen_ID,
         Variant_Classification %in% maf_nonsynonymous
         ) %>%
  dplyr::rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>% 
  left_join(pbta) %>%
  left_join(tmb_df) %>%

  # Calculate VAF 
  dplyr::mutate(VAF = t_alt_count / (t_ref_count + t_alt_count), 
                # Concatenate gene and protein information
                gene_protein = paste(Hugo_Symbol, HGVSp_Short, sep = "_")) %>%  
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, cg_id, tumor_descriptor,
         sample_id, aliquot_id, Chromosome, 
         Start_Position, Reference_Allele, Tumor_Seq_Allele1, Tumor_Seq_Allele2,
         t_ref_count, t_alt_count, 
         tumor_fraction, tumor_ploidy, VAF, gene_protein, Hugo_Symbol,
         mutation_count, tmb) %>% 

  # Remove hypermutants
  filter(!tmb >= 10 #, 
  
  # There are alterations with high number of read counts.
  # We will exclude those with >1000 for now.
        # !t_alt_count >= 1000,
         #!t_ref_count >= 1000
  )  %>% 
  
  # Add `normal_cn`: Total copy number of segment in healthy tissue. 
  # For autosome this will be two and male sex chromosomes one.
  # See: https://github.com/Roth-Lab/pyclone-vi
  mutate(normal_cn = case_when(grepl("chrY", Chromosome) ~ "1", 
                                     TRUE ~ "2")) %>% 
  
  # Calculate CCF
  # CCF() %>% 
  # using formula for CCF calculation
  mutate(CCF = (2*t_alt_count)/ (t_ref_count + t_alt_count)) #%>% 
  #write_tsv(file.path(scratch_dir, "maf_pbta.tsv"))

# Number of samples in the `maf_df`
maf_samples <- print(length(unique(maf_df$Kids_First_Participant_ID)))

# Read patient list
genomic_paired_df <- readr::read_tsv(genomic_paired_file, guess_max = 100000, show_col_types = FALSE) %>% 
  select(!c(cancer_group, experimental_strategy)) %>% 
  left_join(maf_df, by = c("Kids_First_Participant_ID")) %>% # create unique identifiers
  dplyr::mutate(mutation_id = paste(Kids_First_Participant_ID, tumor_descriptor, Kids_First_Biospecimen_ID, Chromosome, Start_Position, Reference_Allele, Tumor_Seq_Allele2, sep = ":"),
                cg_id_kids = paste(cg_id, Kids_First_Participant_ID, sep = "_"),
                cg_id_kids = str_replace(cg_id_kids, "/|-", "_"),
                cg_id_kids = str_replace_all(cg_id_kids, " ", "_"),
                cg_id = str_replace(cg_id, "/|-", "_"),
                cg_id = str_replace_all(cg_id, " ", "_"),
                sample_id = paste(tumor_descriptor, Kids_First_Biospecimen_ID, sep = ":")) #%>% 
  #filter(!is.na(VAF) #,
         #!VAF == 0 #, 
         #!is.na(tmb)
        # )

# Number of samples in the `genomic_paired_df`
genomic_samples <- print(length(unique(genomic_paired_df$Kids_First_Participant_ID)))

# Let's count #specimens per sample 
# We will remove any samples with less than 2 specimens as phylogenies require at least 3 taxa
kids_specimens_n_df <- genomic_paired_df %>% 
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, tumor_descriptor) %>% 
  unique() %>% 
  dplyr::count(Kids_First_Participant_ID) %>% 
  dplyr::mutate(kids_specimens_n = glue::glue("{Kids_First_Participant_ID}  (N={n})")) %>%
  dplyr::rename(kids_specimens_number = n) %>% 
  filter(!kids_specimens_number <= 2) %>%

  left_join(genomic_paired_df, by = c("Kids_First_Participant_ID")) %>%
  #left_join(maf_df) %>% 
  filter(!is.na(Chromosome)) 


# Let's confirm and add the number of timepoints per sample
# In case we lost any from filtering hypermutants and high reads/alteration
timepoints_n_df <- kids_specimens_n_df %>% 
  select(Kids_First_Participant_ID, tumor_descriptor) %>% 
  unique() %>% 
  dplyr::count(Kids_First_Participant_ID) %>% 
  dplyr::mutate(kids_timepoints_n = glue::glue("{Kids_First_Participant_ID}  (N={n})")) %>%
  dplyr::rename(kids_timepoints_number = n) %>% 
  filter(!kids_timepoints_number <= 1)

df <- timepoints_n_df %>% 
  left_join(kids_specimens_n_df, by = c("Kids_First_Participant_ID")) %>% 
  write_tsv(file.path(results_dir, "samples_eligible_for_phylogeny_synonymous.tsv"))

# Number of samples in the `genomic_paired_df` with > 3 biospecimens and >2 timepoints
samples_with_3bs <- print(length(unique(df$Kids_First_Participant_ID)))
samples_with_3bs <- print(unique(df$Kids_First_Participant_ID))
  
# List with samples eligible for phylogeny
# I added the information about to use for phylogenetic inferences
list_df <- df %>% 
  select(Kids_First_Participant_ID) %>% 
  unique() %>% 
  mutate(somatic_germline_phylogeny = case_when(grepl("PT_KZ56XHJT|PT_KTRJ8TFY", Kids_First_Participant_ID) ~ "yes",
                                        TRUE ~ "not_yet")) %>% 
  write_tsv(file.path(results_dir, "samples_eligible_for_phylogeny_list_synonymous.tsv"))

# Remove large files
rm(genomic_paired_df, kids_specimens_n_df, tmb_df)

```

## Add Nautilus location for Deceased specimens

```{r add-Nautilus-location}
nautilus_dec_df <- read_excel(nautilus_dec_file) %>% 
  right_join(df, by = c("sample_id", "aliquot_id", "tumor_descriptor")) %>%
  dplyr::mutate(mutation_id = paste(Kids_First_Participant_ID, tumor_descriptor, Kids_First_Biospecimen_ID, Chromosome, Start_Position, Reference_Allele, Tumor_Seq_Allele2, sep = ":")) %>% 
  write_tsv(file.path(results_dir, "nautilus_dec_synonymous.tsv"))

list_df <- nautilus_dec_df %>% 
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, `Note field from Nautilus of initial parent`) %>% 
  unique() %>% 
  write_tsv(file.path(results_dir, "nautilus_dec_list_synonymous.tsv"))

```

## Other, maybe to delete

``` {r other-things}

# Make list with samples with >2 biospecimens at Deceased timepoint
dec_n_df <- df %>% 
  filter(tumor_descriptor == "Deceased") %>% 
  select(Kids_First_Participant_ID, tumor_descriptor, Kids_First_Biospecimen_ID) %>% 
  unique() %>% 
  dplyr::count(Kids_First_Participant_ID) %>% 
 # dplyr::mutate(kids_dec_n = glue::glue("{Kids_First_Participant_ID}  (N={n})")) %>%
  dplyr::rename(kids_deceased_bs_number = n) %>% 
  filter(!kids_deceased_bs_number <= 1) %>% 
  write_tsv(file.path(results_dir, "kids_dec_multiple_bs_list_synonymous.tsv"))
```


# Create pyclone input files 

We need to generate the input files according to the method's template. 
Phylogenetic methods require at least 2 samples per tumor site (multiregional sampling per anatomical site).
Here, we will consider kids samples with more than 2 timepoints with one or more biospecimens. 
We will compare later differences in samples with single vs multiple biospecimens.

```{r create-pyclone-all-samples-df}
# Create pyclone df for all samples
pyclone_all_samples_df <- nautilus_dec_df %>%
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, cg_id, 
         cg_id_kids, mutation_id, sample_id,
         tumor_descriptor, Chromosome, Start_Position, 
         Reference_Allele, Tumor_Seq_Allele2, t_ref_count, t_alt_count,
         normal_cn, tumor_fraction, mutation_count, Hugo_Symbol) %>% 
  
  # change names to match input requirements
  dplyr::rename("ref_counts" = "t_ref_count", 
                "alt_counts" = "t_alt_count",
                "tumour_content" = "tumor_fraction") 

samples_pyclone <- print(unique(pyclone_all_samples_df$Kids_First_Participant_ID))

# Remove large files
# rm(nautilus_dec_df)

# I will test one HGG dataset for now
# PT_Z4BF2NSB

#PT_Z4BF2NSB_DF <- pyclone_all_samples_df %>% 
#  filter(Kids_First_Participant_ID == "PT_Z4BF2NSB")
                   
```

## Add major_cn, minor_cn columns from cns files

```{r process-cns-files}
data_dir <- dir(path = cns_dir,  pattern = ".call.cns", full.names = TRUE, recursive = TRUE)

# Create list 
data_list <- list()

for (i in 1:length(data_dir) ) { 
  
  # Create sample_name
  sample_name <-  unique(as.character(gsub(".call.cns", "", str_split_fixed(data_dir[i], "/", 13)[,11])))
  #sample_name <-  unique(as.character(gsub(".call.cns", "", str_split_fixed(data_dir[i], "/", 10)[,9])))
  sample_name <- sort(sample_name, decreasing = FALSE)
  print(sample_name)
  
  for (x in seq_along(sample_name) ) { 
    
    data_list[[i]] <- read.csv(data_dir[i], header=T, sep="\t") 
  
    
  # Create file_name
  file_name <- gsub(".call.cns", "", str_split_fixed(data_dir[i], "/", 13)[,12])
  print(file_name)
  
  data_list[[i]] <- data_list[[i]] %>% 
    mutate(Kids_First_Biospecimen_ID = file_name)
  
  # The following code assigns name to df
  # But adds an extra df into the list - dont know yet why but let's remove, it's not necessary
  #df <- assign(file_name, data_list) 
  #data_list[[file_name]] <- df
  }
}

# Bind all df from list 
data_list_bind <- dplyr::bind_rows(data_list) 

# Remove large files
rm(data_list)

# Create and save pyclone_input!
pyclone_input <- data_list_bind %>% 
  
  # Rename to match input format
  # To figure out if the assignment is correct
  dplyr::rename("major_cn" = "cn1", 
                "minor_cn" = "cn2") %>% 

  left_join(pyclone_all_samples_df) %>% 
  
  filter(!is.na(major_cn),
         !is.na(minor_cn)) %>%
  #dplyr::mutate(mutation_id = paste(mutation_id, major_cn, minor_cn, ref_counts, alt_counts, sep = ":")) %>% 
  dplyr::mutate(mutation_id = paste(Kids_First_Participant_ID, tumor_descriptor, Kids_First_Biospecimen_ID, Chromosome, Start_Position, Reference_Allele, Tumor_Seq_Allele2, sep = ":")) %>% 
  select(Kids_First_Participant_ID, tumor_descriptor, cg_id_kids, mutation_id, sample_id, ref_counts, 
         alt_counts, normal_cn, major_cn, minor_cn, tumour_content, mutation_count, Hugo_Symbol) %>% 
  
  # To ensure there are no duplicated entries in the dataframe
  distinct() %>% 
  filter(!is.na(Kids_First_Participant_ID)) 

# Save df with all data together
# We will need this to prepare the input fiel for ClonEvol
pyclone_input_all <- nautilus_dec_df %>%
  select(mutation_id, VAF, CCF, gene_protein, Hugo_Symbol) %>% 
  right_join(pyclone_input) %>% 
  write_tsv(file.path(results_dir, "pyclone_input_all_data_synonymous.tsv"))


samples_pyclone <- print(unique(pyclone_input$Kids_First_Participant_ID))


# To find the position of duplicate elements in x, use this:
# duplicated(pyclone_input)

# Extract duplicate elements:
# pyclone_input[duplicated(pyclone_input)]

```

# Save input files

```{r save-input-files}
for (i in 1:length(data_dir) ) { 
  
  # Create sample_name
  sample_name <-  unique(as.character(gsub(".call.cns", "", str_split_fixed(data_dir[i], "/", 13)[,11])))
  #sample_name <-  unique(as.character(gsub(".call.cns", "", str_split_fixed(data_dir[i], "/", 10)[,9])))
  sample_name <- sort(sample_name, decreasing = FALSE)
  print(sample_name)
  
for (x in seq_along(sample_name) ) { 
  
  # Save input file for pyclone-vi
  pyclone_fname <- paste0(pyclonevi_input_dir, "/", sample_name[x], ".tsv")
  print(pyclone_fname)
  
  pyclone_input_subset <- pyclone_input %>%
    filter(Kids_First_Participant_ID == sample_name[x]) %>% 
    #select(-c(Kids_First_Participant_ID)) %>% 
  write_tsv(file.path(pyclone_fname))
  
  
  # Save input file for FastClone
  fastclone_fname <- paste0(fastclone_input_dir, "/", sample_name[x], ".tsv")
  print(fastclone_fname)
  
  fastclone_input_subset <- pyclone_input %>%
    dplyr::rename("var_counts" = "alt_counts") %>% 
    filter(Kids_First_Participant_ID == sample_name[x]) %>% 
    #select(-c(Kids_First_Participant_ID)) %>% 
  write_tsv(file.path(fastclone_fname))
  
  }
}

```

# Plot depth coverage

```{r define-parameters-for-plots}
# Read color palette
palette_df <- readr::read_tsv(palette_file, guess_max = 100000, show_col_types = FALSE) %>% 
  mutate(tumor_descriptor = color_names)

# Define and order palette
palette <- palette_df$hex_codes
names(palette) <- palette_df$tumor_descriptor

# Define timepoints
timepoints = c("Diagnosis", "Progressive", "Recurrence", "Deceased", "Second Malignancy", "Unavailable")

```

``` {r depth-coverage, fig.width = 6, fig.height = 5, fig.fullwidth = TRUE}
for (i in 1:length(data_dir) ) { 
  
  # Create sample_name
  sample_name <-  unique(as.character(gsub(".call.cns", "", str_split_fixed(data_dir[i], "/", 13)[,11])))
  sample_name <- sort(sample_name, decreasing = FALSE)
  print(sample_name)
  
for (x in seq_along(sample_name) ) { 
  
  pyclone_input_subset <- pyclone_input %>%
    filter(Kids_First_Participant_ID == sample_name[x]) %>% 
    select(-c(Kids_First_Participant_ID)) #%>% 
   # mutate(tumor_descriptor = factor(tumor_descriptor),
    #     tumor_descriptor = fct_relevel(tumor_descriptor, timepoints)) %>% 
    #arrange(tumor_descriptor, sample_id)
  
  # Make this reproducible
  set.seed(2023)

  # Define label for plots
  Timepoint <- factor(x = pyclone_input_subset$tumor_descriptor, levels = timepoints)
  
  #######################
  # Create bxp ref_counts
    p <- print(ggplot(pyclone_input_subset, aes(sample_id, ref_counts, color = Timepoint)) + 
                 geom_jitter(width = 0.15, size = 0.7, alpha = 0.6) +
                 ggplot2::geom_boxplot(color = "black",
                              size = 0.25,
                              alpha = 0,
                              coef = 0) + # remove whiskers
                 theme_Publication() + 
                 scale_color_manual(values = palette,
                                    breaks = sort(names(palette))) +
                 #rotate() +
                 theme(axis.text.x = element_text(angle = 90)) +
                 stat_summary(fun.y=mean,shape=1,col='black',geom='point') +
                 labs(title = sample_name[x],
                      x = "sample_id",
                      y = "ref_counts",
                      color = "Timepoint"))
    
    # Save the plot
    ggsave(filename = paste0(sample_name[x], "-ref_counts.pdf"), 
         path = pyclone_plots_dir, 
         width = 6, 
         height = 5, 
         device = "pdf", 
        useDingbats = FALSE)
    

  #######################
  # Create bxp alt_counts
    p <- print(ggplot(pyclone_input_subset, aes(sample_id, alt_counts, color = Timepoint)) + 
                 geom_jitter(width = 0.15, size = 0.7, alpha = 0.6) +
                 ggplot2::geom_boxplot(color = "black",
                              size = 0.25,
                              alpha = 0,
                              coef = 0) + # remove whiskers
                 theme_Publication() + 
                 scale_color_manual(values = palette,
                                    breaks = sort(names(palette))) +
                 #rotate() +
                 theme(axis.text.x = element_text(angle = 90)) +
                 stat_summary(fun.y=mean,shape=1,col='black',geom='point') +
                 labs(title = sample_name[x],
                      x = "sample_id",
                      y = "alt_counts",
                      color = "Timepoint"))
    
    # Save the plot
    ggsave(filename = paste0(sample_name[x], "-alt_counts.pdf"), 
         path = pyclone_plots_dir, 
         width = 6, 
         height = 5, 
         device = "pdf", 
        useDingbats = FALSE)
  
  
}
}

```


# Total number of mutations across timepoints and biospecimen sample per Patient case

```{r create-barplot-sample, fig.width = 6, fig.height = 5, fig.fullwidth = TRUE}
for (i in 1:length(data_dir) ) { 
  
  # Create sample_name
  sample_name <-  unique(as.character(gsub(".call.cns", "", str_split_fixed(data_dir[i], "/", 13)[,11])))
  sample_name <- sort(sample_name, decreasing = FALSE)
  print(sample_name)
  
for (x in seq_along(sample_name) ) { 
  
  pyclone_input_subset <- pyclone_input %>%
    filter(Kids_First_Participant_ID == sample_name[x]) %>% 
    select(-c(Kids_First_Participant_ID))
  
  # Define parameters for function
  ylim = max(pyclone_input_subset$mutation_count)
  
  
   # Rename legend for timepoints
  Timepoint <- factor(pyclone_input_subset$tumor_descriptor)
  
  # Plot stacked barplot 
  print(ggplot(pyclone_input_subset, aes(x = sample_id, 
                                              y = mutation_count, 
                                              fill = Timepoint)) +  
               geom_col(position = position_stack(reverse = TRUE)) +
               geom_bar(stat = "identity", width = 0.5) + 
               scale_fill_manual(values = palette, breaks = sort(names(palette))) + 
               theme_Publication() + 
               theme(axis.text.x = element_text(angle = 85, 
                                                hjust = 1, 
                                                vjust = 1)) + 
               labs(title = paste(sample_name)) + 
               labs(x = "sample_id", y = "Total Mutations") +
               ylim(0, ylim)) 
  
  
}
}
```

```{r}
sessionInfo()
```


