---
title: "Inference of subclonal architecture of tumors across multiple timepoints for the PBTA Cohort"
author: 'Antonia Chroni <chronia@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---

# Set up

```{r sload-library}
suppressPackageStartupMessages({
  library(tidyverse)
  library(data.table)
  library(R.utils)
  library(clonevol)
  library(devtools)
  library(purrr)
  library(flextable)
  library(plyr)
})
```

# Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory
# Use this as the root directory to ensure proper sourcing of functions 
# no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
analysis_dir <- file.path(root_dir, "analyses", "tumor-clone-inference")
input_dir <- file.path(analysis_dir, "input")
data_dir <- file.path(root_dir, "data")
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")
maf_files_dir <- file.path(root_dir, "analyses", "tmb-vaf-longitudinal", "results")

# Input files
pbta_file <- file.path(files_dir, "pbta.tsv") # file from add-sample-distribution module
maf_file <- file.path(maf_files_dir, "maf.tsv")
genomic_paired_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") # file from add-sample-distribution module

# File path to results directory
results_dir <-
  file.path(analysis_dir, "results", "CloneFinder_input")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# File path to plots directory
plots_dir <-
   file.path(analysis_dir, "plots")
 if (!dir.exists(plots_dir)) {
   dir.create(plots_dir)
 }
```

# Load data

```{r load-inputs-please-wait}
maf_df <- readr::read_tsv(maf_file, guess_max = 100000, show_col_types = FALSE) %>% 
  dplyr::rename("Chr" = "Chromosome", 
                "Position" = "Start_Position",
                "Wild" = "Reference_Allele",
                "Mut" = "Tumor_Seq_Allele2",
                "XX_ref" = "t_ref_count", 
                "XX_alt" = "t_alt_count") # change names to match input requirements for Clonefinder

pbta_df <- readr::read_tsv(pbta_file, guess_max = 100000, show_col_types = FALSE) %>% 
  filter(!experimental_strategy == "RNA-Seq")

genomic_paired_df <- readr::read_tsv(genomic_paired_file, guess_max = 100000, show_col_types = FALSE) %>% 
  select(!c(cancer_group, experimental_strategy)) %>% 
  left_join(pbta_df, by = c("Kids_First_Participant_ID")) %>% 
  mutate(cg_id_kids = paste(cg_id, Kids_First_Participant_ID, sep = "_"),
         td_bs_id = paste(tumor_descriptor, Kids_First_Biospecimen_ID, sep = "_"))

df <- genomic_paired_df %>% 
  left_join(maf_df, by = c("Kids_First_Biospecimen_ID")) %>% 
  mutate(mut_id = paste(Chr, Position, Wild, Mut, sep = "_"))

df_samples <- print(length(unique(df$cg_id_kids)))
```

# Process Data for CloneFinder

We need to generate the input files for CloneFinder according to the method's template.

## Step 1

First though, we need to ensure that we have enough number of tumor samples per time point and patient case.
This is because CloneFinder requires at least 2 tumor samples per anatomical site.
Thus, we will first identify the number of tumor samples per patient and we will keep only those with more than 2 tumor samples.

## Step 2
Input files for CloneFinder should contain the read counts for reference and alteration per each chromosomal position.
We will split these per patient sample based on the filtering from step 1.


## Step 3 

Next we need to create different columns assigning the read counts for ref and alt per tumor sample ("Kids_First_Biospecimen_ID").
We will do this in three steps: (1) create wide df for ref variant, (2) create wide df for alt variant, and (3) merge wide df.

```{r df-cg-id-kids}
# Function
create_df <- function(df, samples, kids_id) {
  # Reference df
    ref_df <- df_sub %>%
      filter(td_bs_id == td_bs_id_names) %>%
      mutate(td_bs_id = paste(td_bs_id_names, "ref", sep = "_")) %>% 
      pivot_wider(names_from = "td_bs_id", values_from = "XX_ref") %>% 
      select(!XX_alt)
  
    # Alteration df
    alt_df <- df_sub %>%
      filter(td_bs_id == td_bs_id_names) %>%
      mutate(td_bs_id = paste(td_bs_id_names, "alt", sep = "_")) %>% 
      pivot_wider(names_from = "td_bs_id", values_from = "XX_alt") %>%
      select(!XX_ref)
  
    # Merge df
    merge_df <- ref_df %>% 
      full_join(alt_df, by = c("cg_id_kids", "mut_id", "Chr", "Position", "Wild", "Mut")) %>% 
      as.data.frame() 
    
    return(merge_df)
}
  
# Save results byc g_id_kids
cg_id_kids_names <- unique(as.character(df$cg_id_kids))
cg_id_kids_names <- sort(cg_id_kids_names, decreasing = FALSE)
print(cg_id_kids_names)

#cg_id_kids_names = "Adamantinomatous Craniopharyngioma_PT_CBTW4E3X"

# Loop through cg_sum
for (x in seq_along(cg_id_kids_names)){
  print(x)

  df_sub <- df %>%
    filter(cg_id_kids == cg_id_kids_names[x]) %>% 
    select("cg_id_kids", "td_bs_id", "mut_id", "Chr", "Position", "Wild", "Mut", "XX_ref", "XX_alt")
 
  # We will define samples based on the "Kids_First_Participant_ID" column 
  td_bs_id_names <- unique(as.character(df_sub$td_bs_id))
  print(td_bs_id_names)
  
  # Create empty list for df
  df_list <- list() 
  
  # td_bs_id_names = "Recurrence_BS_KR6PXZGJ"
  # td_bs_id_names = "Diagnosis_BS_SS7E1QKV"
  
  # Run loop for each td_bs_id_names
  for (i in seq_along(td_bs_id_names)) {
    print(i)
    
    # Print df per pair
    m <- create_df(df = df_sub, 
                   samples = td_bs_id_names[i],
                   kids_id = cg_id_kids_names[x])
  
    # Assign name to df and store all df in list
    df_name <- paste(cg_id_kids_names[x], td_bs_id_names[i], sep = "-")
    df <- assign(df_name, m)
    df_list[[df_name]] <- df
    
    # Merge all df per `cg_kids_id`
    #library(powerjoin)
    cg_kids_df <- power_full_join(df_list, by = c("cg_id_kids", "mut_id", "Chr", "Position", "Wild", "Mut")) %>% 
      replace(is.na(.), 0) %>% 
      write_tsv(file.path(results_dir, "test_df.tsv"))

    }
}

  
    

#####################
# merge data_all[[i]]$df_ref with data_all[[i]]$df_alt
# add columns: 'Chr','Position','Wild', "Mut" that are missing so it matches the CloneFinder input

#data_all[[i]]$merge_read_counts <- list(data_all[[i]]$df_ref_wide, left_join(data_all[[i]]$df_alt_wide)) %>% 
#                                          reduce(left_join, by=c('match_id_unique'))  %>% 
#                                          mutate(Chr =gsub("_..*", "", match_id_unique)) %>% 
#                                          mutate(Position =gsub("_", "", match_id_unique)) 

left_join_ref_alt_wide <- left_join(as.data.frame(df_ref_wide),as.data.frame(df_alt_wide), by = "match_id_unique") %>%
                          mutate(
                            Chr =gsub("_..*", "", 'match_id_unique'), 
                            Position =gsub("_", "", 'match_id_unique')
                          ) 
merge_read_counts <- list(left_join_ref_alt_wide)

# we will replace any NAs with zero
merge_read_counts[is.na(merge_read_counts)] <- 0  
  
# save edited files
write.table(merge_read_counts, paste0(temp_path, "_edited.txt"))
} 

# let's change back to the root_dir for the rest of the module
setwd(root_dir)
```

## Step 4 

```{r Process Data for CloneFinder-Step4}
# let's create a list of the samples and their directory path
# we will use this to run CloneFinder for all samples together

distinct_samples <- read_counts_all_samples_keep %>% distinct(Kids_First_Participant_ID)                   #removes the duplicate values
samples_list_path <-  subset(distinct_samples, 
                             select = c(Kids_First_Participant_ID)) %>%
           mutate(path = temp_path,
                 additional = 1) %>%
  write_csv(file.path(results_dir, "distinct_samples_list.csv"))
```


```{r echo=TRUE}
# Create directory to save CloneFinder results
CloneFinder_results <-
  file.path(results_dir, "CloneFinder_results")
if (!dir.exists(CloneFinder_results)) {
  dir.create(CloneFinder_results)
}
```

```{r}
sessionInfo()
```