---
title: "Selection analyses and cancer driver discovery using dNdScv"
author: 'Antonia Chroni <chronia@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---



The **dNdScv** R package is a suite of maximum-likelihood dN/dS methods designed to quantify selection in cancer and somatic evolution (Martincorena *et al.*, 2017). The package contains functions to quantify dN/dS ratios for missense, nonsense and essential splice mutations, at the level of individual genes, groups of genes or at whole-genome level. The *dNdScv* method was designed to detect cancer driver genes (*i.e.* genes under positive selection in cancer) on datasets ranging from a few samples to thousands of samples, in whole-exome/genome or targeted sequencing studies.

Unlike traditional implementations of dN/dS, the *dNdScv* package uses trinucleotide context-dependent substitution models to avoid common mutation biases affecting dN/dS (Greenman *et al.*, 2006, Martincorena *et al.*, 2017). The package includes two different dN/dS models. *dNdSloc*, like traditional dN/dS implementations, uses the number of synonymous mutations in a gene to infer the local mutation rate, without exploiting additional information from other genes. *dNdScv* offers a much more powerful alternative, combining local information (synonymous mutations in the gene) and global information (variation of the mutation rate across genes, exploiting epigenomic covariates) to estimate the background mutation rate. *dNdScv* should be preferred in most situations.

For more information, see [dndscv](https://github.com/im3sanger/dndscv).

##Driver discovery (positive selection) in cancer exomes/genomes


```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
  #library(devtools); install_github("im3sanger/dndscv")
  library("dndscv")
})
```

# Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory
# Use this as the root directory to ensure proper sourcing of functions 
# no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
analysis_dir <- file.path(root_dir, "analyses", "tumor-clone-inference")
input_dir <- file.path(analysis_dir, "input")
data_dir <- file.path(root_dir, "data")
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")
maf_files_dir <- file.path(root_dir, "analyses", "tmb-vaf-longitudinal", "results")
refdb_dir <- file.path(input_dir, "dndscv_data", "data")


# Input files
pbta_file <- file.path(files_dir, "pbta.tsv") # file from add-sample-distribution module
maf_file <- file.path(data_dir, "snv-consensus-plus-hotspots.maf.tsv.gz")
refdb_file <- file.path(refdb_dir, "RefCDS_human_GRCh38.p12_dNdScv.0.1.0.rda")

# File path to results directory
results_dir <-
  file.path(analysis_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

```


# Preprocess data aligned to GRCh38 genome reference

By default, dNdScv assumes that mutation data is mapped to the GRCh37/hg19 assembly of the human genome.
Users interested in trying dNdScv on a different set of transcripts, a different assembly or a different species can follow this [tutorial](https://htmlpreview.github.io/?http://github.com/im3sanger/dndscv/blob/master/vignettes/buildref.html).


```{r preprocess-data, message=FALSE}
# Read in histologies file and filter for the pbta cohort
pbta <- readr::read_tsv(pbta_file, guess_max = 100000, show_col_types = FALSE) %>%
  filter(!(experimental_strategy == "RNA-Seq")) %>% 
  dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID)

maf_df <- vroom::vroom(maf_file, delim = "\t", col_names = TRUE) 

# Read maf file
maf <- maf_df %>% 
  filter(Tumor_Sample_Barcode %in% pbta$Kids_First_Biospecimen_ID#,
         #Variant_Classification %in% maf_nonsynonymous
         ) %>%
  dplyr::rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) 

# How many genes
genes <- length(unique(maf$Hugo_Symbol))

mutations <- maf %>% 
  dplyr::select(Kids_First_Biospecimen_ID, Gene, Hugo_Symbol, 
         Transcript_ID, Chromosome, Start_Position, 
         End_Position, Strand,
         Reference_Allele, Tumor_Seq_Allele2) %>% 
  left_join(pbta) %>% 
  dplyr::select(-c(Kids_First_Biospecimen_ID)) %>% 
  #mutate(Chromosome = str_replace(Chromosome, "chr", "")) %>% 
  dplyr::rename("sampleID" = "Kids_First_Participant_ID",
                "gene.id" = "Gene",
                "gene.name" = "Hugo_Symbol",
                "cds.id" = "Transcript_ID",
                "chr" = "Chromosome",
                #"chr.coding.start" = "Start_Position",
                #"chr.coding.end" = "End_Position",
                "chr" = "Chromosome",
                "pos" = "Start_Position",
                "ref" = "Reference_Allele",
                "mut" = "Tumor_Seq_Allele2"
                #"ref" = "Reference_Allele",
                #"mut" = "Tumor_Seq_Allele2"
                ) %>% 
  dplyr::select(sampleID, chr, pos, ref, mut)

print(head(mutations))

```
# Run dndscv

With the example dataset provided, the function should take about one minute to run. In this example, the function issues a warning as it detects the same mutation in more than one sample, requesting the user to verify that the input table of mutations only contains independent mutation events. In this case, each sample corresponds to a different patient and so the warning can be safely ignored.

We have run dNdScv with default parameters. This includes removing ultra-hypermutator samples and subsampling mutations when encountering too many mutations per gene in the same sample. These were designed to protect against loss of sensitivity from ultra-hypermutators and from clustered artefacts in the input mutation table, but there are occasions when the user will prefer to relax these (see Example 2).

```{r run-dndscv, message=FALSE}
dndsout = dndscv(mutations, refdb = refdb_file, cv = NULL)
```


#####dndscv outputs: Table of significant genes

The output of the *dndscv* function is a list of objects. For an analysis of exome or genome data, the most relevant output will often be the result of neutrality tests at gene level. *P-values* for substitutions are obtained by Likelihood-Ratio Tests as described in (Martincorena *et al*, 2017) and q-values are obtained by Benjamini-Hodgberg's multiple testing correction. The table also includes information on the number of substitutions of each class observed in each gene, as well as maximum-likelihood estimates (MLEs) of the dN/dS ratios for each gene, for missense (*wmis*), nonsense (*wnon*), essential splice site mutations (*wspl*) and indels (*wind*). The global q-value integrating all mutation types are available in the *qglobal_cv* and *qallsubs_cv* columns for analyses with and without indels, respectively.

```{r}
sel_cv = dndsout$sel_cv %>%   
  write_tsv(file.path(results_dir, "dndsout_sel_cv.tsv"))

print(head(sel_cv), digits = 3)

signif_genes = sel_cv[sel_cv$qglobal_cv<0.1, c("gene_name","qglobal_cv")] # `<0.1` by default; ==0
rownames(signif_genes) = NULL

signif_genes_unique <- signif_genes %>%
  as.data.frame() %>% 
  #select(gene_name) %>% 
  #unique() %>% 
  write_tsv(file.path(results_dir, "signif_genes.tsv"))

print(signif_genes_unique)


signif_genes_unique <- signif_genes_unique %>% 
  mutate(gene_name = case_when(grepl("H3F3A", gene_name) ~ "H3-3A", 
                                     TRUE ~ gene_name)) %>% 
  write_tsv(file.path(results_dir, "signif_genes_modified.tsv"))

```

Note in the table that the dN/dS ratios of significant cancer genes are typically extremely high, often >10 or even >100. This contains information about the fraction of mutations observed in a gene that are genuine drivers under positive selection. For example, for a highly significant gene, a dN/dS of 10 indicates that there are 10 times more non-synonymous mutations in the gene than neutrally expected, suggesting that at least around 90% of these mutations are genuine drivers (Greenman *et al*, 2006; Martincorena *et al*, 2017).

Users can calculate confidence intervals for the dN/dS ratios per gene using the *geneci* function in the package. To use this function you need to run *dndscv* with the optional argument `outmats=T` and then use `ci = geneci(dndsout)`. For more details see `? geneci`.

#####dndscv outputs: Global dN/dS estimates

Another output that can be of interest is a table with the global MLEs for the dN/dS ratios across all genes. dN/dS ratios with associated confidence intervals are calculated for missense, nonsense and essential splice site substitutions separately, as well as for all non-synonymous substitutions (*wall*) and for all truncating substitutions together (*wtru*), which include nonsense and essential splice site mutations.

```{r}
print(dndsout$globaldnds)
```

Global dN/dS ratios in somatic evolution of cancer, and seemingly of healthy somatic tissues, appear to show a near-universal pattern of dN/dS~1, with exome-wide dN/dS ratios typically slightly higher than 1 (Martincorena *et al.*, 2017). Only occasionally, I have found datasets with global dN/dS<1, but upon closer examination, this has typically been caused by contamination of the catalogue of somatic mutations with germline SNPs. An exception are melanoma tumours, which show a bias towards slight underestimation of dN/dS due to the signature of ultraviolet-induced mutations extending beyond the trinucleotide model (Martincorena *et al.*, 2017). In my personal experience, datasets of somatic mutations with global dN/dS<<1 have always reflected a problem of SNP contamination or an inadequate substitution model, and so the evaluation of global dN/dS values can help identify problems in certain datasets.

#####Other useful outputs

The dndscv function also outputs other results that can be of interest, such as an annotated table of coding mutations (*annotmuts*), MLEs of mutation rate parameters (*mle_submodel*), lists of samples and mutations excluded from the analysis and a table with the observed and expected number of mutations per gene (*genemuts*), among others.

```{r}
head(dndsout$annotmuts)

df = dndsout$annotmuts
```

dNdScv relies on a negative binomial regression model across genes to refine the estimated background mutation rate for a gene. This assumes that the variation of the mutation rate across genes that remains unexplained by covariates or by the sequence composition of the gene can be modelled as a Gamma distribution. This model typically works well on clean cancer genomic datasets, but not all datasets may be suitable for this model. In particular, very low estimates of $\theta$ (the overdispersion parameter), particularly $\theta<1$, may reflect problems with the suitability of the dNdScv model for the dataset.

```{r}
print(dndsout$nbreg$theta)
```

##### dNdSloc: local neutrality test

An additional set of neutrality tests per gene are performed using a more traditional dN/dS model in which the local mutation rate for a gene is estimated exclusively from the synonymous mutations observed in the gene (*dNdSloc*) (Wong, Martincorena, *et al*., 2014). This test is typically only powered in very large datasets. For example, in the dataset used in this example, comprising of 196 simulated breast cancer exomes, this model only detects *ARID1A* as significantly mutated.

```{r}
signif_genes_localmodel = as.vector(dndsout$sel_loc$gene_name[dndsout$sel_loc$qall_loc<0.1])
print(signif_genes_localmodel)
```


```{r echo=TRUE}
sessionInfo()
```


