---
title: "Prepare input data for clonevol"
author: 'Antonia Chroni <chronia@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---

# Background

[ClonEvol](https://github.com/hdng/clonevol) is a package for clonal ordering and clonal evolution visualization. It uses the clustering of heterozygous variants identified using other tools as input to infer consensus clonal evolution trees and estimate the cancer cell fraction (also called clonal frequency) of the clones in individual samples. ClonEvol can deal with statistical uncertainty and error in sequencing and data analysis that may distort the cellular prevalence estimate of individual variants.


# Set up

```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
  library(reshape2)
})
```

# Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory
# Use this as the root directory to ensure proper sourcing of functions 
# no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
analysis_dir <- file.path(root_dir, "analyses", "tumor-clone-inference")
results_dir <- file.path(analysis_dir, "results")
input_dir <- file.path(results_dir, "fastclone-output")

# Input files
input_all_data_file <- file.path(results_dir, "input_all_data.tsv") 
signif_genes_file <- file.path(results_dir, "signif_genes.tsv") 

# File path to input directory
clonevol_input_dir <-
  file.path(analysis_dir, "results", "clonevol-input")
if (!dir.exists(clonevol_input_dir)) {
  dir.create(clonevol_input_dir)
}
```

# Load and process data

```{r load-process-inputs}
# It is important to order timepoints 
# so plots and numbers of clusters/clones will appear in the correct order 

# Define timepoints
timepoints = c("Diagnosis", "Progressive", "Recurrence", "Deceased", "Second Malignancy", "Unavailable")

input_all_data_df <- readr::read_tsv(input_all_data_file, guess_max = 100000, show_col_types = FALSE) %>%
  dplyr::rename("mutation.id.unique" = "mutation_id") %>% 
  #filter(Kids_First_Participant_ID == "PT_Z4BF2NSB") %>% 
  dplyr::mutate(tumor_descriptor = factor(tumor_descriptor),
                tumor_descriptor = fct_relevel(tumor_descriptor, timepoints)) %>% 
  arrange(tumor_descriptor) 

#############################################
# Read and process cluster file
# FastClone
data_dir <- dir(path = input_dir,  pattern = "scores.csv", full.names = TRUE, recursive = TRUE)

# Create list 
data_list <- list()

for (i in 1:length(data_dir) ) { 
  # Create sample_name
  sample_name <-  unique(as.character(gsub("scores.csv", "", str_split_fixed(data_dir[i], "/", 13)[,11])))
  sample_name <- sort(sample_name, decreasing = FALSE)
  print(sample_name)
  
  for (x in seq_along(sample_name) ) { 
    
    data_list[[i]] <- read.csv(data_dir[i]) 
  }
}
#############################################

# Bind all df from list 
data_list_bind <- dplyr::bind_rows(data_list) 

colnames(data_list_bind)[1] <- "mutation.id.unique" 
names(data_list_bind)


# Count number of columns
count_col <- ncol(data_list_bind)
count_col

# Duplicate orginalf df
data_list_bind <- data_list_bind %>% 
  
  # Cluster number
  dplyr::rename("1" = "X0",
                "2" = "X1"#,
                #"3" = "X2"
                ) %>% 
  mutate_at(c(2:count_col), as.numeric) 

# Melt df
cluster_df <- data_list_bind  %>%
  melt() %>% 
  dplyr::rename("cluster" = "variable",
                "score" = "value"
                )

#############################################

clonevol_df <- input_all_data_df %>%
  
  # order columns by placing numeric columns at the end of the df
  select(Kids_First_Participant_ID, cg_id_kids, tumor_descriptor, 
         gene_protein, Hugo_Symbol, mutation.id.unique, sample_id, 
         ref_counts, alt_counts, normal_cn, 
         major_cn, minor_cn, tumour_content, 
         mutation_count, VAF, CCF) %>% 
  
  # change names to match input requirements
  dplyr::rename("ref.count" = "ref_counts", 
                "var.count" = "alt_counts",
                "vaf" = "VAF",
                "ccf" = "CCF") %>% 
  unique() %>% 
  mutate_at(c(7:15), as.numeric) %>% 

  # pyclone-vi starts numbering of clones from `0`
  # this will cause an error for clonevol when trying to subset
  # in the `plot.variant.clusters`
  # In addition, `infer.clonal.models` function doesn't accept 0 for clustering
  dplyr::mutate(#cluster = case_when(grepl("Deceased", tumor_descriptor) ~ "3",
                 #                   grepl("Recurrence", tumor_descriptor) ~ "2",
                 #                   TRUE ~ "1"),
                # cluster = case_when(grepl("0", cluster) ~ "1"),
                ccf = (ccf*100),
                vaf = (vaf*100)#,
                #unique_id = paste(sample_id, ref.count, sep = ":")
                ) %>% 
  
  # Add cluster information
  left_join(cluster_df) %>% 
  filter(!is.na(score)) %>% 
  unique() %>% 
  dplyr::mutate(mutation_id = paste(mutation.id.unique, major_cn, minor_cn, sep = ":"))


print(table(clonevol_df$cg_id_kids))

###---------------------------------------------------------------------------------------

# Read and process list with driver genes
signif_genes_df <- readr::read_tsv(signif_genes_file, guess_max = 100000, show_col_types = FALSE) %>% 
  select(gene_name) %>% 
  dplyr::rename("gene.all" = "gene_name"
                ) %>% 
  dplyr::mutate(is.driver.all = "TRUE")


# to make a loop for each `cg_id_kids`
cg_id_kids_names <- unique(as.character(clonevol_df$cg_id_kids))
cg_id_kids_names <- sort(cg_id_kids_names, decreasing = FALSE)
print(cg_id_kids_names)

# Loop through cg_sum
for (x in seq_along(cg_id_kids_names)){
  print(x)

  clonevol_df_sub <- clonevol_df %>%
    filter(cg_id_kids == cg_id_kids_names[x])

  # make long table
  df <- clonevol_df_sub %>%
    # ref.count
    mutate(id = paste(tumor_descriptor, "ref.count", sep = ".")) %>% #sample_id

    # using pivot_wider for multiple columns was introducing a bug so I split it up instead
    # pivot_wider(names_from = "sample_id", values_from = c("ref.count", "var.count", "vaf", "ccf")) %>% #values_fn = list
    pivot_wider(names_from = "id", values_from = c("ref.count"), values_fill = 0) %>% #values_fn = list
  
    # var.count
    left_join(clonevol_df, relationship = "many-to-many") %>% 
    select(-c(ref.count)) %>% 
    mutate(id = paste(tumor_descriptor, "var.count", sep = ".")) %>% 
    pivot_wider(names_from = "id", values_from = c("var.count"), values_fill = 0) %>% #values_fn = list

    # vaf
    left_join(clonevol_df, relationship = "many-to-many") %>% 
    select(-c(ref.count, var.count)) %>% 
    mutate(id = paste(tumor_descriptor, "vaf", sep = ".")) %>% 
    pivot_wider(names_from = "id", values_from = c("vaf"), values_fill = 0) %>% #values_fn = list
  
    # ccf
    left_join(clonevol_df, relationship = "many-to-many") %>% 
    select(-c(ref.count, var.count, vaf)) %>% 
    mutate(id = paste(tumor_descriptor, "ccf", sep = ".")) %>% 
    unique() %>% 
    pivot_wider(names_from = "id", values_from = c("ccf"), values_fill = 0) %>% #values_fn = list
  
    as.data.frame() %>% 
    select(-c(Kids_First_Participant_ID, tumor_descriptor)) 
  
  # Count number of columns
  count_col <- ncol(df)

  # Make input for clonevol
  clonevol_input_df <- df %>% 
    #mutate_at(c(3:count_col), as.numeric) %>%
    dplyr::rename("gene.all" = "Hugo_Symbol") %>% 
    #mutate(id = paste(sample_id, Hugo_Symbol, sep = "_")) %>% 
    #select(-c(Hugo_Symbol)) %>% 
    #left_join(goi_df) %>% 
    left_join(signif_genes_df) %>%
    dplyr::mutate(is.driver.all = replace_na(is.driver.all, "FALSE")) %>% 
    #dplyr::rename("gene.all" = "gene_protein", 
    #              # we will use sym info as driver info for now
    #             "is.driver.all" = "sym") %>%
    select(-c(sample_id)) %>%
    relocate(c(gene.all, is.driver.all), .after = Deceased.ccf) %>% 
    unique %>%
    filter(!is.na(is.driver.all)) %>% 
    replace(is.na(.), "-")
    #filter(!is.na(is.driver.all)) %>% 
    #drop_na() 
  
  
  # Convert labels to display on the plot only the ones in the goi list
  clonevol_input_df$gene.all <- ifelse(clonevol_input_df$is.driver.all == TRUE, clonevol_input_df$gene.all, "-")
  
  # Create unique gene labels for visualizations
  # to avoid multiplets per cluster/clone
  unique_genes_clonevol_input_df =  clonevol_input_df %>% 
    #mutate(mutation.id.unique = paste(mutation.id.unique, cluster, sep = ".")) %>% 
    distinct(mutation.id.unique, cluster, .keep_all = TRUE) %>% 
    dplyr::rename("gene" = "gene.all",
                  "is.driver" = "is.driver.all"
                  ) %>% 
    right_join(clonevol_input_df) %>% 
    replace(is.na(.), "-") %>% 
    mutate(is.driver = case_when(grepl("-", is.driver) ~ "FALSE",
                                 TRUE ~ is.driver)) %>% 
    distinct(gene, cluster, .keep_all = TRUE) %>% 
    right_join(clonevol_input_df) %>% 
    replace(is.na(.), "-") %>% 
    mutate(is.driver = case_when(grepl("-", is.driver) ~ "FALSE",
                                 TRUE ~ is.driver))
  
  
  # Name file to save
  fname <- paste0(clonevol_input_dir, "/", cg_id_kids_names[x], "-clonevol_input.tsv")
  print(fname)
  
  # Save the df
  unique_genes_clonevol_input_df <- unique_genes_clonevol_input_df %>%
    #mutate(gene = paste(gene, Diagnosis.ref.count, sep = "_")) %>% 
    write_tsv(file.path(fname))
  
}
```


```{r}
sessionInfo()
```

