---
title: "Estimation of VAF of tumors across multiple timepoints for the autopsy samples in the PBTA Cohort"
author: 'Antonia Chroni <chronia@chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---


#### Tumor evolution project 

# Background
We are investigating the change of VAFs across multiple timepoints for the autopsy samples in the PBTA cohort.

### Data used 
In this notebook, we are looking into the v12 histologies file (last updates on May 1st, 2023) and subset to the PBTA cohort.

# Set up

```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
})
```

## Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory
# Use this as the root directory to ensure proper sourcing of functions 
# no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
scratch_dir <- file.path(root_dir, "scratch")
data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "create-corplots")
input_dir <- file.path(analysis_dir, "input")
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")

# Input files
v12_file <- file.path(data_dir, "histologies.tsv")
maf_file <- file.path(data_dir, "snv-consensus-plus-hotspots.maf.tsv.gz")
tmb_file <- file.path(input_dir, "snv-mutation-tmb-coding.tsv")
matched_time_points_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") # file from add-sample-distribution module
```

## Read in data and process

```{r load-process-inputs-please-wait}
# Read in histologies file and filter for the pbta cohort
pbta <- readr::read_tsv(v12_file, guess_max = 100000, show_col_types = FALSE) %>%
  filter(cohort == "PBTA",
         !is.na(pathology_diagnosis),
         !composition %in% c("Derived Cell Line", "PDX"),
         !experimental_strategy %in% c("Methylation", "Targeted Sequencing")) %>%
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, experimental_strategy, tumor_descriptor, age_at_diagnosis_days, age_last_update_days, cancer_predispositions, broad_histology, short_histology, cancer_group) 

# Make a vector of coding gene mutations
maf_nonsynonymous <- c(
  "Missense_Mutation",
  "Frame_Shift_Del",
  "In_Frame_Ins",
  "Frame_Shift_Ins",
  "Splice_Site",
  "Nonsense_Mutation",
  "In_Frame_Del",
  "Nonstop_Mutation",
  "Translation_Start_Site"
)


# Read in maf file, calculate VAF, and retain only coding gene mutations
maf <- data.table::fread(maf_file, data.table = FALSE) %>% 
  filter(Tumor_Sample_Barcode %in% pbta$Kids_First_Biospecimen_ID,
         Variant_Classification %in% maf_nonsynonymous) %>%
  dplyr::rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>% # change name of the biospecimen to match the one from the histologies files
  dplyr::mutate(VAF = t_alt_count / (t_ref_count + t_alt_count), # calculate VAF 
                gene_protein = paste(Hugo_Symbol, HGVSp_Short, sep = "_")) %>%  # concatenate gene and protein information
  select(Kids_First_Biospecimen_ID, Hugo_Symbol, Chromosome, Start_Position, Variant_Classification, VAF, HGVSc, HGVSp, HGVSp_Short, gene_protein, match_id) 


# Read in tmb file
tmb <- readr::read_tsv(tmb_file, guess_max = 100000, show_col_types = FALSE)

# Read in matched_time_points_file and get the list of patients with matched time points
patient_list <- readr::read_tsv(matched_time_points_file, guess_max = 100000, show_col_types = FALSE) %>%
  subset(select = c("Kids_First_Participant_ID", "descriptors"))

# How many patients have paired time samples?
paired_time_samples_number <- length(unique(patient_list$Kids_First_Participant_ID))

# Add metadata from pbta to patient list and keep only autopsy samples
patient_list_metadata <- patient_list %>% 
  left_join(pbta, by = "Kids_First_Participant_ID") %>%
  filter(!(experimental_strategy == "RNA-Seq")) %>%
  mutate(tumor_descriptor = case_when(grepl("Primary Tumor", tumor_descriptor) ~ "Diagnosis", 
                                      grepl("Initial CNS Tumor", tumor_descriptor) ~ "Diagnosis", 
                                      grepl("Progressive Disease Post-Mortem", tumor_descriptor) ~ "Deceased", 
                                      TRUE ~ tumor_descriptor)) %>%
  mutate(Deceased = case_when(grepl("Deceased", descriptors) ~ "Deceased")) %>%
  filter(Deceased == "Deceased") 

# How many patients have paired time samples with autopsy time point present?
paired_time_samples_number_autopsy <- length(unique(patient_list_metadata$Kids_First_Participant_ID))
```

## Number of patient cases with paired time and autopsy samples

`r paired_time_samples_number` of patient cases have paired time samples in genomic assays and `r paired_time_samples_number_autopsy` have autopsy sample.


# Create maf file for autopsy samples

```{r maf_autopsy}
# There are patient cases with more than one time points in addition to Deceased, e.g. Diagnosis and Progressive.
# We will create plots per each time point as a variable while considering Deceased as constant.
# First we will create "timepoints_all_number" based on "timepoints_all" to account for time points = 1, 2, 3.
maf_autopsy <- patient_list_metadata %>% 
  left_join(maf, by = "Kids_First_Biospecimen_ID") %>%
  filter(!is.na(Kids_First_Participant_ID),
         !is.na(VAF),
         !VAF == 0, 
         !gene_protein == "Unknown") %>%
  mutate(timepoints_all = case_when(descriptors == "Deceased, Diagnosis, Diagnosis, Progressive, Progressive" ~ "dx_pro", 
                                     descriptors == "Deceased, Diagnosis, Recurrence" ~ "dx_rec", descriptors %in% c("Deceased, Progressive, Recurrence", "Deceased, Progressive, Progressive, Recurrence", "Deceased, Progressive, Recurrence, Recurrence") ~ "pro_rec",
                                     descriptors == "Deceased, Diagnosis, Progressive, Recurrence" ~ "dx_pro_rec",
                                     descriptors %in% c("Deceased, Diagnosis", "Deceased, Diagnosis, Diagnosis", "Deceased, Diagnosis, Diagnosis, Diagnosis, Diagnosis") ~ "dx",
                                     descriptors %in% c("Deceased, Progressive", "Deceased, Progressive, Progressive") ~ "pro",
                                     descriptors == "Deceased, Recurrence" ~ "rec"),
         timepoints_all_number = case_when(timepoints_all %in% c("dx_pro", "dx_rec", "pro_rec") ~ "2",
                                            timepoints_all == "dx_pro_rec" ~ "3",
                                            timepoints_all %in% c("dx", "pro", "rec") ~ "1")) %>%
  select(Kids_First_Participant_ID, tumor_descriptor, timepoints_all, timepoints_all_number, Hugo_Symbol, gene_protein, VAF) %>%
  mutate(match_id = paste(Kids_First_Participant_ID, tumor_descriptor, gene_protein, VAF, sep = "_")) %>%
  unique ()

# Next, we will create df for cases for timepoint = 1, 2, 3.
# That we will allow us to will create a column to identify time points to use for the plots functions (timepoints).

# Subset the cases with one time point 
# Cases with 1 TIME POINT
timepoint_1 <- maf_autopsy %>%
  mutate(col_rm = ifelse(tumor_descriptor == "Deceased" & timepoints_all_number %in% c("2", "3"), "remove", "keep")) %>%
  filter(col_rm == "keep") %>%
  subset(select = -c(col_rm)) 

# Subset to the cases with only the Deceased samples to be used for both timepoint_2 and timepoint_3
maf_autopsy_subset <- maf_autopsy %>%
  filter(tumor_descriptor == "Deceased")

# Cases with 2 TIME POINTS
# Subset the cases with two more time points and duplicate these rows 
timepoint_2 <- maf_autopsy_subset %>%
  filter(timepoints_all_number == "2")
timepoint_2 <- timepoint_2[rep(seq_len(nrow(timepoint_2)), each = 2), ]  

# Rename duplicate rows to distinguish between duplicates
timepoint_2 <- timepoint_2 %>% 
  group_by(match_id) %>%
  dplyr::mutate(tumor_descriptor = paste("Deceased", row_number(), sep = "_")) 

# Cases with 3 TIME POINTS
# Subset the cases with three more time points and duplicate these rows 
timepoint_3 <- maf_autopsy_subset %>%
  filter(timepoints_all_number == "3")
timepoint_3 <- timepoint_3[rep(seq_len(nrow(timepoint_3)), each = 3), ]  

# Rename duplicate rows to distinguish between duplicates
timepoint_3 <- timepoint_3 %>% 
  group_by(match_id) %>% 
  dplyr::mutate(tumor_descriptor = paste("Deceased", row_number(), sep = "_")
  )

# Bind each df with time points 1, 2, 3
maf_autopsy.bind <- rbind(timepoint_1, timepoint_2, timepoint_3) %>%
  mutate(timepoints = case_when(timepoints_all == "dx" & tumor_descriptor %in% c("Diagnosis", "Deceased") ~ "Diagnosis",
                                          timepoints_all == "pro" & tumor_descriptor %in% c("Progressive", "Deceased") ~ "Progressive",
                                          timepoints_all == "rec" & tumor_descriptor %in% c("Recurrence", "Deceased") ~ "Recurrence",
                                          timepoints_all == "dx_pro" & tumor_descriptor %in% c("Diagnosis", "Deceased_1") ~ "Diagnosis",
                                          timepoints_all == "dx_pro" & tumor_descriptor %in% c("Progressive", "Deceased_2") ~ "Progressive",
                                          timepoints_all == "dx_rec" & tumor_descriptor %in% c("Diagnosis", "Deceased_1") ~ "Diagnosis",
                                          timepoints_all == "dx_rec" & tumor_descriptor %in% c("Recurrence", "Deceased_2") ~ "Recurrence",
                                          timepoints_all == "pro_rec" & tumor_descriptor %in% c("Progressive", "Deceased_1") ~ "Progressive",
                                          timepoints_all == "pro_rec" & tumor_descriptor %in% c("Recurrence", "Deceased_2") ~ "Recurrence",
                                          timepoints_all == "dx_pro_rec" & tumor_descriptor %in% c("Diagnosis", "Deceased_1") ~ "Diagnosis",
                                          timepoints_all == "dx_pro_rec" & tumor_descriptor %in% c("Progressive", "Deceased_2") ~ "Progressive",
                                          timepoints_all == "dx_pro_rec" & tumor_descriptor %in% c("Recurrence", "Deceased_3") ~ "Recurrence"),
         tumor_descriptor = case_when(tumor_descriptor %in% c("Deceased_1", "Deceased_2", "Deceased_3") ~ "Deceased",
                                      TRUE ~ tumor_descriptor)) %>%
  write_tsv(file.path(scratch_dir, "maf_autopsy.tsv"))
```

```{r echo=TRUE}
sessionInfo()
```
