---
title: "Longitudinal analysis for the PBTA Cohort (pbta-tumor-evolution project)"
author: 'Antonia Chroni <chronia@chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---


#### Tumor evolution project 

# Background
This is an exploratory analysis of longitudinal data in the PBTA cohort.
We are investigating the number of histologies (cancer_type column): (1) per assay (experimental_strategy) and (2) pairs of genomic and transcriptomic assays.
We are also looking at the number of patient samples available per histology with paired genomic and transcriptomic assays for (1) each time point (disease_stage) and (2) when Diagnosis is present and paired with another point in time.

### Data used 
In this notebook, we are looking into the v12 histologies file (last updates on May 1st, 2023) and subset to the PBTA cohort.
We use tumor_descriptor and focus on time points recorded.

### Note
The colors in the plots do not match the ones of the PBTA cohort yet. ---TO DO---


### Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/01_sample_distribution_analysis/01-tumor-descriptor-and-assay-count-plot.Rmd', clean = TRUE)"
```

# Set up

```{r load-library}

suppressPackageStartupMessages({
  library(tidyverse)
  library(ggplot2)
  library(flextable)
})


```

## Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no
# matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
setwd(root_dir)


# File path to input directory
analysis_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis")
input_dir <- file.path(analysis_dir, "input")
  file.path(root_dir, "input")


# Inputs
input_metadata <- file.path(input_dir, "histologies.tsv")
input_cols <- file.path(input_dir, "PBTA-germline-histology-groups-plot-mapping.tsv")


# File path to results directory
results_dir <-
  file.path(analysis_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}


# Outputs
output_cohort_pbta <- file.path(results_dir, "cohort_pbta.tsv")
output_paired_assays <- file.path(results_dir, "paired_assays.tsv")
output_paired_Diagnosis <- file.path(results_dir, "paired_Diagnosis.tsv")

output_genomic_df <- file.path(results_dir, "genomic_df.tsv")

output_LGG <- file.path(results_dir, "LGG.tsv")
output_Medulloblastoma <- file.path(results_dir, "Medulloblastoma.tsv")
output_HGG <- file.path(results_dir, "HGG.tsv")
output_DML <- file.path(results_dir, "DML.tsv")
output_Ependymoma <- file.path(results_dir, "Ependymoma.tsv")

```


## Load data
```{r load-inputs-please-wait}
# Inputs
## Read in metadata
v12 <- readr::read_tsv(input_metadata, guess_max = 100000, show_col_types = FALSE)

```

# Process Data

All tumor samples have matched normal samples. 
We need to filter normal samples out for further analysis. 
We are using "pathology_diagnosis" instead of "sample_type". This way we discard normal samples (as shown in "sample_type") in addition to undiagnosed tumor samples.
Also, we will discard Derived Cell Line, PDX from "composition".


```{r cohort PBTA, echo=TRUE}
# Note
# The following samples (sample_id) are marked as Not Reported tumor_descriptor: 7316UP-3705, C4106355-Tumor, and C4113612-Tumor
# Jo Lynne found the following status for 2/3 in the data warehouse:
# C4106355-Tumor = Recurrence
# C4113612-Tumor = Initial CNS Tumor
# 7316UP-3705 = Not Reported
# manually annotate them here and these will be updated in the v13 - then discard that part of the code

# I also added the cancer_group into the match_id to use this later in the script
# Also added the experimental_strategy to create unique events tagged to each assay 
# so we won't miss samples with paired assays across multiple time points
# if you remove this, print(length(unique(pbta$match_id))) will result to 2258, instead of 4293 thatis now
# hope this makes and it is correct rational...

pbta <- v12 %>%
  filter(cohort == "PBTA",
         !is.na(pathology_diagnosis),
         !composition %in% c("Derived Cell Line", "PDX"),
         !experimental_strategy %in% c("Methylation", "Targeted Sequencing")) %>%
           # !is.na(cancer_group) # Discard NAs in cancer type
  mutate(match_id = paste(cancer_group, sample_id, composition, experimental_strategy, tumor_descriptor, sep = "_")) %>% 
  mutate(tumor_descriptor_final = case_when(grepl("C4106355-Tumor", sample_id) ~ "Recurrence", # to be updated in v13 as "Recurrence"
                                            grepl("C4113612-Tumor", sample_id) ~ "Diagnosis", # to be updated in v13 as "Initial CNS Tumor"
                                            grepl("Primary Tumor", tumor_descriptor) ~ "Diagnosis",
                                            grepl("Initial CNS Tumor", tumor_descriptor) ~ "Diagnosis",
                                            grepl('Progressive Disease Post-Mortem', tumor_descriptor) ~ 'Deceased', 
                                            TRUE ~ tumor_descriptor))

print(pbta %>% count(tumor_descriptor_final))

# Save pbta after filter 
write_tsv(pbta, path = output_cohort_pbta)
```


## Filter based on both genomic and transcriptomic assays
Let's select all patient samples containing both genomic and transcriptomic assays.

```{r assays, echo=TRUE}

print(length(unique(pbta$Kids_First_Participant_ID)))
print(length(unique(pbta$match_id)))

assays <- pbta %>%
  select(Kids_First_Participant_ID, tumor_descriptor_final, match_id, experimental_strategy, cancer_group) %>%
  unique() %>%
  arrange(Kids_First_Participant_ID)


# we perform join here using the Kids_First_Participant_ID and match_id
# THIS PART I AM NOT SURE OF 
paired_assays <- assays %>%
  group_by(Kids_First_Participant_ID, match_id) %>%
  summarise(experimental_strategy = str_c(experimental_strategy, collapse = ";")) %>%
  mutate(experimental_strategy_sum = case_when(grepl("RNA-Seq", experimental_strategy) & grepl("WGS|WXS", experimental_strategy) ~ "DNA,RNA",
                                                grepl("WXS|WGS", experimental_strategy) ~ "DNA", 
                                                grepl("RNA-Seq", experimental_strategy) ~ "RNA"
                                                )) 
  
# Save paired_assays 
write_tsv(paired_assays, path = output_paired_assays)
```


# Summary

In the following section, we will provide summary table across different disease stages, histologies, and patients samples.

## Number of each assay

First, we’ll examine how many of each type of assay we have (tumors only). This information is stored in the experimental_strategy column.

```{r}
paired_assays %>%
  group_by(experimental_strategy) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

paired_assays %>%
  group_by(experimental_strategy_sum) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```



# Diagnosis vs. any other point in time (recurrence, progressive etc)

We’ll look at the breakdown of the tumor_descriptor column, separating the genomic assays from transcriptomic assays.

```{r}
tumor_descriptor_df <- paired_assays %>%
    select(Kids_First_Participant_ID, match_id, experimental_strategy, experimental_strategy_sum) %>%
    mutate(cancer_group = match_id) %>%
    mutate(cancer_group =gsub("_..*", "", cancer_group)) %>% 
    mutate(tumor_descriptor = match_id) %>% 
    mutate(tumor_descriptor =gsub("..*_", "", tumor_descriptor)) %>% 
    mutate(tumor_descriptor = case_when(grepl("Primary Tumor", tumor_descriptor) ~ "Diagnosis",
                                 grepl("Initial CNS Tumor", tumor_descriptor) ~ "Diagnosis",
                                 grepl('Progressive Disease Post-Mortem', tumor_descriptor) ~ 'Deceased', 
                                 TRUE ~ tumor_descriptor)) %>%
  arrange(Kids_First_Participant_ID)


tumor_descriptor_df %>%
  group_by(tumor_descriptor, experimental_strategy_sum, ) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

```

The following table was supposed to provide matched different time points and summarize the information under the "tumor_descriptor_sum" column.
But it doesn't work for now.

Also, notice how for the first patient with PT_00G007DM, there are two different cancer_group assigned, which creates double events per assay.
Why does this happen? Maybe to exclude cancer_group in match_id?


```{r}

tumor_descriptor_df_Diagnosis <- tumor_descriptor_df  %>%
  mutate(tumor_descriptor_sum=case_when(
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="Progressive" ~ "Diagnosis-Progressive",
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="Recurrence" ~ "Diagnosis-Recurrence",
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="Deceased" ~ "Diagnosis-Deceased",
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="Second Malignancy" ~ "Diagnosis-Second Malignancy",
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="Not Reported" ~ "Diagnosis-Not Reported",
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="NA" ~ "Diagnosis-NA",
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="Residual" ~ "Diagnosis-Residual",
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="Metastatic" ~ "Diagnosis-Metastatic",
    tumor_descriptor=="Diagnosis" & tumor_descriptor=="Unvailable" ~ "Diagnosis-Unvailable",
    TRUE ~ "Unmatched"
  ))


# Save tumor_descriptor_df_Diagnosis
write_tsv(tumor_descriptor_df_Diagnosis, path = output_paired_Diagnosis)


tumor_descriptor_df_Diagnosis %>%
  regulartable() %>%
  fontsize(size = 12, part = "all")
``` 

!!!!!!!!!
ATTENTION
This is not working properly. They are matched time points within patients with paired assays.
For example, patient with Kids_First_Participant_ID = PT_00G007DM.
!!!!!!!!!

## Genomic assays
Setting aside the RNA-Seq for the moment and only looking at the patient samples that contain WXS and WGS assays. We’re collapsing the different values in tumor_descriptor to form a single descriptor when there are multiple types of tumors from the same individual.


```{r}
genomic_df <- tumor_descriptor_df %>%
  filter(!(experimental_strategy_sum == "RNA")) %>% 
  group_by(Kids_First_Participant_ID) %>%
  summarize(descriptors = paste(sort(tumor_descriptor),  
                                collapse = ", "),
            experimental_strategy = paste(sort(experimental_strategy),
                                          collapse = ", "),
            cancer_group = paste(sort(cancer_group),
                                          collapse = ", "))

# Save tumor_descriptor_df_Diagnosis
write_tsv(genomic_df, path = output_genomic_df)
```


### All time points with genomic assays
Count the examples where all time points have both kinds of assays.

For time points where both WGS and WXS exist, it will show as a pair of the same time point.
For example, Diagnosis-Diagnosis.

```{r}
# no NAs = both kinds of assays are present
genomic_df %>%
  group_by(descriptors) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")
```



```{r}
genomic_df %>%
  group_by(Kids_First_Participant_ID, descriptors, experimental_strategy) %>%
  tally() %>%
  arrange(desc(n)) %>%
  regulartable() %>%
  fontsize(size = 12, part = "all")


```


## Transcriptomic assays
Looking only at RNA-seq samples and performing the same collapsing of the tumor_descriptor column.

```{r}
transcriptomic_df <- tumor_descriptor_df %>%
  filter(experimental_strategy == "RNA-Seq") %>%
  group_by(Kids_First_Participant_ID) %>%
  summarize(descriptors = paste(sort(unique(tumor_descriptor)),  
                                collapse = ", "),
            experimental_strategy = unique(experimental_strategy),
            cancer_group = paste(sort(unique(cancer_group)),
                                          collapse = ", "))
```


```{r}
transcriptomic_df %>%
  group_by(descriptors) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")
```



```{r}
transcriptomic_df  %>%
  group_by(Kids_First_Participant_ID, descriptors, experimental_strategy) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")
```


## Identify duplicate events - paired assays
We will identify duplicate events.
We may want to discard them from further analysis.
This identifies paired assays as well.

```{r paired_assays_mult_events, echo=TRUE}

# identify multiple events per patient sample
paired_assays_mult_events <- paired_assays %>%
  group_by(Kids_First_Participant_ID) %>%
  summarise(match_id = str_c(match_id, collapse = "~"),
            experimental_strategy = str_c(experimental_strategy, collapse = "~"),
            experimental_strategy_sum = str_c(experimental_strategy_sum, collapse = "~")) %>%
  mutate(has_mult_events = case_when(grepl("~", match_id) ~ "yes", 
                               TRUE ~ "no"))

table(paired_assays_mult_events$has_mult_events)

# Save paired_assays_mult_events
# write_tsv(paired_assays_mult_events, path = output_paired_assays_mult_events)

# filter out any multiple events based on has_mult_events column
# paired_assays_mult_events <- paired_assays_mult_events %>%
#  filter(has_mult_events == "no")

# table(paired_assays_mult_events$has_mult_events)
```


## Paired genomic, transcriptomic assays
How many participants have paired genomic and transcriptomic samples for the same tumor_descriptor values?

```{r}
# if we perform a full join here using the Kids_First_Participant_ID and descriptors, we 
# will get NAs in columns where pairs don't exist and can use this information 
# to count
paired_df <-
  full_join(genomic_df, transcriptomic_df,
                   by = c("cancer_group", "Kids_First_Participant_ID", "descriptors", "experimental_strategy"))

```

### All time points
Count the examples where all time points have both kinds of assays.

```{r}
# no NAs = both kinds of assays are present
paired_df %>%
  filter(complete.cases(.)) %>%
  group_by(descriptors) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part="all")
```



```{r}
paired_df %>%
  group_by(Kids_First_Participant_ID, descriptors, experimental_strategy) %>%
  tally() %>%
  arrange(desc(n)) %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")
```



### Genomic assays and cancer_group
Let's investigate the number of tumor samples with genomic assays per cancer_group.

```{r}
list <- genomic_df %>% 
      count(cancer_group) %>% 
      arrange(desc(n)) 
print (list, n)

```
We will look into the first five cancer groups with the highest number of tumor samples.
These are: Low-grade glioma, Medulloblastoma, High-grade glioma, Diffuse midline glioma, and Ependymoma.
BUT there are cases that the second time point has a different cancer_group. 
This creates combinations of the same subtype, but they won't show up in the following.
I need to update the code to reflect this. 
For more details, see "genomic_df.tsv".

#### Low-grade glioma
Per patient sample

```{r}
cohort <- genomic_df %>% 
  filter(cancer_group == "Low-grade glioma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

# Save table
LGG <- cohort
write_tsv(LGG, path = output_LGG)

```

#### Medulloblastoma
Per patient sample

```{r}
cohort <- tumor_descriptor_df_Diagnosis %>% 
  filter(cancer_group == "Medulloblastoma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

# Save table
Medulloblastoma <- cohort
write_tsv(LGG, path = output_Medulloblastoma)

```


#### High-grade glioma
Per patient sample

```{r}
cohort <- tumor_descriptor_df_Diagnosis %>% 
  filter(cancer_group == "High-grade glioma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

# Save table
HGG <- cohort
write_tsv(HGG, path = output_HGG)


```

#### Diffuse midline glioma
Per patient sample

```{r}
cohort <- tumor_descriptor_df_Diagnosis %>% 
  filter(cancer_group == "Diffuse midline glioma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

# Save table
DML <- cohort
write_tsv(DML, path = output_DML)

```

#### Ependymoma
Per patient sample

```{r}
cohort <- tumor_descriptor_df_Diagnosis %>% 
  filter(cancer_group == "Ependymoma")


cohort %>% 
  regulartable() %>%
  fontsize(size = 12, part = "all")

# Save table
Ependymoma <- cohort
write_tsv(Ependymoma, path = output_Ependymoma)

```

```{r}
sessionInfo()
```
  