---
title: "GSVA Score Modeling"
author: "Stephanie J. Spielman for ALSF CCDL"
date: '2020'
output:
  html_document:
    df_print: paged
    toc: yes
  html_notebook:
    df_print: paged
    toc: yes
    toc_float: yes
params:
  plot_ci: yes
  is_ci: FALSE
---

### Purpose

The purpose of this analysis is to assess significant differences in GSVA scores for each hallmark pathways. Using ANOVA and subsequent Tukey tests, we ask:

+ For each pathway, are GSVA scores significantly different across `cancer_group`? If so, which histologies are significantly different?

+ For each pathway, are GSVA scores significantly different across `tumor_descriptor`? If so, which types are significantly different?

We perform this using both GSVA scores calculated from RNA-seq libraries. Code is also flexible enough to test a different variable besides `cancer_group`, etc.

### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('01-model-gsea.Rmd', clean = TRUE)" 
```
_This assumes you are in the top directory of the repository._

### Setup

Load libraries and define certain constants:

```{r, lib-load, warning=FALSE, message=FALSE}
library(tidyverse)
library(broom)
library(ggpubr)
#library(Seurat)
#library(scater)
# How to install EnhancedVolcano
# https://github.com/kevinblighe/EnhancedVolcano
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")


# How to install enrichplot
#BiocManager::install("EnhancedVolcano")
#library(EnhancedVolcano)

#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("enrichplot")
#library(enrichplot)

`%>%` <- dplyr::`%>%`

# This script contains functions used to modeling GSVA scores
source(file.path("util", "hallmark_models.R"))
source(paste0(root_dir, "/figures/scripts/theme.R"))

# Significance testing universal threshold
SIGNIFICANCE_THRESHOLD <- 0.01

# Assigning params$is_ci to running_in_ci avoids a locked binding error
running_in_ci <- params$is_ci

# Are we testing? In case of a non 0/1 number, we recast as logical, and then ensure logical.
if (running_in_ci %in% c(0,1)) running_in_ci <- as.logical(running_in_ci)
if (!(is.logical(running_in_ci)))
{
  stop("\n\nERROR: The parameter `is_ci` should be FALSE/TRUE (or 0/1).")
}


# File path to plot directory
plots_dir <-
  file.path(analysis_dir, "plots")
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```


Next, define directories and load data files:
```{r, data-load, message=FALSE, warning=FALSE}
### Define directories
# Establish base dir
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir    <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "gene-set-enrichment-analysis")

files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")
input_dir <- file.path(analysis_dir, "input")

######### Define input files
## Metadata file (histologies/clinical data)
metadata_file <- file.path(files_dir, "pbta.tsv") # file from add-sample-distribution module
matched_genomic_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") # file from add-sample-distribution module

## GSEA scores
scores_file <- file.path(input_dir, "gsva_scores.tsv")

# File path to results directory
results_dir <-
  file.path(analysis_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

######## Load input files
pbta <- readr::read_tsv(metadata_file, guess_max = 100000) %>% 
  filter(experimental_strategy == "RNA-Seq") %>% 
  filter(!is.na(RNA_library))


# Read in matched_genomic_file and get the list of patients 
# with matched time points for the genomic assays
patient_list <- readr::read_tsv(matched_genomic_file, guess_max = 100000, show_col_types = FALSE) %>%
  subset(select = c("Kids_First_Participant_ID", "descriptors"))

# Add metadata from pbta to patient list
metadata <- patient_list %>% 
  left_join(pbta, by = "Kids_First_Participant_ID")

# Read scores file
scores_file <- readr::read_tsv(scores_file) 

######## Find out unique RNA library types
rna_library_list <- scores_file %>% pull(data_type) %>% unique()

######## Define output files for each unique rna library
cancer_group_anova_outpaths <- lapply(rna_library_list, function(x){
  x<-gsub(" ", "_", x)
  x<-stringr::str_to_lower(gsub("-", "", x))
  file.path(results_dir, paste0("gsva_anova_", x, "_cancer_group.tsv"))
})

cancer_group_tukey_outpaths <- lapply(rna_library_list, function(x){
  x<-gsub(" ", "_", x)
  x<-stringr::str_to_lower(gsub("-", "", x))
  file.path(results_dir, paste0("gsva_tukey_", x, "_cancer_group.tsv"))
})

tumor_descriptor_anova_outpaths <- lapply(rna_library_list, function(x){
  x<-gsub(" ", "_", x)
  x<-stringr::str_to_lower(gsub("-", "", x))
  file.path(results_dir, paste0("gsva_anova_", x, "_tumor_descriptor.tsv"))
})

tumor_descriptor_tukey_outpaths <- lapply(rna_library_list, function(x){
  x<-gsub(" ", "_", x)
  x<-stringr::str_to_lower(gsub("-", "", x))
  file.path(results_dir, paste0("gsva_tukey_", x, "_tumor_descriptor.tsv"))
})

```


### ANOVA and Tukey analysis of GSVA scores

Here we perform a series of ANOVAs, for polyA and stranded libraries separately, to determine whether mean GSVA scores for a given grouping are significantly different across hallmarks (pathways). The given groupings examined here are `cancer_group` and `tumor_descriptor`. 
In other words, we perform an ANOVA (and associated posthoc Tukey test) for each hallmark as, `hallmark ~ grouping`. Users can specify the grouping variable.

First, prepare the data for modeling:

```{r aov-prep}

### Merge histology metadata with each set of gsea scores

metadata_with_gsva <- metadata %>%
  inner_join(scores_file, by = "Kids_First_Biospecimen_ID") 
```


Now, model:

```{r, aov-perform, fig.width = 25, fig.height = 10, fig.fullwidth = TRUE, echo =TRUE}
for(i in 1:length(rna_library_list)){
  
  rna_library = rna_library_list[i]
  # find out the number of cancer group with this RNA library
  cancer_group_n <- metadata_with_gsva %>%
    dplyr::filter(data_type == rna_library) %>%
    dplyr::pull(cancer_group) %>% 
    unique() %>% length()
  
  # find out the number of harmonized diagnosis with this RNA library
  tumor_descriptor_n <- metadata_with_gsva %>%
    dplyr::filter(data_type == rna_library) %>%
    dplyr::pull(tumor_descriptor) %>% 
    unique() %>% length()
  
  # anova can only be run on factors with >=2 levels, so to avoid error, we give a if statement 
  if(cancer_group_n>=2){
    cancer_group_model_results <- gsva_anova_tukey(metadata_with_gsva, cancer_group, rna_library, SIGNIFICANCE_THRESHOLD)
    
    # plot results
    p <- print(ggplot(cancer_group_model_results[["tukey"]], aes(x = hallmark_name, 
                                y = tukey_p_value,
                                fill = tukey_p_value)) +  
               geom_col(position = position_stack(reverse = TRUE)) +
               rotate() +
               geom_bar(stat = "identity", width = 0.5) + 
               scale_fill_gradient(low="blue",high="red") +
               #scale_fill_manual(values = palette, breaks = sort(names(palette))) + 
               theme_Publication() + 
               theme(axis.text.x = element_text(angle = 85, 
                                                hjust = 1, 
                                                vjust = 1)) + 
               ggplot2::facet_wrap(~comparison, ncol = 3) +
               labs(title = "Hallmark pathways") + 
               labs(x = "Hallmark pathways", y = "Count") +
               ylim(0, 3)) 
  
    # Save the plot
    ggsave(filename = paste0(rna_library_list[i], "-cancer_group_model_results_tukey.pdf"), 
           path = plots_dir, 
           width = 25, 
           height = 10, 
           device = "pdf", 
           useDingbats = FALSE)
  
   # plot results
    p <- print(ggplot(cancer_group_model_results[["anova"]], aes(x = hallmark_name, 
                                y = anova_p_value,
                                fill = anova_p_value)) +  
               geom_col(position = position_stack(reverse = TRUE)) +
               rotate() +
               geom_bar(stat = "identity", width = 0.5) + 
               scale_fill_gradient(low="blue",high="red") +
               #scale_fill_manual(values = palette, breaks = sort(names(palette))) + 
               theme_Publication() + 
               theme(axis.text.x = element_text(angle = 85, 
                                                hjust = 1, 
                                                vjust = 1)) + 
               #ggplot2::facet_wrap(~comparison, ncol = 3) +
               labs(title = "Hallmark pathways") + 
               labs(x = "Hallmark pathways", y = "Count") +
               ylim(0, 3)) 
  
    # Save the plot
    ggsave(filename = paste0(rna_library_list[i], "-cancer_group_model_results_anova.pdf"), 
           path = plots_dir, 
           width = 25, 
           height = 10, 
           device = "pdf", 
           useDingbats = FALSE)
    
    
    # write out results
    readr::write_tsv(cancer_group_model_results[["anova"]], cancer_group_anova_outpaths[[i]])
    readr::write_tsv(cancer_group_model_results[["tukey"]],  cancer_group_tukey_outpaths[[i]])
    
    # print results for viewing 
    print(rna_library)
    print(head(cancer_group_model_results))
  }
  
  # anova can only be run on factors with >=2 levels, so to avoid error, we give a if statement
  if(tumor_descriptor_n>=2){
  tumor_descriptor_model_results <- gsva_anova_tukey(metadata_with_gsva, tumor_descriptor, rna_library, SIGNIFICANCE_THRESHOLD)
  
  # plot results
    p <- print(ggplot(tumor_descriptor_model_results[["tukey"]], aes(x = hallmark_name, 
                                y = tukey_p_value,
                                fill = tukey_p_value)) +  
               geom_col(position = position_stack(reverse = TRUE)) +
               rotate() +
               geom_bar(stat = "identity", width = 0.5) + 
               scale_fill_gradient(low="blue",high="red") +
               #scale_fill_manual(values = palette, breaks = sort(names(palette))) + 
               theme_Publication() + 
               theme(axis.text.x = element_text(angle = 85, 
                                                hjust = 1, 
                                                vjust = 1)) + 
               ggplot2::facet_wrap(~comparison, ncol = 3) +
               labs(title = "Hallmark pathways") + 
               labs(x = "Hallmark pathways", y = "Count") +
               ylim(0, 3)) 
  
    # Save the plot
    ggsave(filename = paste0(rna_library_list[i], "-tumor_descriptor_model_results_tukey.pdf"), 
           path = plots_dir, 
           width = 25, 
           height = 10, 
           device = "pdf", 
           useDingbats = FALSE)
  
   # plot results
    p <- print(ggplot(tumor_descriptor_model_results[["anova"]], aes(x = hallmark_name, 
                                y = anova_p_value,
                                fill = anova_p_value)) +  
               geom_col(position = position_stack(reverse = TRUE)) +
               rotate() +
               geom_bar(stat = "identity", width = 0.5) + 
               scale_fill_gradient(low="blue",high="red") +
               #scale_fill_manual(values = palette, breaks = sort(names(palette))) + 
               theme_Publication() + 
               theme(axis.text.x = element_text(angle = 85, 
                                                hjust = 1, 
                                                vjust = 1)) + 
               #ggplot2::facet_wrap(~comparison, ncol = 3) +
               labs(title = "Hallmark pathways") + 
               labs(x = "Hallmark pathways", y = "Count") +
               ylim(0, 3)) 
  
    # Save the plot
    ggsave(filename = paste0(rna_library_list[i], "-tumor_descriptor_model_results_anova.pdf"), 
           path = plots_dir, 
           width = 25, 
           height = 10, 
           device = "pdf", 
           useDingbats = FALSE)
    
  # write out results
  readr::write_tsv(tumor_descriptor_model_results[["anova"]], tumor_descriptor_anova_outpaths[[i]])
  readr::write_tsv(tumor_descriptor_model_results[["tukey"]], tumor_descriptor_tukey_outpaths[[i]])
  
  # print results for viewing
  print(rna_library)
  print(head(tumor_descriptor_model_results))
  }
}

# CI part is commented out since we want to run in CAVATICA later
### Don't run polyA samples in CI due to data limitations, won't be enough levels for ANOVA
# if (!(running_in_ci)){
#   cancer_group_polya_model_results <- gsva_anova_tukey(metadata_with_gsva, cancer_group, 'polya', SIGNIFICANCE_THRESHOLD)
#   tumor_descriptor_polya_model_results <- gsva_anova_tukey(metadata_with_gsva, tumor_descriptor, 'polya', SIGNIFICANCE_THRESHOLD)
#   head(cancer_group_polya_model_results)
#   head(tumor_descriptor_polya_model_results)
# 
#   ### Save polya library results
#   write_tsv(cancer_group_polya_model_results[["anova"]],     file_anova_polya_cancer_group)
#   write_tsv(tumor_descriptor_polya_model_results[["anova"]],    file_anova_polya_tumor_descriptor)
#   write_tsv(cancer_group_polya_model_results[["tukey"]],     file_tukey_polya_cancer_group)
#   write_tsv(tumor_descriptor_polya_model_results[["tukey"]],    file_tukey_polya_tumor_descriptor)
#   }

```

**How many `cancer_group` have significant ANOVAs across hallmark pathways, for each library?**
```{r}

for(i in 1:length(rna_library_list)){
  
  rna_library = rna_library_list[i]
   cancer_group_n <- metadata_with_gsva %>%
    filter(data_type == rna_library) %>%
    pull(cancer_group) %>% 
    unique() %>% length()
  
  if(cancer_group_n>=2 ){
    cancer_group_model_results <- gsva_anova_tukey(metadata_with_gsva, cancer_group, rna_library, SIGNIFICANCE_THRESHOLD)
    num_sig <- cancer_group_model_results[["anova"]] %>% 
      count(significant_anova) 
    
    # print out information
   num_sig
  }
}

# CI part is commented out since we want to run in CAVATICA later
# if (!(running_in_ci))
# {
#   cancer_group_polya_model_results[["anova"]] %>% count(significant_anova)
# }
```

> All are significantly different for stranded, none for polya. Likely due to data/power limitations.

**How many `tumor_descriptor` have significant ANOVAs across hallmark pathways, for each library?**
```{r}
for(i in 1:length(rna_library_list)){
  # model for each library type
  rna_library = rna_library_list[i]
  
  tumor_descriptor_n <- metadata_with_gsva %>%
    filter(data_type == rna_library) %>%
    pull(tumor_descriptor) %>% 
    unique() %>% length()
  
  if(tumor_descriptor_n>=2){
    tumor_descriptor_model_results <- gsva_anova_tukey(metadata_with_gsva, tumor_descriptor, rna_library, SIGNIFICANCE_THRESHOLD)
    # count the number of significant 
    num_sig <- tumor_descriptor_model_results[["anova"]] %>% 
      count(significant_anova) 
  # print out information
  num_sig
  }
}

# CI part is commented out since we want to run in CAVATICA later
# if (!(running_in_ci))
# {
#   tumor_descriptor_polya_model_results[["anova"]] %>% count(significant_anova)
# }
```


**Comparison of Diagnosis vs Progressive across hallmark pathways, for each library?**
```{r plot-comparisons-Dx-Pro}
for(i in 1:length(rna_library_list)){
  # model for each library type
  rna_library = rna_library_list[i]

  EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue') 
  
  
  
   
top_up <- DE_genes_volcano[[names(seurat_obj.list)[i]]] %>% dplyr::filter(avg_log2FC > 0.25) %>% dplyr::arrange(p_val_adj)
      top_down <- DE_genes_volcano[[names(seurat_obj.list)[i]]] %>% dplyr::filter(avg_log2FC < 0.25) %>% dplyr::arrange(avg_log2FC) %>% dplyr::arrange(p_val_adj)

      p <- EnhancedVolcano(DE_genes_volcano[[names(seurat_obj.list)[i]]],
        lab = rownames(DE_genes_volcano[[names(seurat_obj.list)[i]]]),
        x = 'avg_log2FC',
        y = 'p_val_adj',
        selectLab= c(rownames(top_up)[1:8],rownames(top_down)[1:8]),
        title = paste0(names(seurat_obj.list)[i],' - ',a,' NPM1',n,' vs Controls'),
        drawConnectors=TRUE,
        col=c("black","black","black","red3"),
        ylab = bquote(~-Log[10]~ 'P adj'),
        subtitle = NULL,
        pCutoff = 5e-2,
        FCcutoff=0.25)

      print(p)
      ggsave(paste0("Volcano_plot_",names(seurat_obj.list)[i],"_",n,"_",a,"_NPM1_vs_Controls_MAST.pdf"), plot=p, path=temp_dir, device="pdf", height=7, width=10)

    # Heatmap_top_up_50
      # add gene
      top_up <- cbind(rownames(top_up),top_up)
      rownames(top_up) <- NULL
      colnames(top_up) <- c(names(top_up)) #to not write all the column names
      colnames(top_up)[1] <- "gene"
      head(top_up)
    
      top50up <- top_up[1:50,]
      print(DoHeatmap(seurat_obj.list[[names(seurat_obj.list)[i]]], features = top50up$gene, 
          group.by= "samples",
          label = TRUE))
      ggsave(paste0("Heatmap_top_up_50_",names(seurat_obj.list)[i],"_",n,"_",a,"_NPM1_vs_Controls_MAST.pdf"), path=temp_dir, device="pdf", height=7, width=10)
    
    # Heatmap_top_down_50
      # add gene
      top_down <- cbind(rownames(top_down),top_down)
      rownames(top_down) <- NULL
      colnames(top_down) <- c(names(top_down)) #to not write all the column names
      colnames(top_down)[1] <- "gene"
      head(top_down)

      top50down <- top_down[1:50,]
      print(DoHeatmap(seurat_obj.list[[names(seurat_obj.list)[i]]], features = top50down$gene, 
          group.by= "samples",
          label = TRUE))
      ggsave(paste0("Heatmap_top_down_50_",names(seurat_obj.list)[i],"_",n,"_",a,"_NPM1_vs_Controls_MAST.pdf"), path=temp_dir, device="pdf", height=7, width=10)

    # Heatmap_top_up_down_25
      top25up <- top_up[1:25,]
      top25down <- top_down[1:25,]
      genes <- rbind(top25up$gene, top25down$gene)  
      print(DoHeatmap(seurat_obj.list[[names(seurat_obj.list)[i]]], features = genes, 
          group.by= "samples",
          label = TRUE))
      ggsave(paste0("Heatmap_top_up_down_25_",names(seurat_obj.list)[i],"_",n,"_",a,"_NPM1_vs_Controls_MAST.pdf"), path=temp_dir, device="pdf", height=7, width=10)


 

    
}
```


**Comparison of Diagnosis vs Deceased across hallmark pathways, for each library?**
```{r plot-comparisons-Dx-Dec}


```



```{r}
sessionInfo()
```



