---
title: "GSVA Score Modeling"
author: "Antonia Chroni <chronia@chop.edu> for D3B"
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
params:
  plot_ci: yes
  is_ci: FALSE
---

### Purpose

The purpose of this analysis is to assess significant differences in GSVA scores for each hallmark pathways. Using ANOVA and subsequent Tukey tests, we ask:

+ For each pathway, are GSVA scores significantly different across `tumor_descriptor`? If so, which timepoints are significantly different?

We perform this using both GSVA scores calculated from RNA-seq libraries. Code is also flexible enough to test a different variable besides `tumor_descriptor`, etc.


Plots are inspired from here: https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html

### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('01-model-gsea.Rmd', clean = TRUE)" 
```
_This assumes you are in the top directory of the repository._

### Setup

Load libraries and define certain constants:

```{r, lib-load, warning=FALSE, message=FALSE}
library(tidyverse)
library(broom)
library(ggpubr)
library(pheatmap)

`%>%` <- dplyr::`%>%`

# This script contains functions used to modeling GSVA scores
source(file.path("util", "hallmark_models.R"))

# Significance testing universal threshold
SIGNIFICANCE_THRESHOLD <- 0.01

# Assigning params$is_ci to running_in_ci avoids a locked binding error
running_in_ci <- params$is_ci

# Are we testing? In case of a non 0/1 number, we recast as logical, and then ensure logical.
if (running_in_ci %in% c(0,1)) running_in_ci <- as.logical(running_in_ci)
if (!(is.logical(running_in_ci)))
{
  stop("\n\nERROR: The parameter `is_ci` should be FALSE/TRUE (or 0/1).")
}

```


Next, define directories and load data files:
```{r, data-load, message=FALSE, warning=FALSE}
### Define directories
# Establish base dir
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir    <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "gene-set-enrichment-analysis")

files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")
input_dir <- file.path(analysis_dir, "input")

######### Define input files
## Metadata file (histologies/clinical data)
metadata_file <- file.path(files_dir, "pbta.tsv") # file from add-sample-distribution module
matched_genomic_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") # file from add-sample-distribution module

## GSEA scores
scores_file <- file.path(input_dir, "gsva_scores.tsv")

# File path to results directory
results_dir <-
  file.path(analysis_dir, "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# File path to plot directory
plots_dir <-
  file.path(analysis_dir, "plots")
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

source(paste0(root_dir, "/figures/scripts/theme.R"))

######## Load input files
pbta <- readr::read_tsv(metadata_file, guess_max = 100000) %>% 
  filter(experimental_strategy == "RNA-Seq") %>% 
  filter(!is.na(RNA_library))


# Read in matched_genomic_file and get the list of patients 
# with matched time points for the genomic assays
patient_list <- readr::read_tsv(matched_genomic_file, guess_max = 100000, show_col_types = FALSE) %>%
  subset(select = c("Kids_First_Participant_ID", "descriptors"))

# Add metadata from pbta to patient list
metadata <- patient_list %>% 
  left_join(pbta, by = "Kids_First_Participant_ID")

# Read scores file
scores_file <- readr::read_tsv(scores_file) 

######## Find out unique RNA library types
rna_library_list <- scores_file %>% pull(data_type) %>% unique()

######## Define output files for each unique rna library
cancer_group_anova_outpaths <- lapply(rna_library_list, function(x){
  x<-gsub(" ", "_", x)
  x<-stringr::str_to_lower(gsub("-", "", x))
  file.path(results_dir, paste0("gsva_anova_", x, "_cancer_group.tsv"))
})

cancer_group_tukey_outpaths <- lapply(rna_library_list, function(x){
  x<-gsub(" ", "_", x)
  x<-stringr::str_to_lower(gsub("-", "", x))
  file.path(results_dir, paste0("gsva_tukey_", x, "_cancer_group.tsv"))
})

tumor_descriptor_anova_outpaths <- lapply(rna_library_list, function(x){
  x<-gsub(" ", "_", x)
  x<-stringr::str_to_lower(gsub("-", "", x))
  file.path(results_dir, paste0("gsva_anova_", x, "_tumor_descriptor.tsv"))
})

tumor_descriptor_tukey_outpaths <- lapply(rna_library_list, function(x){
  x<-gsub(" ", "_", x)
  x<-stringr::str_to_lower(gsub("-", "", x))
  file.path(results_dir, paste0("gsva_tukey_", x, "_tumor_descriptor.tsv"))
})

```


### ANOVA and Tukey analysis of GSVA scores

Here we perform a series of ANOVAs, for polyA and stranded libraries separately, to determine whether mean GSVA scores for a given grouping are significantly different across hallmarks (pathways). The given groupings examined here are `cancer_group` and `tumor_descriptor`. 
In other words, we perform an ANOVA (and associated posthoc Tukey test) for each hallmark as, `hallmark ~ grouping`. Users can specify the grouping variable.

First, prepare the data for modeling:

```{r aov-prep}

### Merge histology metadata with each set of gsea scores

metadata_with_gsva <- metadata %>%
  inner_join(scores_file, by = "Kids_First_Biospecimen_ID") 
```


Now, model:

```{r, aov-perform, fig.width = 12, fig.height = 10, fig.fullwidth = TRUE, echo = TRUE}
for(i in 1:length(rna_library_list)){
  
  rna_library = rna_library_list[i]
  
  # find out the number of `tumor_descriptor` with this RNA library
  tumor_descriptor_n <- metadata_with_gsva %>%
    dplyr::filter(data_type == rna_library) %>%
    dplyr::pull(tumor_descriptor) %>% 
    unique() %>% length()
  
  # anova can only be run on factors with >=2 levels, so to avoid error, we give a if statement
  if(tumor_descriptor_n>=2){
  tumor_descriptor_model_results <- gsva_anova_tukey(metadata_with_gsva, tumor_descriptor, rna_library, SIGNIFICANCE_THRESHOLD)
  
  # plot results
  # tukey
  name <- paste0(plots_dir, "/",rna_library_list[i], "-tumor_descriptor_tukey_barplot.pdf")
  print(name)
  p <- create_barplot(df = tumor_descriptor_model_results[["tukey"]],
                      y_value = tumor_descriptor_model_results[["tukey"]]$tukey_p_value,
                      title_value = "tukey_p_value")
  pdf(file = name, width = 12, height = 10)
  print(p)
  dev.off()
  
  # anova
  name <- paste0(plots_dir, "/",rna_library_list[i], "-tumor_descriptor_anova_barplot.pdf")
  print(name)
  p <- create_barplot(df = tumor_descriptor_model_results[["anova"]],
                      y_value = tumor_descriptor_model_results[["anova"]]$anova_p_value,
                      title_value = "anova_p_value")
  pdf(file = name, width = 12, height = 10)
  print(p)
  dev.off()
  
  
  
    
   ################ dot plots ########################################################
   # plot results
    p <- print(ggplot(tumor_descriptor_model_results[["tukey"]], 
                       mapping = aes(x = reorder(hallmark_name, tukey_p_value), tukey_p_value)) +  
              # geom_dotplot(binwidth = 1.5)  +
               geom_line(aes(group = hallmark_name)) +
               geom_point(aes(color = tukey_p_value, size = tukey_p_value)) +
               rotate() +
               scale_color_gradient(low = "blue", high = "red") +
               #scale_fill_manual(values = palette, breaks = sort(names(palette))) + 
               theme_Publication() + 
               theme(axis.text.x = element_text(angle = 85, 
                                                hjust = 1, 
                                                vjust = 1)) + 
               ggplot2::facet_wrap(~comparison, ncol = 3) +
               labs(title = paste0("Tukey dotplot for ", rna_library_list[i], " using tukey")) + 
               labs(x = "", y = "tukey_p_value") +
               ylim(0, 1)) 
  
    # Save the plot
    ggsave(filename = paste0(rna_library_list[i], "-tumor_descriptor_tukey_dotplot.pdf"), 
           path = plots_dir, 
           width = 12, 
           height = 10, 
           device = "pdf", 
           useDingbats = FALSE)
  
   # plot results
   p <- print(ggplot(tumor_descriptor_model_results[["anova"]], 
                     mapping = aes(x = reorder(hallmark_name, anova_p_value), anova_p_value)) +  
              # geom_dotplot(binwidth = 1.5)  +
               geom_line(aes(group = hallmark_name)) +
               geom_point(aes(color = anova_p_value, size = anova_p_value)) +
               rotate() +
               scale_color_gradient(low = "blue", high = "red") +
               #scale_fill_manual(values = palette, breaks = sort(names(palette))) + 
               theme_Publication() + 
               theme(axis.text.x = element_text(angle = 85, 
                                                hjust = 1, 
                                                vjust = 1)) + 
               #ggplot2::facet_wrap(~comparison, ncol = 3) +
               labs(title = paste0("Dotplot for ", rna_library_list[i], " using anova")) + 
               labs(x = "", y = "anova_p_value") +
               ylim(0, 1)) 
  
  
    # Save the plot
    ggsave(filename = paste0(rna_library_list[i], "-tumor_descriptor_anova_dotplot.pdf"), 
           path = plots_dir, 
           width = 12, 
           height = 10, 
           device = "pdf", 
           useDingbats = FALSE)  
    
    
   
    
    
    
    
    
    
    
  # write out results
  readr::write_tsv(tumor_descriptor_model_results[["anova"]], tumor_descriptor_anova_outpaths[[i]])
  readr::write_tsv(tumor_descriptor_model_results[["tukey"]], tumor_descriptor_tukey_outpaths[[i]])
  
  # print results for viewing
  print(rna_library)
  print(head(tumor_descriptor_model_results))
  }
}
```


```{r}
sessionInfo()
```

