---
title: "VAF estimation of tumors across multiple timepoints for the autopsy samples in the PBTA Cohort"
author: 'Antonia Chroni <chronia@chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---

#### Tumor evolution project 

### Data used 
In this notebook, we are looking into the v12 histologies file (last updates on May 1st, 2023) and subset to the PBTA cohort.

# Set up
```{r load-library}
suppressPackageStartupMessages({
  library(tidyverse)
})
```

## Directories and File Inputs/Outputs
```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will be in the project root directory
# Use this as the root directory to ensure proper sourcing of functions 
# no matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
scratch_dir <- file.path(root_dir, "scratch")
data_dir <- file.path(root_dir, "data")
analysis_dir <- file.path(root_dir, "analyses", "tmb-vaf-longitudinal")
input_dir <- file.path(analysis_dir, "input")
files_dir <- file.path(root_dir, "analyses", "sample-distribution-analysis", "results")

# Input files
v12_file <- file.path(data_dir, "histologies.tsv")
maf_file <- file.path(data_dir, "snv-consensus-plus-hotspots.maf.tsv.gz")
tmb_file <- file.path(input_dir, "snv-mutation-tmb-coding.tsv")
matched_time_points_file <- file.path(files_dir, "genomic_assays_matched_time_points.tsv") # file from add-sample-distribution module
```

## Read in data and process
```{r load-process-inputs-please-wait}
# Read in histologies file and filter for the pbta cohort
pbta <- readr::read_tsv(v12_file, guess_max = 100000, show_col_types = FALSE) %>%
  filter(cohort == "PBTA",
         !is.na(pathology_diagnosis),
         !composition %in% c("Derived Cell Line", "PDX"),
         !experimental_strategy %in% c("Methylation", "Targeted Sequencing")) %>%
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, experimental_strategy, tumor_descriptor, age_at_diagnosis_days, age_last_update_days, cancer_predispositions, broad_histology, short_histology, cancer_group) 

# Read in tmb file
tmb <- readr::read_tsv(tmb_file, guess_max = 100000, show_col_types = FALSE) %>%
  dplyr::rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) # change name of the biospecimen to match the one from the histologies files

# Make a vector of coding gene mutations
maf_nonsynonymous <- c(
  "Missense_Mutation",
  "Frame_Shift_Del",
  "In_Frame_Ins",
  "Frame_Shift_Ins",
  "Splice_Site",
  "Nonsense_Mutation",
  "In_Frame_Del",
  "Nonstop_Mutation",
  "Translation_Start_Site"
)

# Read in maf file, calculate VAF, and retain only coding gene mutations
# We will use this column to label gene mutations in the final graphs
maf <- data.table::fread(maf_file, data.table = FALSE) %>% 
  filter(Tumor_Sample_Barcode %in% pbta$Kids_First_Biospecimen_ID,
         Variant_Classification %in% maf_nonsynonymous) %>%
  dplyr::rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>% # change name of the biospecimen to match the one from the histologies files
  dplyr::mutate(VAF = t_alt_count / (t_ref_count + t_alt_count), # calculate VAF 
                gene_protein = paste(Hugo_Symbol, HGVSp_Short, sep = "_")) %>%  # concatenate gene and protein information
  select(Kids_First_Biospecimen_ID, Hugo_Symbol, Chromosome, Start_Position, Variant_Classification, VAF, HGVSc, HGVSp, HGVSp_Short, gene_protein) 

# Read in matched_time_points_file and get the list of patients 
# with matched time points for the genomic assays
patient_list <- readr::read_tsv(matched_time_points_file, guess_max = 100000, show_col_types = FALSE) %>%
  subset(select = c("Kids_First_Participant_ID", "descriptors"))

# How many patients have paired time samples?
paired_time_samples <- length(unique(patient_list$Kids_First_Participant_ID))

# Add metadata from pbta to patient list
patient_list_metadata <- patient_list %>% 
  left_join(pbta, by = "Kids_First_Participant_ID") %>%
  filter(!(experimental_strategy == "RNA-Seq")) %>%
  mutate(tumor_descriptor = case_when(grepl("Primary Tumor|Initial CNS Tumor", tumor_descriptor) ~ "Diagnosis", 
                                      grepl("Progressive Disease Post-Mortem", tumor_descriptor) ~ "Deceased", 
                                      TRUE ~ tumor_descriptor)) 
```

We are interested in how VAF values change over the time and specifically for the autopsy samples.
We will create corplots by keeping Deceased timepoint as constant and any other timepoint paired with it will be the variable.

# Create maf file for autopsy samples
```{r maf_autopsy}
# First, we will filter the df to keep paired samples with Deceased timepoint (autopsy samples).
# There are patient cases with more than one time points in addition to Deceased, e.g. Diagnosis and Progressive.
# There are also multiple biospecimen samples per timepoint and patient case.
# So, next we will create unique identifiers for each of those cases `match_id_bs`.
# Next we will leverage this information to create one column for the constant variable (`timepoints_deceased`)
# and one column for the variable (`timepoints_other`)
# We will replace wiht numbers any timpoint not relevant to those columns, 
# e.g., there should be only deceased samples in the `timepoints_deceased` so we will replace with numbers 1-3
# Same logic applies for `timepoints_other` only there, we have one value to replace (used 4 for continuity).
# This system allows us to build a timepoint model to be used for loops that will iterate 
# and create graphs based on any possible combination between timepoints and biospecimen samples per `Kids_First_Participant_ID`.

maf_autopsy <- patient_list_metadata %>%
  mutate(Deceased = case_when(grepl("Deceased", descriptors) ~ "Deceased")) %>%
  filter(Deceased == "Deceased") %>% 
  left_join(maf, by = "Kids_First_Biospecimen_ID") %>%
  filter(!is.na(Kids_First_Participant_ID),
         !is.na(VAF),
         !VAF == 0, 
         !gene_protein == "Unknown") %>%
  select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID, descriptors, tumor_descriptor, experimental_strategy, Hugo_Symbol, gene_protein, VAF) %>%
  mutate(match_id = paste(Kids_First_Biospecimen_ID, tumor_descriptor, gene_protein, VAF, sep = "_"),
         match_id_bs = paste(tumor_descriptor, Kids_First_Biospecimen_ID, sep = "_"),
         timepoints_deceased = match_id_bs,
         timepoints_deceased = case_when(grepl("Diagnosis", match_id_bs) ~ "1", 
                                         grepl("Recurrence", match_id_bs) ~ "2",
                                         grepl("Progressive", match_id_bs) ~ "3",
                                         TRUE ~ timepoints_deceased),
         timepoints_other = match_id_bs,
         timepoints_other = case_when(grepl("Deceased", match_id_bs) ~ "4",
                                      TRUE ~ timepoints_other)) %>%
  unique() %>%
  write_tsv(file.path(scratch_dir, "maf_autopsy.tsv"))

# How many patients have paired time samples with autopsy time point present?
paired_time_samples_autopsy <- length(unique(maf_autopsy$Kids_First_Participant_ID))
```
## How many patient cases with autopsy samples in the paired longitudinal data?
There are `r paired_time_samples_autopsy` autopsy samples out of the total `r paired_time_samples` patient cases.

# Create tmb file for all patient samples with genomic assays
```{r tmb_all}
# We will create similar unique identifiers as we did for the maf file.
# We will also create identifiers for each timepoint (timepoints_deceased and timepoints_other).
tmb_genomic <- patient_list_metadata %>% 
  left_join(tmb, by = c("Kids_First_Biospecimen_ID", "experimental_strategy")) %>%
  filter(!is.na(Kids_First_Participant_ID),
         !is.na(tmb)) %>%
   mutate(match_id_bs = paste(tumor_descriptor, Kids_First_Biospecimen_ID, sep = "_"),
          timepoints_deceased = match_id_bs,
          timepoints_deceased = case_when(grepl("Diagnosis", match_id_bs) ~ "1",
                                          grepl("Recurrence", match_id_bs) ~ "2",
                                          grepl("Progressive", match_id_bs) ~ "3",
                                          TRUE ~ timepoints_deceased),
          timepoints_other = match_id_bs,
          timepoints_other = case_when(grepl("Deceased", match_id_bs) ~ "4",
                                       TRUE ~ timepoints_other),
          log2_mutation_count = log2(mutation_count)) 

paired_time_samples_tmb_genomic <- length(unique(tmb_genomic$Kids_First_Participant_ID))

# There might be missign tmb information for some `Kids_First_Biospecimen_ID`
# This will reuslt into cases with no paired `tumor_descriptor` per `Kids_First_Participant_ID`
# Let's account for this and exclude thse samples
paired_tumor_descriptor_list <- tmb_genomic %>%
  unique() %>% 
  arrange(Kids_First_Participant_ID,tumor_descriptor)  %>%
  group_by(Kids_First_Participant_ID) %>%
  summarise(tumor_descriptor_sum = str_c(tumor_descriptor, collapse = ";")) %>% 
  filter(!tumor_descriptor_sum %in% c("Diagnosis", "Progressive", "Recurrence"))

tmb_genomic_filter <- paired_tumor_descriptor_list %>% 
  left_join(tmb_genomic, by = "Kids_First_Participant_ID") %>%
  write_tsv(file.path(scratch_dir, "tmb_genomic.tsv"))

paired_time_samples_tmb_genomic_filter <- length(unique(tmb_genomic_filter$Kids_First_Participant_ID))
```

## How many patient cases with paired time and autopsy samples and TMB data
There are `r paired_time_samples_tmb_genomic_filter` (out of the `r paired_time_samples_tmb_genomic`) patient cases with TMB information.
Any discrepancy in patient numbers between maf and tmb data might be caused from an issue related to the MAF file (see https://github.com/d3b-center/bixu-tracker/issues/2049).

```{r echo=TRUE}
sessionInfo()
```
